<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq analysis with Bioconductor: Introduction to RNA-seq</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team." style="text-decoration: unset">
           
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link" style="color: #383838">Beta
            <i aria-hidden="true" class="icon" data-feather="alert-circle" style="color: #001483; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
        </abbr>

      </div>
    </div>
    <div class="selector-container ">


      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/01-intro-to-rnaseq.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq analysis with Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq analysis with Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq analysis with Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 0%" class="percentage">
    0%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 0%" aria-valuenow="0" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/01-intro-to-rnaseq.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->

            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        1. Introduction to RNA-seq
        </span>
      
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-setup.html">2. RStudio Project and Experimental Data</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-import-annotate.html">3. Importing and annotating quantified data into R</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-exploratory-qc.html">4. Exploratory analysis and quality control</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-differential-expression.html">5. Differential expression analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-extra-design.html">6. Extra exploration of design matrices</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-gene-set-analysis.html">7. Gene set enrichment analysis</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-next-steps.html">8. Next steps</a>
    </div><!--/div.accordion-header-->

  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html"></a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">

            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="index.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="02-setup.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="index.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="02-setup.html" rel="next">Next: RStudio Project and... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Introduction to RNA-seq</h1>
        <p> Last updated on 2025-07-01 |

        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/01-intro-to-rnaseq.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>



        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>What are the different choices to consider when planning an RNA-seq
experiment?</li>
<li>How does one process the raw fastq files to generate a table with
read counts per gene and sample?</li>
<li>Where does one find information about annotated genes for a given
organism?</li>
<li>What are the typical steps in an RNA-seq analysis?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain what RNA-seq is.</li>
<li>Describe some of the most common design choices that have to be made
before running an RNA-seq experiment.</li>
<li>Provide an overview of the procedure to go from the raw data to the
read count matrix that will be used for downstream analysis.</li>
<li>Show some common types of results and visualizations generated in
RNA-seq analyses.</li>
</ul></div>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="what-are-we-measuring-in-an-rna-seq-experiment">What are we measuring in an RNA-seq experiment?<a class="anchor" aria-label="anchor" href="#what-are-we-measuring-in-an-rna-seq-experiment"></a></h1>
<figure><img src="fig/alternative_splicing_sub.png" alt="Illustration of part of the central dogma of molecular biology, where DNA is transcribed to RNA, and intronic sequences are spliced out" class="figure mx-auto d-block"></figure><p>In order to produce an RNA molecule, a stretch of DNA is first
transcribed into mRNA. Subsequently, intronic regions are spliced out,
and exonic regions are combined into different <em>isoforms</em> of a
gene.</p>
<figure><img src="fig/rnaseq_martin_wang.png" alt="Illustration of the major experimental steps of an RNA-seq experiment" class="figure mx-auto d-block"></figure><p>(figure adapted from <a href="https://www.nature.com/articles/nrg3068" class="external-link">Martin &amp; Wang
(2011)</a>).</p>
<p>In a typical RNA-seq experiment, RNA molecules are first collected
from a sample of interest. After a potential enrichment for molecules
with polyA tails (predominantly mRNA), or depletion of otherwise highly
abundant ribosomal RNA, the remaining molecules are fragmented into
smaller pieces (there are also long-read protocols where entire
molecules are considered, but those are not the focus of this lesson).
It is important to keep in mind that because of the splicing excluding
intronic sequences, an RNA molecule (and hence a generated fragment) may
not correspond to an uninterrupted region of the genome. The RNA
fragments are then reverse transcribed into cDNA, whereafter sequencing
adapters are added to each end. These adapters allow the fragments to
attach to the flow cell. Once attached, each fragment will be heavily
amplified to generate a cluster of identical sequences on the flow cell.
The sequencer then determines the sequence of the first 50-200
nucleotides of the cDNA fragments in each such cluster, starting from
one end, which corresponds to a <em>read</em>. Many data sets are
generated with so called paired-end protocols, in which the fragments
are read from both ends. Millions of such reads (or pairs of reads) will
be generated in an experiment, and these will be represented in a (pair
of) FASTQ files. Each read is represented by four consecutive lines in
such a file: first a line with a unique read identifier, next the
inferred sequence of the read, then another identifier line, and finally
a line containing the base quality for each inferred nucleotide,
representing the probability that the nucleotide in the corresponding
position has been correctly identified.</p>
<div id="challenge-discuss-the-following-points-with-your-neighbor" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-discuss-the-following-points-with-your-neighbor" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss the following points with your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>What are potential advantages and disadvantages of paired-end
protocols compared to only sequencing one end of a fragment?</li>
<li>What quality assessment can you think of that would be useful to
perform on the FASTQ files with read sequences?</li>
</ol></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="experimental-design-considerations">Experimental design considerations<a class="anchor" aria-label="anchor" href="#experimental-design-considerations"></a></h1>
<p>Before starting to collect data, it is essential to take some time to
think about the design of the experiment. Experimental design concerns
the organization of experiments with the purpose of making sure that the
right type of data, and enough of it, is available to answer the
questions of interest as efficiently as possible. Aspects such as which
conditions or sample groups to consider, how many replicates to collect,
and how to plan the data collection in practice are important questions
to consider. Many high-throughput biological experiments (including
RNA-seq) are sensitive to ambient conditions, and it is often difficult
to directly compare measurements that have been done on different days,
by different analysts, in different centers, or using different batches
of reagents. For this reason, it is very important to design experiments
properly, to make it possible to disentangle different types of (primary
and secondary) effects.</p>
<figure><img src="fig/factors-affecting-measurements.png" alt="A classification of many different factors affecting measurements obtained from an experiment into treatment, biological, technical and error effects" class="figure mx-auto d-block"></figure><p>(figure from <a href="https://www.amazon.com/Experimental-Design-Laboratory-Biologists-Reproducibility/dp/1107424887" class="external-link">Lazic
(2017)</a>).</p>
<div id="challenge-discuss-with-your-neighbor" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-discuss-with-your-neighbor" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss with your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Why is it essential to have replicates?</li>
</ol></div>
</div>
</div>
<p>Importantly, not all replicates are equally useful, from a
statistical point of view. One common way to classify the different
types of replicates is as ‘biological’ and ‘technical’, where the latter
are typically used to test the reproducibility of the measurement
device, while biological replicates inform about the variability between
different samples from a population of interest. Another scheme
classifies replicates (or units) into ‘biological’, ‘experimental’ and
‘observational’. Here, biological units are entities we want to make
inferences about (e.g., animals, persons). Replication of biological
units is required to make a general statement of the effect of a
treatment - we can not draw a conclusion about the effect of drug on a
population of mice by studying a single mouse only. Experimental units
are the smallest entities that can be <em>independently assigned to a
treatment</em> (e.g., animal, litter, cage, well). Only replication of
experimental units constitute true replication. Observational units,
finally, are entities at which measurements are made.</p>
<p>To explore the impact of experimental design on the ability to answer
questions of interest, we are going to use an interactive application,
provided in the <a href="https://csoneson.github.io/ConfoundingExplorer/" class="external-link">ConfoundingExplorer</a>
package.</p>
<div id="challenge" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Launch the ConfoundingExplorer application and familiarize yourself
with the interface.</p>
</div>
</div>
</div>
<div id="challenge-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>For a balanced design (equal distribution of replicates between the
two groups in each batch), what is the effect of increasing the strength
of the batch effect? Does it matter whether one adjusts for the batch
effect or not?</li>
<li>For an increasingly unbalanced design (most or all replicates of one
group coming from one batch), what is the effect of increasing the
strength of the batch effect? Does it matter whether one adjusts for the
batch effect or not?</li>
</ol></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="rna-seq-quantification-from-reads-to-count-matrix">RNA-seq quantification: from reads to count matrix<a class="anchor" aria-label="anchor" href="#rna-seq-quantification-from-reads-to-count-matrix"></a></h1>
<figure><img src="fig/rna-quant.jpeg" alt="Illustration of a set of reads generated by a sequencer, and genomic and transcriptomic reference sequences" class="figure mx-auto d-block"></figure><p>The read sequences contained in the FASTQ files from the sequencer
are typically not directly useful as they are, since we do not have the
information about which gene or transcript they originate from. Thus,
the first processing step is to attempt to identify the location of
origin for each read, and use this to obtain an estimate of the number
of reads originating from a gene (or another features, such as an
individual transcript). This can then be used as a proxy for the
abundance, or expression level, of the gene. There is a plethora of RNA
quantification pipelines, and the most common approaches can be
categorized into three main types:</p>
<ol style="list-style-type: decimal"><li><p>Align reads to the genome, and count the number of reads that map
within the exons of each gene. This is the one of simplest methods. For
species for which the transcriptome is poorly annotated, this would be
the preferred approach. Example: <code>STAR</code> alignment to GRCm39 +
<code>Rsubread</code> <a href="https://doi.org/10.1093%2Fnar%2Fgkz114" class="external-link">featureCounts</a></p></li>
<li><p>Align reads to the transcriptome, quantify transcript expression,
and summarize transcript expression into gene expression. This approach
can produce accurate quantification results (<a href="https://doi.org/10.1186/s13059-016-0940-1" class="external-link">independent
benchmarking</a>), particularly for high-quality samples without DNA
contamination. Example: RSEM quantification using
<code>rsem-calculate-expression --star</code> on the GENCODE GRCh38
transcriptome + <code>tximport</code></p></li>
<li><p>Pseudoalign reads against the transcriptome, using the
corresponding genome as a decoy, quantifying transcript expression in
the process, and summarize the transcript-level expression into
gene-level expression. The advantages of this approach include:
computational efficiency, mitigation of the effect of DNA contamination,
and GC bias correction. Example: <code>salmon quant --gcBias</code> +
<code>tximport</code></p></li>
</ol><p>At typical sequencing read depth, gene expression quantification is
often more accurate than transcript expression quantification. However,
differential gene expression analyses can be <a href="https://doi.org/10.12688/f1000research.7563.1" class="external-link">improved</a> by
having access also to transcript-level quantifications.</p>
<p>Other tools used in RNA-seq quantification include: TopHat2, bowtie2,
kallisto, HTseq, among many others.</p>
<p>The choice of an appropriate RNA-seq quantification depends on the
quality of the transcriptome annotation, the quality of the RNA-seq
library preparation, the presence of contaminating sequences, among many
factors. Often, it can be informative to compare the quantification
results of multiple approaches.</p>
<p>Because the best quantification method is species- and
experiment-dependent, and often requires large amounts of computing
resources, this workshop will not cover any specifics of how to generate
the counts. Instead, we recommend checking out the references above and
consulting with a local bioinformatics expert if you need help.</p>
<div id="challenge-discuss-the-following-points-with-your-neighbor-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-discuss-the-following-points-with-your-neighbor-1" class="callout-inner">
<h3 class="callout-title">Challenge: Discuss the following points with your neighbor</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>Which of the mentioned RNA-Seq quantification tools have you heard
about? Do you know other pros and cons of the methods?</li>
<li>Have you done your own RNA-Seq experiment? If so, what
quantification tool did you use and why did you choose it?</li>
<li>Do you have access to specific tools / local bioinformatics expert /
computational resources for quantification? If you don’t, how might you
gain access?</li>
</ol></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="finding-the-reference-sequences">Finding the reference sequences<a class="anchor" aria-label="anchor" href="#finding-the-reference-sequences"></a></h1>
<p>In order to quantify abundances of known genes or transcripts from
RNA-seq data, we need a reference database informing us of the sequence
of these features, to which we can then compare our reads. This
information can be obtained from different online repositories. It is
highly recommended to choose one of these for a particular project, and
not mix information from different sources. Depending on the
quantification tool you have chosen, you will need different types of
reference information. If you are aligning your reads to the genome and
investigating the overlap with known annotated features, you will need
the full genome sequence (provided in a fasta file) and a file that
tells you the genomic location of each annotated feature (typically
provided in a gtf file). If you are mapping your reads to the
transcriptome, you will instead need a file with the sequence of each
transcript (again, a fasta file).</p>
<ul><li>If you are working with mouse or human samples, the <a href="https://www.gencodegenes.org/" class="external-link">GENCODE project</a> provides
well-curated reference files.</li>
<li>
<a href="https://www.ensembl.org/info/data/ftp/index.html" class="external-link">Ensembl</a>
provides reference files for a large set of organisms, including <a href="https://plants.ensembl.org/info/data/ftp/index.html" class="external-link">plants</a>
and <a href="https://fungi.ensembl.org/info/data/ftp/index.html" class="external-link">fungi</a>.</li>
<li>
<a href="https://hgdownload.soe.ucsc.edu/downloads.html" class="external-link">UCSC</a>
also provides reference files for many organisms.</li>
</ul><div id="challenge-2" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge</h3>
<div class="callout-content">
<p>Download the latest mouse transcriptome fasta file from GENCODE. What
do the entries look like? Tip: to read the file into R, consider the
<code>readDNAStringSet()</code> function from the
<code>Biostrings</code> package.</p>
</div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="where-are-we-heading-towards-in-this-workshop">Where are we heading towards in this workshop?<a class="anchor" aria-label="anchor" href="#where-are-we-heading-towards-in-this-workshop"></a></h1>
<p>During the coming two days, we will discuss and practice how to
perform differential expression analysis with Bioconductor, and how to
interpret the results. We will start from a count matrix, and thus
assume that the initial quality assessment and quantification of gene
expression have already been done. The outcome of a differential
expression analysis is often represented using graphical
representations, such as MA plots and heatmaps (see below for
examples).</p>
<p><img src="fig/maPlot-example.png" alt="An example MA plot" class="figure"><img src="fig/heatmap-example.png" alt="An example heatmap" class="figure"></p>
<p>In the following episodes we will learn, among other things, how to
generate and interpret these plots. It is also common to perform
follow-up analyses to investigate whether there is a functional
relationship among the top-ranked genes, so called gene set (enrichment)
analysis, which will also be covered in a later episode.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Key Points</h3>
<div class="callout-content">
<ul><li>RNA-seq is a technique of measuring the amount of RNA expressed
within a cell/tissue and state at a given time.</li>
<li>Many choices have to be made when planning an RNA-seq experiment,
such as whether to perform poly-A selection or ribosomal depletion,
whether to apply a stranded or an unstranded protocol, and whether to
sequence the reads in a single-end or paired-end fashion. Each of the
choices have consequences for the processing and interpretation of the
data.</li>
<li>Many approaches exist for quantification of RNA-seq data. Some
methods align reads to the genome and count the number of reads
overlapping gene loci. Other methods map reads to the transcriptome and
use a probabilistic approach to estimate the abundance of each gene or
transcript.</li>
<li>Information about annotated genes can be accessed via several
sources, including Ensembl, UCSC and GENCODE.</li>
</ul></div>
</div>
</div>
</div>



      </div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="index.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="02-setup.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="index.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="02-setup.html" rel="next">Next: RStudio Project and... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>

        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/01-intro-to-rnaseq.Rmd" class="external-link">Edit on GitHub</a>

	
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a>
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:charlotte.soneson@fmi.ch">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">

        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>

        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.16.12" class="external-link">sandpaper (0.16.12)</a>,
        <a href="https://github.com/carpentries/pegboard/tree/0.7.9" class="external-link">pegboard (0.7.9)</a>,
      and <a href="https://github.com/mblue9/varnish/tree/3403ad0dd37f7641c6442330fcc182937e7964c5" class="external-link">varnish (0.2.16)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "LearningResource",
  "@id": "https://carpentries-incubator.github.io/bioc-rnaseq/01-intro-to-rnaseq.html",
  "inLanguage": "en",
  "dct:conformsTo": "https://bioschemas.org/profiles/LearningResource/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Introduction to RNA-seq",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-rnaseq/01-intro-to-rnaseq.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-rnaseq/01-intro-to-rnaseq.html",
  "dateCreated": "2020-09-15",
  "dateModified": "2025-07-01",
  "datePublished": "2025-07-01"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

