<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>RNA-seq analysis with Bioconductor: Gene set analysis</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="assets/styles.css"><script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png"><link rel="manifest" href="site.webmanifest"><link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
<!--
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
-->
          <a href="https://bioconductor.github.io/2023-07-31-BiocRnaseq-Bioc2023/" class="external-link alert-link" style="color: #383838">Under active development for BioC2023 workshop July 31 - August 1 2023
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>

          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/06-gene-set-analysis.html';">Instructor View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      RNA-seq analysis with Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            RNA-seq analysis with Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"><li><a class="dropdown-item" href="discuss.html">Discussion</a></li><li><a class="dropdown-item" href="reference.html"></a></li>
          </ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq analysis with Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 62%" class="percentage">
    62%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 62%" aria-valuenow="62" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/06-gene-set-analysis.html">Instructor View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-setup.html">2. RStudio Project and Experimental Data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-import-annotate.html">3. Importing and annotating quantified data into R</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-exploratory-qc.html">4. Exploratory analysis and quality control</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-differential-expression.html">5. Differential expression analysis</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        6. Gene set analysis
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#enrichment-of-a-curated-gene-set">Enrichment of a curated gene set</a></li>
<li><a href="#gene-ontology-analysis-with-clusterprofiler">Gene ontology analysis with clusterProfiler</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-next-steps.html">7. Next steps</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-extra-design.html">8. Extra exploration of design matrices</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li><li><a href="reference.html"></a></li>
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="05-differential-expression.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="07-next-steps.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="05-differential-expression.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Differential</a>
        <a class="chapter-link float-end" href="07-next-steps.html" rel="next">Next: Next steps... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Gene set analysis</h1>
        <p> Last updated on 2023-05-17 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/06-gene-set-analysis.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>Why do we perform gene set enrichment analysis?</li>
<li>How does one retrieve predefined sets of genes with shared
characteristics for gene set testing?</li>
<li>What different classes of methods are available for gene set
analysis?</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Explain how to find differentially expressed pathways with gene set
analysis in R.</li>
<li>Understand how differentially expressed genes can enrich a gene
set.</li>
<li>Explain how to perform a gene set analysis in R, using
clusterProfiler.</li>
</ul></div>
</div>
</div>
</div>
</div>
<div id="contribute" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute"></a>
</h3>
<div class="callout-content">
<p>This episode is intended to introduce the concept of how to carry out
a functional analysis of a subset of differentially expressed (DE)
genes, by means of assessing how significantly DE genes enrich gene sets
of our interest.</p>
</div>
</div>
</div>
<p>First, we are going to explore the basic concept of enriching a gene
set with differentially expressed (DE) genes. Recall the differential
expression analysis.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span><span class="op">[</span>, <span class="va">se</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"Cerebellum"</span><span class="op">]</span>,</span>
<span>                            design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in DESeq2::DESeqDataSet(se[, se$tissue == "Cerebellum"], design = ~sex
+ : some variables in design formula are characters, converting to factors</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<p>Fetch results for the contrast between male and female mice.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resSex</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">resSex</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 32652 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 53, 0.16%
LFC &lt; 0 (down)     : 71, 0.22%
outliers [1]       : 10, 0.031%
low counts [2]     : 13717, 42%
(mean count &lt; 6)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<p>Select DE genes between males and females with FDR &lt; 5%.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sexDE</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resSex</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">sexDE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 54  6</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sexDE</span> <span class="op">&lt;-</span> <span class="va">sexDE</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="fu">abs</span><span class="op">(</span><span class="va">sexDE</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">sexDE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               baseMean log2FoldChange     lfcSE      stat        pvalue
Eif2s3y       1410.8750       12.62514 0.5652155  22.33685 1.620659e-110
Kdm5d          692.1672       12.55386 0.5936267  21.14773  2.895664e-99
Uty            667.4375       12.01728 0.5935911  20.24505  3.927797e-91
Ddx3y         2072.9436       11.87241 0.3974927  29.86825 5.087220e-196
Xist         22603.0359      -11.60429 0.3362822 -34.50761 6.168523e-261
LOC105243748    52.9669        9.08325 0.5976242  15.19893  3.594320e-52
                      padj
Eif2s3y      1.022366e-106
Kdm5d         1.370011e-95
Uty           1.486671e-87
Ddx3y        4.813782e-192
Xist         1.167393e-256
LOC105243748  1.133708e-48</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sexDEgenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">sexDE</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">sexDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Eif2s3y"      "Kdm5d"        "Uty"          "Ddx3y"        "Xist"        
[6] "LOC105243748"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">sexDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 54</code></pre>
</div>
<section id="enrichment-of-a-curated-gene-set"><h2 class="section-heading">Enrichment of a curated gene set<a class="anchor" aria-label="anchor" href="#enrichment-of-a-curated-gene-set"></a>
</h2>
<hr class="half-width"><div id="contribute-1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute-1" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute-1"></a>
</h3>
<div class="callout-content">
<p>Here we illustrate how to assess the enrichment of one gene set we
curate ourselves with our subset of DE genes with sex-specific
expression. Here we form such a gene set with genes from sex
chromosomes. Could you think of another more accurate gene set formed by
genes with sex-specific expression?</p>
</div>
</div>
</div>
<p>Build a gene set formed by genes located in the sex chromosomes X and
Y.</p>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">xygenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">[</span><span class="fu">decode</span><span class="op">(</span><span class="fu">seqnames</span><span class="op">(</span><span class="fu">rowRanges</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">length</span><span class="op">(</span><span class="va">xygenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2123</code></pre>
</div>
<p>Build a contingency table and conduct a one-tailed Fisher’s exact
test that verifies the association between genes being DE between male
and female mice and being located in a sex chromosome.</p>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">N</span> <span class="op">&lt;-</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">sexDEgenes</span><span class="op">)</span></span>
<span><span class="va">m</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">xygenes</span><span class="op">)</span></span>
<span><span class="va">k</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">xygenes</span>, <span class="va">sexDEgenes</span><span class="op">)</span><span class="op">)</span> </span>
<span><span class="va">dnames</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>GS<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"inside"</span>, <span class="st">"outside"</span><span class="op">)</span>, DE<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"yes"</span>, <span class="st">"no"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">k</span>, <span class="va">n</span><span class="op">-</span><span class="va">k</span>, <span class="va">m</span><span class="op">-</span><span class="va">k</span>, <span class="va">N</span><span class="op">+</span><span class="va">k</span><span class="op">-</span><span class="va">n</span><span class="op">-</span><span class="va">m</span><span class="op">)</span>,</span>
<span>                        nrow<span class="op">=</span><span class="fl">2</span>, ncol<span class="op">=</span><span class="fl">2</span>, dimnames<span class="op">=</span><span class="va">dnames</span><span class="op">)</span></span>
<span><span class="va">t</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>         DE
GS        yes    no
  inside   18  2105
  outside  36 39627</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fisher.test</span><span class="op">(</span><span class="va">t</span>, alternative<span class="op">=</span><span class="st">"greater"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Fisher's Exact Test for Count Data

data:  t
p-value = 7.944e-11
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 5.541517      Inf
sample estimates:
odds ratio 
  9.411737 </code></pre>
</div>
</section><section id="gene-ontology-analysis-with-clusterprofiler"><h2 class="section-heading">Gene ontology analysis with clusterProfiler<a class="anchor" aria-label="anchor" href="#gene-ontology-analysis-with-clusterprofiler"></a>
</h2>
<hr class="half-width"><div id="contribute-2" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute-2" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute-2"></a>
</h3>
<div class="callout-content">
<p>Here we illustrate how to assess the enrichment on the entire
collection of Gene Ontology (GO) gene sets using the package
clusterProfiler. Could we illustrate any missing important feature of
this package for this analysis objective? Could we briefly mention other
packages that may be useful for this task?</p>
</div>
</div>
</div>
<p>Second, let’s perform a gene set analysis for an entire collection of
gene sets using the Bioconductor package <a href="https://bioconductor.org/packages/clusterProfiler" class="external-link">clusterProfiler</a>.
For this purpose, we will fetch the results for the contrast between two
time points.</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTime</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">resTime</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 32652 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 4472, 14%
LFC &lt; 0 (down)     : 4276, 13%
outliers [1]       : 10, 0.031%
low counts [2]     : 8732, 27%
(mean count &lt; 1)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<p>Select DE genes between <code>Day0</code> and <code>Day8</code> with
FDR &lt; 5% and minimum 1.5-fold change.</p>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">timeDE</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resTime</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu">abs</span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">timeDE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2110    6</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">timeDE</span> <span class="op">&lt;-</span> <span class="va">timeDE</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="fu">abs</span><span class="op">(</span><span class="va">timeDE</span><span class="op">$</span><span class="va">log2FoldChange</span><span class="op">)</span>, decreasing<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span>, <span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">timeDE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>              baseMean log2FoldChange     lfcSE      stat       pvalue
LOC105245444  2.441873       4.768938 0.9013067  5.291138 1.215573e-07
LOC105246405  9.728219       4.601505 0.6101832  7.541186 4.657174e-14
4933427D06Rik 1.480365       4.556126 1.0318402  4.415535 1.007607e-05
A930006I01Rik 2.312732      -4.353155 0.9176026 -4.744053 2.094837e-06
LOC105245223  3.272536       4.337202 0.8611255  5.036666 4.737099e-07
A530053G22Rik 1.554735       4.243903 1.0248977  4.140806 3.460875e-05
                      padj
LOC105245444  1.800765e-06
LOC105246405  2.507951e-12
4933427D06Rik 9.169093e-05
A930006I01Rik 2.252139e-05
LOC105245223  6.047199e-06
A530053G22Rik 2.720142e-04</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">timeDEgenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">timeDE</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">timeDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "LOC105245444"  "LOC105246405"  "4933427D06Rik" "A930006I01Rik"
[5] "LOC105245223"  "A530053G22Rik"</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">timeDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 2110</code></pre>
</div>
<p>Call the <code>enrichGO()</code> function from <a href="https://bioconductor.org/packages/clusterProfiler" class="external-link">clusterProfiler</a>
as follows.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">enrichplot</span><span class="op">)</span></span>
<span></span>
<span><span class="va">resTimeGO</span> <span class="op">&lt;-</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>,</span>
<span>                      keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                      universe <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,</span>
<span>                      OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                      ont <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                      pvalueCutoff <span class="op">=</span> <span class="fl">0.01</span>,</span>
<span>                      qvalueCutoff <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="fu">dim</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 18  9</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                   ID                                           Description
GO:0071674 GO:0071674                            mononuclear cell migration
GO:0035456 GO:0035456                           response to interferon-beta
GO:0050900 GO:0050900                                   leukocyte migration
GO:0030595 GO:0030595                                  leukocyte chemotaxis
GO:0035458 GO:0035458                  cellular response to interferon-beta
GO:0002523 GO:0002523 leukocyte migration involved in inflammatory response
           GeneRatio   BgRatio       pvalue     p.adjust       qvalue
GO:0071674   32/1245 178/21025 1.612273e-08 7.986143e-05 7.581201e-05
GO:0035456   17/1245  59/21025 3.142295e-08 7.986143e-05 7.581201e-05
GO:0050900   50/1245 374/21025 5.961138e-08 1.010016e-04 9.588020e-05
GO:0030595   34/1245 221/21025 2.991816e-07 3.801850e-04 3.609074e-04
GO:0035458   14/1245  48/21025 4.368673e-07 4.441193e-04 4.216000e-04
GO:0002523   10/1245  25/21025 7.368444e-07 6.242300e-04 5.925780e-04
                                                                                                                                                                                                                                                                                           geneID
GO:0071674                                                                                                  Tnfsf18/Aire/Ccl17/Ccr7/Nlrp12/Ccl2/Retnlg/Apod/Il12a/Ccl5/Fut7/Ccl7/Spn/Itgb3/Grem1/Ptk2b/Lgals3/Adam8/Dusp1/Ch25h/Nbl1/Alox5/Padi2/Plg/Calr/Ager/Slamf9/Ccl6/Mdk/Itga4/Hsd3b7/Trpm4
GO:0035456                                                                                                                                                                          Tgtp1/Tgtp2/F830016B08Rik/Iigp1/Ifitm6/Igtp/Gm4951/Bst2/Oas1c/Irgm1/Gbp6/Ifi47/Aim2/Ifitm7/Irgm2/Ifit1/Ifi204
GO:0050900 Tnfsf18/Aire/Ccl17/Ccr7/Nlrp12/Bst1/Ccl2/Retnlg/Ppbp/Cxcl5/Apod/Il12a/Ccl5/Fut7/Ccl7/Ccl28/Spn/Sell/Itgb3/Grem1/Cxcl1/Ptk2b/Lgals3/Adam8/Pf4/Dusp1/Ch25h/S100a8/Nbl1/Alox5/Padi2/Plg/Edn3/Il33/Ptn/Ada/Emp2/Enpp1/Calr/Ager/Slamf9/Ccl6/Prex1/Aoc3/Itgam/Mdk/Itga4/Hsd3b7/P2ry12/Trpm4
GO:0030595                                                                                      Tnfsf18/Ccl17/Ccr7/Bst1/Ccl2/Retnlg/Ppbp/Cxcl5/Il12a/Ccl5/Ccl7/Sell/Grem1/Cxcl1/Ptk2b/Lgals3/Adam8/Pf4/Dusp1/Ch25h/S100a8/Nbl1/Alox5/Padi2/Edn3/Ptn/Calr/Slamf9/Ccl6/Prex1/Itgam/Mdk/Hsd3b7/Trpm4
GO:0035458                                                                                                                                                                                             Tgtp1/Tgtp2/F830016B08Rik/Iigp1/Igtp/Gm4951/Oas1c/Irgm1/Gbp6/Ifi47/Aim2/Irgm2/Ifit1/Ifi204
GO:0002523                                                                                                                                                                                                                                   Ccl2/Ppbp/Fut7/Adam8/S100a8/Alox5/Ptn/Aoc3/Itgam/Mdk
           Count
GO:0071674    32
GO:0035456    17
GO:0050900    50
GO:0030595    34
GO:0035458    14
GO:0002523    10</code></pre>
</div>
<p>Let’s build a more readable table of results.</p>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">kableExtra</span><span class="op">)</span></span>
<span></span>
<span><span class="va">resTimeGOtab</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">)</span></span>
<span><span class="va">resTimeGOtab</span><span class="op">$</span><span class="va">ID</span> <span class="op">&lt;-</span> <span class="cn">NULL</span></span>
<span><span class="va">resTimeGOtab</span><span class="op">$</span><span class="va">geneID</span> <span class="op">&lt;-</span> <span class="fu">sapply</span><span class="op">(</span><span class="fu">strsplit</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">$</span><span class="va">geneID</span>, <span class="st">"/"</span><span class="op">)</span>, <span class="va">paste</span>, collapse<span class="op">=</span><span class="st">", "</span><span class="op">)</span></span>
<span><span class="va">ktab</span> <span class="op">&lt;-</span> <span class="fu">kable</span><span class="op">(</span><span class="va">resTimeGOtab</span>, row.names<span class="op">=</span><span class="cn">TRUE</span>, caption<span class="op">=</span><span class="st">"GO results for DE genes between time points."</span><span class="op">)</span></span>
<span><span class="fu">kable_styling</span><span class="op">(</span><span class="va">ktab</span>, bootstrap_options<span class="op">=</span><span class="fu">c</span><span class="op">(</span><span class="st">"stripped"</span>, <span class="st">"hover"</span>, <span class="st">"responsive"</span><span class="op">)</span>, fixed_thead<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<table class="table table table-hover table-responsive" style="margin-left: auto; margin-right: auto;"><caption>
GO results for DE genes between time points.
</caption>
<thead><tr><th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
Description
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
GeneRatio
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
BgRatio
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
pvalue
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
p.adjust
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
qvalue
</th>
<th style="text-align:left;position: sticky; top:0; background-color: #FFFFFF;">
geneID
</th>
<th style="text-align:right;position: sticky; top:0; background-color: #FFFFFF;">
Count
</th>
</tr></thead><tbody><tr><td style="text-align:left;">
<a href="GO:0071674" class="uri">GO:0071674</a>
</td>
<td style="text-align:left;">
mononuclear cell migration
</td>
<td style="text-align:left;">
32/1245
</td>
<td style="text-align:left;">
178/21025
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:right;">
0.0000799
</td>
<td style="text-align:right;">
0.0000758
</td>
<td style="text-align:left;">
Tnfsf18, Aire, Ccl17, Ccr7, Nlrp12, Ccl2, Retnlg, Apod, Il12a, Ccl5,
Fut7, Ccl7, Spn, Itgb3, Grem1, Ptk2b, Lgals3, Adam8, Dusp1, Ch25h, Nbl1,
Alox5, Padi2, Plg, Calr, Ager, Slamf9, Ccl6, Mdk, Itga4, Hsd3b7, Trpm4
</td>
<td style="text-align:right;">
32
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0035456" class="uri">GO:0035456</a>
</td>
<td style="text-align:left;">
response to interferon-beta
</td>
<td style="text-align:left;">
17/1245
</td>
<td style="text-align:left;">
59/21025
</td>
<td style="text-align:right;">
0.00e+00
</td>
<td style="text-align:right;">
0.0000799
</td>
<td style="text-align:right;">
0.0000758
</td>
<td style="text-align:left;">
Tgtp1, Tgtp2, F830016B08Rik, Iigp1, Ifitm6, Igtp, Gm4951, Bst2, Oas1c,
Irgm1, Gbp6, Ifi47, Aim2, Ifitm7, Irgm2, Ifit1, Ifi204
</td>
<td style="text-align:right;">
17
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0050900" class="uri">GO:0050900</a>
</td>
<td style="text-align:left;">
leukocyte migration
</td>
<td style="text-align:left;">
50/1245
</td>
<td style="text-align:left;">
374/21025
</td>
<td style="text-align:right;">
1.00e-07
</td>
<td style="text-align:right;">
0.0001010
</td>
<td style="text-align:right;">
0.0000959
</td>
<td style="text-align:left;">
Tnfsf18, Aire, Ccl17, Ccr7, Nlrp12, Bst1, Ccl2, Retnlg, Ppbp, Cxcl5,
Apod, Il12a, Ccl5, Fut7, Ccl7, Ccl28, Spn, Sell, Itgb3, Grem1, Cxcl1,
Ptk2b, Lgals3, Adam8, Pf4, Dusp1, Ch25h, S100a8, Nbl1, Alox5, Padi2,
Plg, Edn3, Il33, Ptn, Ada, Emp2, Enpp1, Calr, Ager, Slamf9, Ccl6, Prex1,
Aoc3, Itgam, Mdk, Itga4, Hsd3b7, P2ry12, Trpm4
</td>
<td style="text-align:right;">
50
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0030595" class="uri">GO:0030595</a>
</td>
<td style="text-align:left;">
leukocyte chemotaxis
</td>
<td style="text-align:left;">
34/1245
</td>
<td style="text-align:left;">
221/21025
</td>
<td style="text-align:right;">
3.00e-07
</td>
<td style="text-align:right;">
0.0003802
</td>
<td style="text-align:right;">
0.0003609
</td>
<td style="text-align:left;">
Tnfsf18, Ccl17, Ccr7, Bst1, Ccl2, Retnlg, Ppbp, Cxcl5, Il12a, Ccl5,
Ccl7, Sell, Grem1, Cxcl1, Ptk2b, Lgals3, Adam8, Pf4, Dusp1, Ch25h,
S100a8, Nbl1, Alox5, Padi2, Edn3, Ptn, Calr, Slamf9, Ccl6, Prex1, Itgam,
Mdk, Hsd3b7, Trpm4
</td>
<td style="text-align:right;">
34
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0035458" class="uri">GO:0035458</a>
</td>
<td style="text-align:left;">
cellular response to interferon-beta
</td>
<td style="text-align:left;">
14/1245
</td>
<td style="text-align:left;">
48/21025
</td>
<td style="text-align:right;">
4.00e-07
</td>
<td style="text-align:right;">
0.0004441
</td>
<td style="text-align:right;">
0.0004216
</td>
<td style="text-align:left;">
Tgtp1, Tgtp2, F830016B08Rik, Iigp1, Igtp, Gm4951, Oas1c, Irgm1, Gbp6,
Ifi47, Aim2, Irgm2, Ifit1, Ifi204
</td>
<td style="text-align:right;">
14
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0002523" class="uri">GO:0002523</a>
</td>
<td style="text-align:left;">
leukocyte migration involved in inflammatory response
</td>
<td style="text-align:left;">
10/1245
</td>
<td style="text-align:left;">
25/21025
</td>
<td style="text-align:right;">
7.00e-07
</td>
<td style="text-align:right;">
0.0006242
</td>
<td style="text-align:right;">
0.0005926
</td>
<td style="text-align:left;">
Ccl2, Ppbp, Fut7, Adam8, S100a8, Alox5, Ptn, Aoc3, Itgam, Mdk
</td>
<td style="text-align:right;">
10
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0050953" class="uri">GO:0050953</a>
</td>
<td style="text-align:left;">
sensory perception of light stimulus
</td>
<td style="text-align:left;">
26/1245
</td>
<td style="text-align:left;">
164/21025
</td>
<td style="text-align:right;">
4.10e-06
</td>
<td style="text-align:right;">
0.0029769
</td>
<td style="text-align:right;">
0.0028259
</td>
<td style="text-align:left;">
Aipl1, Vsx2, Nxnl2, Lrit3, Cryba2, Bfsp2, Lrat, Gabrr2, Lum, Rlbp1,
Pde6g, Gpr179, Col1a1, Cplx3, Best1, Ush1g, Rs1, Rdh5, Guca1b, Th,
Ppef2, Rbp4, Olfm2, Rom1, Vsx1, Rpe65
</td>
<td style="text-align:right;">
26
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0071675" class="uri">GO:0071675</a>
</td>
<td style="text-align:left;">
regulation of mononuclear cell migration
</td>
<td style="text-align:left;">
21/1245
</td>
<td style="text-align:left;">
117/21025
</td>
<td style="text-align:right;">
4.70e-06
</td>
<td style="text-align:right;">
0.0029799
</td>
<td style="text-align:right;">
0.0028288
</td>
<td style="text-align:left;">
Tnfsf18, Aire, Ccr7, Ccl2, Apod, Il12a, Ccl5, Ccl7, Spn, Itgb3, Grem1,
Ptk2b, Lgals3, Adam8, Dusp1, Nbl1, Padi2, Calr, Ager, Mdk, Itga4
</td>
<td style="text-align:right;">
21
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0060326" class="uri">GO:0060326</a>
</td>
<td style="text-align:left;">
cell chemotaxis
</td>
<td style="text-align:left;">
38/1245
</td>
<td style="text-align:left;">
298/21025
</td>
<td style="text-align:right;">
7.00e-06
</td>
<td style="text-align:right;">
0.0039475
</td>
<td style="text-align:right;">
0.0037473
</td>
<td style="text-align:left;">
Tnfsf18, Ccl17, Ccr7, Bst1, Ccl2, Retnlg, Ppbp, Cxcl5, Nr4a1, Il12a,
Ccl5, Ccl7, Ccl28, Sell, Grem1, Cxcl1, Ptk2b, Lgals3, Adam8, Pf4, Dusp1,
Ch25h, S100a8, Nbl1, Alox5, Padi2, Edn3, Ptn, Plxnb3, Calr, Lpar1,
Slamf9, Ccl6, Prex1, Itgam, Mdk, Hsd3b7, Trpm4
</td>
<td style="text-align:right;">
38
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0007601" class="uri">GO:0007601</a>
</td>
<td style="text-align:left;">
visual perception
</td>
<td style="text-align:left;">
25/1245
</td>
<td style="text-align:left;">
160/21025
</td>
<td style="text-align:right;">
8.10e-06
</td>
<td style="text-align:right;">
0.0041224
</td>
<td style="text-align:right;">
0.0039134
</td>
<td style="text-align:left;">
Aipl1, Vsx2, Nxnl2, Lrit3, Cryba2, Bfsp2, Lrat, Gabrr2, Lum, Rlbp1,
Pde6g, Gpr179, Col1a1, Cplx3, Best1, Rs1, Rdh5, Guca1b, Th, Ppef2, Rbp4,
Olfm2, Rom1, Vsx1, Rpe65
</td>
<td style="text-align:right;">
25
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0002460" class="uri">GO:0002460</a>
</td>
<td style="text-align:left;">
adaptive immune response based on somatic recombination of immune
receptors built from immunoglobulin superfamily domains
</td>
<td style="text-align:left;">
42/1245
</td>
<td style="text-align:left;">
350/21025
</td>
<td style="text-align:right;">
1.09e-05
</td>
<td style="text-align:right;">
0.0050165
</td>
<td style="text-align:right;">
0.0047621
</td>
<td style="text-align:left;">
Tnfsf18, Aire, Ccr7, Gzmb, Cd70, H2-Q6, H2-Q7, Il12a, Il12rb1, H2-Q4,
Il18rap, Fut7, Spn, Icosl, Pirb, H2-Q2, Irf7, Cd274, Tnfrsf13c, Il2rb,
Masp2, C8g, C4b, Il33, H2-Q1, Ada, Emp2, C3, Enpp1, Tfrc, Cd46, H2-K1,
Rorc, Csf2rb, Ager, H2-T23, Tap2, Tnfsf13b, Pla2g4a, Trpm4, Parp3, Klhl6
</td>
<td style="text-align:right;">
42
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:1990266" class="uri">GO:1990266</a>
</td>
<td style="text-align:left;">
neutrophil migration
</td>
<td style="text-align:left;">
21/1245
</td>
<td style="text-align:left;">
125/21025
</td>
<td style="text-align:right;">
1.36e-05
</td>
<td style="text-align:right;">
0.0057697
</td>
<td style="text-align:right;">
0.0054771
</td>
<td style="text-align:left;">
Ccl17, Ccr7, Bst1, Ccl2, Ppbp, Cxcl5, Ccl5, Fut7, Ccl7, Sell, Cxcl1,
Lgals3, Adam8, Pf4, S100a8, Edn3, Emp2, Ccl6, Prex1, Itgam, Mdk
</td>
<td style="text-align:right;">
21
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0036336" class="uri">GO:0036336</a>
</td>
<td style="text-align:left;">
dendritic cell migration
</td>
<td style="text-align:left;">
9/1245
</td>
<td style="text-align:left;">
27/21025
</td>
<td style="text-align:right;">
1.54e-05
</td>
<td style="text-align:right;">
0.0057975
</td>
<td style="text-align:right;">
0.0055036
</td>
<td style="text-align:left;">
Tnfsf18, Ccr7, Nlrp12, Retnlg, Il12a, Alox5, Calr, Slamf9, Trpm4
</td>
<td style="text-align:right;">
9
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0002685" class="uri">GO:0002685</a>
</td>
<td style="text-align:left;">
regulation of leukocyte migration
</td>
<td style="text-align:left;">
31/1245
</td>
<td style="text-align:left;">
230/21025
</td>
<td style="text-align:right;">
1.60e-05
</td>
<td style="text-align:right;">
0.0057975
</td>
<td style="text-align:right;">
0.0055036
</td>
<td style="text-align:left;">
Tnfsf18, Aire, Ccr7, Bst1, Ccl2, Apod, Il12a, Ccl5, Fut7, Ccl7, Ccl28,
Spn, Sell, Itgb3, Grem1, Ptk2b, Lgals3, Adam8, Dusp1, Nbl1, Padi2, Edn3,
Il33, Ptn, Ada, Calr, Ager, Aoc3, Mdk, Itga4, P2ry12
</td>
<td style="text-align:right;">
31
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0030593" class="uri">GO:0030593</a>
</td>
<td style="text-align:left;">
neutrophil chemotaxis
</td>
<td style="text-align:left;">
18/1245
</td>
<td style="text-align:left;">
100/21025
</td>
<td style="text-align:right;">
2.11e-05
</td>
<td style="text-align:right;">
0.0071625
</td>
<td style="text-align:right;">
0.0067993
</td>
<td style="text-align:left;">
Ccl17, Ccr7, Bst1, Ccl2, Ppbp, Cxcl5, Ccl5, Ccl7, Sell, Cxcl1, Lgals3,
Pf4, S100a8, Edn3, Ccl6, Prex1, Itgam, Mdk
</td>
<td style="text-align:right;">
18
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0016264" class="uri">GO:0016264</a>
</td>
<td style="text-align:left;">
gap junction assembly
</td>
<td style="text-align:left;">
7/1245
</td>
<td style="text-align:left;">
17/21025
</td>
<td style="text-align:right;">
2.88e-05
</td>
<td style="text-align:right;">
0.0086355
</td>
<td style="text-align:right;">
0.0081976
</td>
<td style="text-align:left;">
Gjd3, Aplnr, Gja5, Agt, Hopx, Cav1, Ace
</td>
<td style="text-align:right;">
7
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:1903028" class="uri">GO:1903028</a>
</td>
<td style="text-align:left;">
positive regulation of opsonization
</td>
<td style="text-align:left;">
6/1245
</td>
<td style="text-align:left;">
12/21025
</td>
<td style="text-align:right;">
2.89e-05
</td>
<td style="text-align:right;">
0.0086355
</td>
<td style="text-align:right;">
0.0081976
</td>
<td style="text-align:left;">
Pla2g5, Masp2, C4b, Colec11, C3, Cfp
</td>
<td style="text-align:right;">
6
</td>
</tr><tr><td style="text-align:left;">
<a href="GO:0034341" class="uri">GO:0034341</a>
</td>
<td style="text-align:left;">
response to type II interferon
</td>
<td style="text-align:left;">
22/1245
</td>
<td style="text-align:left;">
142/21025
</td>
<td style="text-align:right;">
3.18e-05
</td>
<td style="text-align:right;">
0.0089664
</td>
<td style="text-align:right;">
0.0085118
</td>
<td style="text-align:left;">
Ccl17, Gbp4, Ccl2, Tgtp1, H2-Q7, Il12rb1, Ifitm6, Ccl5, Ccl7, Igtp,
Nos2, Nlrc5, Bst2, Irgm1, Gbp6, Capg, Ifitm7, Gbp9, Gbp5, Irgm2, Ccl6,
Aqp4
</td>
<td style="text-align:right;">
22
</td>
</tr></tbody></table><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Gene set enrichment analysis helps to answer “what are the common
biological functions affected in the experiment”.</li>
<li>Gene sets can be retrieved from several online resources, including
mSigDB, KEGG, reactome, Gene Ontology.</li>
<li>A rich set of gene set analysis methods are available. Some methods
require genes to be classified into ‘significant’ and ‘non-significant’
before the gene set analysis, while others ingest the full ranking of
genes.</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="05-differential-expression.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="07-next-steps.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="05-differential-expression.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Differential</a>
        <a class="chapter-link float-end" href="07-next-steps.html" rel="next">Next: Next steps... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/06-gene-set-analysis.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:charlotte.soneson@fmi.ch">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.12.3" class="external-link">sandpaper (0.12.3)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/mblue9/varnish/tree/3403ad0dd37f7641c6442330fcc182937e7964c5" class="external-link">varnish (0.2.16)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/bioc-rnaseq/06-gene-set-analysis.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Gene set analysis",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-rnaseq/06-gene-set-analysis.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-rnaseq/06-gene-set-analysis.html",
  "dateCreated": "2020-09-15",
  "dateModified": "2023-05-17",
  "datePublished": "2023-06-06"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

