<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<title>RNA-seq analysis with Bioconductor: All in One View</title>
<meta name="viewport" content="width=device-width, initial-scale=1">
<link rel="stylesheet" type="text/css" href="assets/styles.css">
<script src="assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="apple-touch-icon.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="manifest" href="site.webmanifest">
<link rel="mask-icon" href="safari-pinned-tab.svg" color="#5bbad5">
<meta name="msapplication-TileColor" content="#da532c">
<meta name="theme-color" content="#ffffff">
</head>
<body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught." style="text-decoration: unset">
           
<!--
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#early-development-pre-alpha-through-alpha" class="alert-link" style="color: #383838">Pre-Alpha
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>
-->
          <a href="https://bioconductor.github.io/2023-07-31-BiocRnaseq-Bioc2023/" class="external-link alert-link" style="color: #383838">Under active development for BioC2023 workshop July 31 - August 1 2023
            <i aria-hidden="true" class="icon" data-feather="alert-octagon" style="color: #FF4955; border-radius: 5px"></i>
          </a>

          <span class="visually-hidden">This lesson is in the pre-alpha phase, which means that it is in early development, but has not yet been taught.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Learner View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1">
<li><button class="dropdown-item" type="button" onclick="window.location.href='instructor/aio.html';">Instructor View</button></li>
        </ul>
</div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="assets/images/incubator-logo-sm.svg">
</div>
    <div class="lesson-title-md">
      RNA-seq analysis with Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0">
<li class="nav-item">
          <span class="lesson-title">
            RNA-seq analysis with Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="reference.html#glossary">Glossary</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="profiles.html">Learner Profiles</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown">
<li><a class="dropdown-item" href="discuss.html">Discussion</a></li>
<li><a class="dropdown-item" href="reference.html"></a></li>
          </ul>
</li>
      </ul>
</div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled>
<input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset>
</form>
  </div>
<!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  RNA-seq analysis with Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: %" class="percentage">
    %
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: %" aria-valuenow="" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Learner View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="instructor/aio.html">Instructor View</a>
                      </div>
                    </div>
                  </div>
<!--/div.accordion-item-->
                </div>
<!--/div.accordion-flush-->
              </div>
<!--div.sidenav-view-selector -->
            </div>
<!--/div.col -->
      
            <hr>
</div>
<!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Setup</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="01-intro-to-rnaseq.html">1. Introduction to RNA-seq</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="02-setup.html">2. RStudio Project and Experimental Data</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="03-import-annotate.html">3. Importing and annotating quantified data into R</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="04-exploratory-qc.html">4. Exploratory analysis and quality control</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush6">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading6">
        <a href="05-differential-expression.html">5. Differential expression analysis</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="06-gene-set-analysis.html">6. Gene set enrichment analysis</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="07-next-steps.html">7. Next steps</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush9">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading9">
        <a href="08-extra-design.html">8. Extra exploration of design matrices</a>
    </div>
<!--/div.accordion-header-->
        
  </div>
<!--/div.accordion-item-->
</div>
<!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width">
<div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul>
<li>
                        <a href="key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="reference.html#glossary">Glossary</a>
                      </li>
                      <li>
                        <a href="profiles.html">Learner Profiles</a>
                      </li>
                      <li><a href="discuss.html">Discussion</a></li>
<li><a href="reference.html"></a></li>
                    </ul>
</div>
                </div>
              </div>
            </div>
            <hr class="half-width resources">
<a href="aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none">
<div class="d-grid gap-1">
            
            </div>
          </div>
<!-- /div.accordion -->
        </div>
<!-- /div.sidebar-inner -->
      </nav>
</div>
<!-- /div.sidebar -->
  </div>
<!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-extra.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <main id="main-content" class="main-content"><div class="container lesson-content">
        
        
<section id="aio-01-intro-to-rnaseq"><p>Content from <a href="01-intro-to-rnaseq.html">Introduction to RNA-seq</a></p>
<hr>
<p> Last updated on 2023-07-17 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/01-intro-to-rnaseq.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the different choices to consider when planning an RNA-seq
experiment?</li>
<li>How does one process the raw fastq files to generate a table with
read counts per gene and sample?</li>
<li>Where does one find information about annotated genes for a given
organism?</li>
<li>Which are the most commonly used Bioconductor packages for
statistical analysis of RNA-seq data?</li>
<li>What are the typical steps in an RNA-seq analysis?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain what RNA-seq is.</li>
<li>Describe some of the most common design choices that have to be made
before running an RNA-seq experiment.</li>
<li>Provide an overview of the procedure to go from the raw data to the
read count matrix that will be used for downstream analysis.</li>
<li>Show some common types of results and visualizations generated in
RNA-seq analyses.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div id="contribute" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute"></a>
</h3>
<div class="callout-content">
<p>This episode is intended to introduce important concepts in RNA-seq,
such as the biology of RNA-seq, typical research questions, experimental
design, data processing, QC issues and ways around them, functional
analysis of the expressed genes and use of Bioconductor for RNA-seq
analysis pipelines to bring everyone up to speed. See this <a href="https://docs.google.com/document/d/12hUqVo2MhgYH9IT8ShJKSZgeMLzjeUh2EKdxM_siWwM" class="external-link">link</a>
on how to contribute</p>
</div>
</div>
</div>
<div class="section level1">
<h1 id="rna-seq-quantification-from-reads-to-count-matrix">RNA-seq quantification: from reads to count matrix<a class="anchor" aria-label="anchor" href="#rna-seq-quantification-from-reads-to-count-matrix"></a>
</h1>
<p>There is a plethroa of RNA quantification pipelines, and the most
common approaches can be categorized into three main types:</p>
<ol style="list-style-type: decimal">
<li><p>Align reads to the genome, and count the number of reads that map
to each gene. This is the one of simplest methods. For species for which
the transcriptome is poorly annotated, this would be the preferred
approach. Example: <code>STAR</code> alignment to GRCz11 +
<code>Rsubread</code> <a href="https://doi.org/10.1093%2Fnar%2Fgkz114" class="external-link">featureCounts</a></p></li>
<li><p>Align reads to the transcriptome, quantify transcript expression,
and summarize transcript expression into gene expression. This approach
can produce accurate quantification results based <a href="https://doi.org/10.1186/s13059-016-0940-1" class="external-link">independent
benchmarking</a>, particularly for high-quality samples without DNA
contamination. Example: RSEM quantification using
<code>rsem-calculate-expression --star</code> on the GENCODE GRCh38
transcriptome + <code>tximport</code></p></li>
<li><p>Pseudoalign reads against the transcriptome, using the
corresponding genome as a decoy, quantifying transcript expression in
the process, and summarize the transcript-level expression into
gene-level expression. The advantages of this approach include:
computational efficiency, mitigation of the effect of DNA contamination,
and GC bias correction. Example: <code>salmon quant --gcBias</code> +
<code>tximport</code></p></li>
</ol>
<p>At typical sequencing read depth, gene expression quantification is
often more accurate than transcript expression quantification. However,
gene expression quantification can be <a href="https://doi.org/10.12688/f1000research.7563.1" class="external-link">improved</a> by
first quantifying transcript expression and then summarizing it to gene
expression.</p>
<p>Other tools used in RNA-seq quantification include: TopHat2, bowtie2,
kallisto, HTseq, among many others.</p>
<p>Choosing the appropriate RNA-seq quantification would depend on the
quality of the transcriptome annotation, the quality of the RNA-seq
library preparation, the presence of contaminating sequences, among many
factors. Often time, it would be important to compare the quantification
results of multiple approaches.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>RNA-seq is a technique of measuring the amount of RNA expressed
within a cell/tissue and state at a given time.</li>
<li>Many choices have to be made when planning an RNA-seq experiment,
such as whether to perform poly-A selection or ribosomal depletion,
whether to apply a stranded or an unstranded protocol, and whether to
sequence the reads in a single-end or paired-end fashion. Each of the
choices have consequences for the processing and interpretation of the
data.</li>
<li>Many approaches exist for quantification of RNA-seq data. Some
methods align reads to the genome and count the number of reads
overlapping gene loci. Other methods map reads to the transcriptome and
use a probabilistic approach to estimate the abundance of each gene or
transcript.</li>
<li>Information about annotated genes can be accessed via several
sources, including Ensembl, UCSC and GENCODE.</li>
<li>The most commonly used Bioconductor packages for statistical
analysis of RNA-seq data are <code>DESeq2</code>, <code>edgeR</code> and
<code>limma</code>.</li>
</ul>
</div>
</div>
</div>
</div></section><section id="aio-02-setup"><p>Content from <a href="02-setup.html">RStudio Project and Experimental Data</a></p>
<hr>
<p> Last updated on 2023-07-17 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/02-setup.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How do you use RStudio project to manage your analysis project?</li>
<li>What is the most effective way for organising directories for an
analysis project?</li>
<li>How to download a dataset from the internet and save as a file.</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Create an RStudio project and the the directories required for
storing the files pertinent to an analysis project.</li>
<li>Download the data set that will be used for the subsequent
episodes.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="introduction"><h2 class="section-heading">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<hr class="half-width">
<p>Typically, an analysis project begins with dataset files, a handful
of R scripts and output files in a directory. As the project advances,
complexity inevitably rises with the addition of more scripts, output
files, and possibly new datasets. The complexity is further amplified
when dealing with multiple versions of scripts and output files,
necessitating efficient organisation. If these are not well-managed from
the beginning, resuming the project after a break, or sharing the
project with someone else becomes challenging and time-consuming, as we
struggle to recall the project’s status and to navigate the directory
tree. Additionally without proper organisation, the project’s complexity
could lead to frequent use of the <code>setwd</code> function to switch
between different working directories, resulting in a disorganised
workspace.</p>
<p>In this lesson, we will first focus on an effective strategy for
managing the files, both used and generated by our data analysis
project, within a <em>working directory</em>.</p>
<div id="what-is-a-working-directory" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<div id="what-is-a-working-directory" class="callout-inner">
<h3 class="callout-title">What is a working directory?<a class="anchor" aria-label="anchor" href="#what-is-a-working-directory"></a>
</h3>
<div class="callout-content">
<p>A working directory in R is the default location on your computer
where R looks for files to load or store any data you wish to save. More
information can be found in our <a href="https://carpentries-incubator.github.io/bioc-intro/20-r-rstudio.html#the-working-directory" class="external-link">Introduction
to data analysis with R and Bioconductor</a> lesson.</p>
</div>
</div>
</div>
<p>Secondly, we will also learn how to leverage RStudio projects, a
feature built-in to RStudio for managing our analysis project.</p>
<div id="what-is-rstudio" class="callout prereq">
<div class="callout-square">
<i class="callout-icon" data-feather="check"></i>
</div>
<div id="what-is-rstudio" class="callout-inner">
<h3 class="callout-title">What is RStudio?<a class="anchor" aria-label="anchor" href="#what-is-rstudio"></a>
</h3>
<div class="callout-content">
<p>RStudio is a freely available integrated development environment
(IDE), widely used by scientists and software developers for developing
software or analysing datasets in R. If you require assistance with
RStudio or its general usage, please refer to our <a href="https://carpentries-incubator.github.io/bioc-intro/20-r-rstudio.html" class="external-link">Introduction
to data analysis with R and Bioconductor</a> lesson.</p>
</div>
</div>
</div>
<p>Finally, in this lesson, we will also learn to use the R function
<code>download.files</code> to download the data for the subsequent
episodes.</p>
</section><section id="structuring-your-working-directory"><h2 class="section-heading">Structuring your working directory<a class="anchor" aria-label="anchor" href="#structuring-your-working-directory"></a>
</h2>
<hr class="half-width">
<p>For a more streamlined workflow, we suggest storing all files
associated with an analysis in a specific directory, which will serve as
your project’s working directory. Initially, this working directory
should contain four distinct directories:</p>
<ul>
<li>
<code>data</code>: dedicated to storing raw data. This folder should
ideally only house raw data and not be modified unless you receive a new
dataset (even then, if you have the storage capacity, we suggest you
retain the previous dataset as well in case you need it again in the
future). For RNA-seq data analysis, this directory will typically
contain <code>*.fastq</code> files and any related metadata files for
the experiment.</li>
<li>
<code>scripts</code>: for storing the R scripts you’ve written and
utilised for analysing the data.</li>
<li>
<code>documents</code>: for storing documents related to your
analysis, such as a manuscript outline or meeting notes with your
team.</li>
<li>
<code>output</code>: for storing intermediate or final results
generated by the R scripts in the <code>scripts</code> directory.
Importantly, if you carry out data cleaning or pre-processing, the
output should ideally be stored in this directory, as these no longer
represent raw data.</li>
</ul>
<p>As your project grows in complexity, you might find it necessary to
create more directories or sub-directories. Nevertheless, the
aforementioned four directories should serve as the foundation of your
working directory.</p>
<div id="create-the-directories-for-subsequent-episodes" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="create-the-directories-for-subsequent-episodes" class="callout-inner">
<h3 class="callout-title">Create the directories for subsequent
episodes<a class="anchor" aria-label="anchor" href="#create-the-directories-for-subsequent-episodes"></a>
</h3>
<div class="callout-content">
<p>Select a directory on your computer to serve as the working directory
for the rest of this episode and lesson. Then, within this chosen
directory, create the four fundamental directories previously discussed
(<code>data</code>, <code>scripts</code>, <code>documents</code>, and
<code>output</code>).</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<figure style="text-align: center"><img src="fig/02_setup/folders_working_dir.png" alt="Your working directory should look like this" width="1221" class="figure mx-auto d-block"><figcaption>
Your working directory should look like this
</figcaption></figure>
</div>
</div>
</div>
</div>
</section><section id="using-rstudio-project-to-manage-your-working-directory"><h2 class="section-heading">Using RStudio project to manage your working directory<a class="anchor" aria-label="anchor" href="#using-rstudio-project-to-manage-your-working-directory"></a>
</h2>
<hr class="half-width">
<p>As previously highlighted, RStudio project is a feature built-into
RStudio for managing your analysis project. It does so by storing
project-specific settings in an <code>.Rproj</code> file stored in your
project’s working directory. Loading these settings up into RStudio by
either opening the <code>.Rproj</code> file directly or through
RStudio’s open project option (from the menu bar, select
<code>File</code> &gt; <code>Open Project...</code>) will automatically
set your working directory in R to the location of the
<code>.Rproj</code> file, essentially your project’s working
directory.</p>
<p>To create an RStudio project:</p>
<ol style="list-style-type: decimal">
<li>Start RStudio.</li>
<li>Navigate to the menu bar and select <code>File</code> &gt;
<code>New Project...</code>.</li>
<li>Choose <code>Existing Directory</code>.</li>
<li>Click <code>Browse...</code> button, and select the directory you
have previously chosen as the working directory for the analysis (i.e.,
the directory where the 4 essential directories reside).</li>
<li>Click <code>Create Project</code> at the bottom right of the
window.</li>
</ol>
<p>Upon completion of the steps above, you will find a
<code>.Rproj</code> file within your project’s working directory.</p>
<figure style="text-align: center"><img src="fig/02_setup/rstudio_project.png" alt="A new .Rproj file should be created in your chosen working directory." width="1621" class="figure mx-auto d-block"><figcaption>
A new .Rproj file should be created in your chosen working directory.
</figcaption></figure><p>Moreover, the heading of your RStudio console should now also display
the absolute path of your project’s working directory, i.e., where the
<code>.Rproj</code> file resides, indicating that RStudio has set this
directory as your working directory in R.</p>
<figure style="text-align: center"><img src="fig/02_setup/rstudio_console.png" alt="Your R working directory should now be set to where the .Rproj file resides." width="1211" class="figure mx-auto d-block"><figcaption>
Your R working directory should now be set to where the .Rproj file
resides.
</figcaption></figure><p>From this point forward, any R code that you execute that involves
reading in data from a file, or saving some data into a file will, by
default, be directe to a path relative to your project’s working
directory.</p>
<p>If you wish to close the project, perhaps to open another project,
create a new one, or take a break from the project, you can do so by
using to <code>File</code> &gt; <code>Close Project</code> option
located in the menu bar.</p>
</section><section id="download-the-rna-seq-data-for-subsequent-episodes"><h2 class="section-heading">Download the RNA-seq data for subsequent episodes<a class="anchor" aria-label="anchor" href="#download-the-rna-seq-data-for-subsequent-episodes"></a>
</h2>
<hr class="half-width">
<p>Finally, we will learn how to use R to download the RNA-seq data
required for the subsequent episodes of the lesson. The dataset we will
be using was generated to investigate the impact upper-respiratory
infection have on changes in RNA transcription in the cerebellum and
spinal cord of mice. This dataset was produced as part of the following
study:</p>
<blockquote>
<p>Blackmore, Stephen, et al. “Influenza infection triggers disease in a
genetic model of experimental autoimmune encephalomyelitis.” Proceedings
of the National Academy of Sciences 114.30 (2017): E6107-E6116.</p>
</blockquote>
<p>The dataset is available at <a href="https://www.ncbi.nlm.nih.gov/geo/" class="external-link">Gene Expression Omnibus
(GEO)</a>, under the accession number <a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE96870" class="external-link">GSE96870</a>.
Downloading data from GEO is not straightforward (and won’t be covered
in this lesson). Hence, we have made the data available on our Github
repository for easier access.</p>
<p>To download the files, we will use the R function
<code>download.file</code>, which necessitates at least two parameters:
<code>url</code> and <code>destfile</code>. The <code>url</code>
parameter is used to specify the address on the internet to download the
data from. The <code>destfile</code> parameter indicates where to save
the downloaded file and what the downloaded file should be named.</p>
<p>Let’s download one of the four data files needed for the remainder of
this lesson. The data file is located at <a href="https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_counts_cerebellum.csv" class="external-link uri">https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_counts_cerebellum.csv</a>.
We shall save the downloaded file in the <code>data</code> folder of our
working directory with the name
<code>GSE96870_counts_cerebellum.csv</code>.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_counts_cerebellum.csv"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/GSE96870_counts_cerebellum.csv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>If you navigate to the <code>data</code> folder in your working
directory, you should now find a file named
<code>GSE96870_counts_cerebellum.csv</code>.</p>
<figure style="text-align: center"><img src="fig/02_setup/data_folder.png" alt="A file named GSE96870_counts_cerebellum.csv should now reside in the data folder." width="1632" class="figure mx-auto d-block"><figcaption>
A file named GSE96870_counts_cerebellum.csv should now reside in the
data folder.
</figcaption></figure><div id="download-the-remaining-data-set-files" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="download-the-remaining-data-set-files" class="callout-inner">
<h3 class="callout-title">Download the remaining data set files<a class="anchor" aria-label="anchor" href="#download-the-remaining-data-set-files"></a>
</h3>
<div class="callout-content">
<p>There are three more data set files we need to download for the
remainder of this lesson.</p>
<table class="table">
<colgroup>
<col width="77%">
<col width="22%">
</colgroup>
<thead><tr class="header">
<th>URL</th>
<th>Filename</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><a href="https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_cerebellum.csv" class="external-link uri">https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_cerebellum.csv</a></td>
<td>GSE96870_coldata_cerebellum.csv</td>
</tr>
<tr class="even">
<td><a href="https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_all.csv" class="external-link uri">https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_all.csv</a></td>
<td>GSE96870_coldata_all.csv</td>
</tr>
<tr class="odd">
<td><a href="https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_rowranges.tsv" class="external-link uri">https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_rowranges.tsv</a></td>
<td>GSE96870_rowranges.tsv</td>
</tr>
</tbody>
</table>
<p>Use the <code>download.file</code> function to download the files
into the <code>data</code> folder in your working directory.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_cerebellum.csv"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/GSE96870_coldata_cerebellum.csv"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_coldata_all.csv"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/GSE96870_coldata_all.csv"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu">download.file</span><span class="op">(</span></span>
<span>    url <span class="op">=</span> <span class="st">"https://github.com/carpentries-incubator/bioc-rnaseq/raw/main/episodes/data/GSE96870_rowranges.tsv"</span>, </span>
<span>    destfile <span class="op">=</span> <span class="st">"data/GSE96870_rowranges.tsv"</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Proper organisation of the files required for your project in a
working directory is crucial for maintaining order and ensuring easy
access in the future.</li>
<li>RStudio project serves as a valuable tool for managing your
project’s working directory and facilitating analysis.</li>
<li>The <code>download.files</code> function in R can be used for
downloading datasets from the internet.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-03-import-annotate"><p>Content from <a href="03-import-annotate.html">Importing and annotating quantified data into R</a></p>
<hr>
<p> Last updated on 2023-07-17 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/03-import-annotate.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can one import quantified gene expression data into an object
suitable for downstream statistical analysis in R?</li>
<li>What types of gene identifiers are typically used, and how are
mappings between them done?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn how to import the quantifications into a SummarizedExperiment
object.</li>
<li>Learn how to add additional gene annotations to the object.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div id="contribute" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute"></a>
</h3>
<div class="callout-content">
<p>This episode is intended to show how we can assemble a
SummarizedExperiment starting from individual count, rowdata and coldata
files. Moreover, we will practice adding annotations for the genes, and
discuss related concepts and things to keep in mind (annotation sources,
versions, ‘helper’ packages like tximeta).</p>
</div>
</div>
</div>
<section id="read-the-data"><h2 class="section-heading">Read the data<a class="anchor" aria-label="anchor" href="#read-the-data"></a>
</h2>
<hr class="half-width">
<p>Let’t read in the data files that we downloaded in the last
episode:</p>
<div class="section level3">
<h3 id="counts">Counts<a class="anchor" aria-label="anchor" href="#counts"></a>
</h3>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/GSE96870_counts_cerebellum.csv"</span>, </span>
<span>                   row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="sample-annotations">Sample annotations<a class="anchor" aria-label="anchor" href="#sample-annotations"></a>
</h3>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">coldata</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/GSE96870_coldata_cerebellum.csv"</span>,</span>
<span>                    row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre>
</div>
</div>
<div class="section level3">
<h3 id="gene-annotations">Gene annotations<a class="anchor" aria-label="anchor" href="#gene-annotations"></a>
</h3>
<p>Need to be careful - the descriptions contain both commas and ’
(e.g., 5’)</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rowranges</span> <span class="op">&lt;-</span> <span class="fu">read.delim</span><span class="op">(</span><span class="st">"data/GSE96870_rowranges.tsv"</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, </span>
<span>                        colClasses <span class="op">=</span> <span class="fu">c</span><span class="op">(</span>ENTREZID <span class="op">=</span> <span class="st">"character"</span><span class="op">)</span>,</span>
<span>                        header <span class="op">=</span> <span class="cn">TRUE</span>, quote <span class="op">=</span> <span class="st">""</span>, row.names <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<p>Mention other ways of getting annotations, and practice querying org
package. Important to use the right annotation source/version.</p>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">suppressPackageStartupMessages</span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Mm.eg.db</span>, keys <span class="op">=</span> <span class="st">"497097"</span>, column <span class="op">=</span> <span class="st">"SYMBOL"</span>, keytype <span class="op">=</span> <span class="st">"ENTREZID"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code><span><span class="fl">497097</span> </span>
<span><span class="st">"Xkr4"</span> </span></code></pre>
</div>
<p>Check feature types</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">rowranges</span><span class="op">$</span><span class="va">gbkey</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
     C_region     D_segment          exon     J_segment      misc_RNA 
           20            23          4008            94          1988 
         mRNA         ncRNA precursor_RNA          rRNA          tRNA 
        21198         12285          1187            35           413 
    V_segment 
          535 </code></pre>
</div>
</div>
</section><section id="assemble-summarizedexperiment"><h2 class="section-heading">Assemble SummarizedExperiment<a class="anchor" aria-label="anchor" href="#assemble-summarizedexperiment"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">stopifnot</span><span class="op">(</span><span class="fu">rownames</span><span class="op">(</span><span class="va">rowranges</span><span class="op">)</span> <span class="op">==</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span>,</span>
<span>          <span class="fu">rownames</span><span class="op">(</span><span class="va">coldata</span><span class="op">)</span> <span class="op">==</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">SummarizedExperiment</span><span class="op">(</span></span>
<span>    assays <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>counts <span class="op">=</span> <span class="fu">as.matrix</span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    rowRanges <span class="op">=</span> <span class="fu">as</span><span class="op">(</span><span class="va">rowranges</span>, <span class="st">"GRanges"</span><span class="op">)</span>,</span>
<span>    colData <span class="op">=</span> <span class="va">coldata</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
</section><section id="save-summarizedexperiment"><h2 class="section-heading">Save SummarizedExperiment<a class="anchor" aria-label="anchor" href="#save-summarizedexperiment"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">saveRDS</span><span class="op">(</span><span class="va">se</span>, <span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span></code></pre>
</div>
</section><section id="session-info"><h2 class="section-heading">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.3.0 (2023-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] stats4    stats     graphics  grDevices utils     datasets  methods  
[8] base     

other attached packages:
 [1] org.Mm.eg.db_3.17.0         AnnotationDbi_1.62.2       
 [3] SummarizedExperiment_1.30.2 Biobase_2.60.0             
 [5] MatrixGenerics_1.12.2       matrixStats_1.0.0          
 [7] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        
 [9] IRanges_2.34.1              S4Vectors_0.38.1           
[11] BiocGenerics_0.46.0        

loaded via a namespace (and not attached):
 [1] Matrix_1.5-1            bit_4.0.5               compiler_4.3.0         
 [4] BiocManager_1.30.21     renv_1.0.0              crayon_1.5.2           
 [7] blob_1.2.4              Biostrings_2.68.1       bitops_1.0-7           
[10] png_0.1-8               fastmap_1.1.1           yaml_2.3.7             
[13] lattice_0.21-8          R6_2.5.1                XVector_0.40.0         
[16] S4Arrays_1.0.4          knitr_1.43              DelayedArray_0.26.6    
[19] GenomeInfoDbData_1.2.10 DBI_1.1.3               rlang_1.1.1            
[22] KEGGREST_1.40.0         cachem_1.0.8            xfun_0.39              
[25] bit64_4.0.5             RSQLite_2.3.1           memoise_2.0.1          
[28] cli_3.6.1               zlibbioc_1.46.0         grid_4.3.0             
[31] rstudioapi_0.15.0       vctrs_0.6.3             evaluate_0.21          
[34] RCurl_1.98-1.12         httr_1.4.6              pkgconfig_2.0.3        
[37] tools_4.3.0            </code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Depending on the gene expression quantification tool used, there are
different ways (often distributed in Bioconductor packages) to read the
output into a SummarizedExperiment or DGEList object for further
processing in R.</li>
<li>Stable gene identifiers such as Ensembl or Entrez IDs should
preferably be used as the main identifiers throughout an RNA-seq
analysis, with gene symbols added for easier interpretation.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-04-exploratory-qc"><p>Content from <a href="04-exploratory-qc.html">Exploratory analysis and quality control</a></p>
<hr>
<p> Last updated on 2023-07-17 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/04-exploratory-qc.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>Why is exploratory analysis an essential part of an RNA-seq
analysis?</li>
<li>How should one preprocess the raw count matrix for exploratory
analysis?<br>
</li>
<li>Are two dimensions sufficient to represent your data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn how to explore the gene expression matrix and perform common
quality control steps.</li>
<li>Learn how to set up an interactive application for exploratory
analysis.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div id="contribute" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute"></a>
</h3>
<div class="callout-content">
<p>This episode is intended to introduce various types of exploratory
analysis and QC steps taken before a formal statistical analysis is
done.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">suppressPackageStartupMessages</span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">vsn</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ComplexHeatmap</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">RColorBrewer</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">hexbin</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span></code></pre>
</div>
<p>Exploratory analysis is crucial for quality control and to get to
know our data. It can help us detect quality problems, sample swaps and
contamination, as well as give us a sense of the most salient patterns
present in the data. In this episode, we will learn about two common
ways of performing exploratory analysis for RNA-seq data; namely
clustering and principal component analysis (PCA). These tools are in no
way limited to (or developed for) analysis of RNA-seq data. However,
there are certain characteristics of count assays that need to be taken
into account when they are applied to this type of data.</p>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">se</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">5</span>, <span class="op">]</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span><span class="op">[</span>, <span class="va">se</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"Cerebellum"</span><span class="op">]</span>,</span>
<span>                            design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in DESeq2::DESeqDataSet(se[, se$tissue == "Cerebellum"], design = ~sex
+ : some variables in design formula are characters, converting to factors</code></pre>
</div>
<section id="library-size-differences"><h2 class="section-heading">Library size differences<a class="anchor" aria-label="anchor" href="#library-size-differences"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span>sample <span class="op">=</span> <span class="fu">colnames</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>, </span>
<span>                  libSize <span class="op">=</span> <span class="fu">colSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>       <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">sample</span>, y <span class="op">=</span> <span class="va">libSize</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Sample"</span>, y <span class="op">=</span> <span class="st">"Total count"</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">theme</span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu">element_text</span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span>, hjust <span class="op">=</span> <span class="fl">1</span>, vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/04-exploratory-qc-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Differences in the total number of reads assigned to genes between
samples typically occur for technical reasons. In practice, it means
that we can not simply compare the raw read count directly between
samples and conclude that a sample with a higher read count also
expresses the gene more strongly - the higher count may be caused by an
overall higher number of reads in that sample. In the rest of this
section, we will use the term <em>library size</em> to refer to the
total number of reads assigned to genes for a sample. We need to adjust
for the differences in library size between samples, to avoid drawing
incorrect conclusions. The way this is typically done for RNA-seq data
can be described as a two-step procedure. First, we estimate <em>size
factors</em> - sample-specific correction factors such that if the raw
counts were to be divided by these factors, the resulting values would
be more comparable across samples. Next, these size factors are
incorporated into the statistical analysis of the data. It is important
to pay close attention to how this is done in practice for a given
analysis method. Sometimes the division of the counts by the size
factors needs to be done explicitly by the analyst. Other times (as we
will see for the differential expression analysis) it is important that
they are provided separately to the analysis tool, which will then use
them appropriately in the statistical model.</p>
<p>With <code>DESeq2</code>, size factors are calculated using the
<code>estimateSizeFactors()</code> function. The size factors estimated
by this function combines an adjustment for differences in library sizes
with an adjustment for differences in the RNA composition of the
samples. The latter is important due to the compositional nature of
RNA-seq data. There is a fixed number of reads to distribute between the
genes, and if a single (or a few) very highly expressed gene consume a
large part of the reads, all other genes will consequently receive very
low counts.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">estimateSizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">data.frame</span><span class="op">(</span>libSize <span class="op">=</span> <span class="fu">colSums</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">dds</span>, <span class="st">"counts"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                  sizeFactor <span class="op">=</span> <span class="fu">sizeFactors</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span><span class="op">)</span>,</span>
<span>       <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">libSize</span>, y <span class="op">=</span> <span class="va">sizeFactor</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">geom_point</span><span class="op">(</span>size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span> <span class="fu">theme_bw</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">labs</span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Library size"</span>, y <span class="op">=</span> <span class="st">"Size factor"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/04-exploratory-qc-rendered-unnamed-chunk-6-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="transform-data"><h2 class="section-heading">Transform data<a class="anchor" aria-label="anchor" href="#transform-data"></a>
</h2>
<hr class="half-width">
<p>There is a rich literature on methods for exploratory analysis. Most
of these work best in situations where the variance of the input data
(here, each gene) is relatively independent of the average value. For
read count data such as RNA-seq, this is not the case. In fact, the
variance increases with the average read count.</p>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">meanSdPlot</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span>, ranks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/04-exploratory-qc-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>There are two ways around this: either we develop methods
specifically adapted to count data, or we adapt (transform) the count
data so that the existing methods are applicable. Both ways have been
explored; however, at the moment the second approach is arguably more
widely applied in practice.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">vst</span><span class="op">(</span><span class="va">dds</span>, blind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu">meanSdPlot</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span>, ranks <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/04-exploratory-qc-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="heatmaps-and-clustering"><h2 class="section-heading">Heatmaps and clustering<a class="anchor" aria-label="anchor" href="#heatmaps-and-clustering"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dst</span> <span class="op">&lt;-</span> <span class="fu">dist</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu">colorRampPalette</span><span class="op">(</span><span class="fu">brewer.pal</span><span class="op">(</span><span class="fl">9</span>, <span class="st">"Blues"</span><span class="op">)</span><span class="op">)</span><span class="op">(</span><span class="fl">255</span><span class="op">)</span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu">Heatmap</span><span class="op">(</span></span>
<span>    <span class="fu">as.matrix</span><span class="op">(</span><span class="va">dst</span><span class="op">)</span>, </span>
<span>    col <span class="op">=</span> <span class="va">colors</span>,</span>
<span>    name <span class="op">=</span> <span class="st">"Euclidean\ndistance"</span>,</span>
<span>    cluster_rows <span class="op">=</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">dst</span><span class="op">)</span>,</span>
<span>    cluster_columns <span class="op">=</span> <span class="fu">hclust</span><span class="op">(</span><span class="va">dst</span><span class="op">)</span>,</span>
<span>    bottom_annotation <span class="op">=</span> <span class="fu">columnAnnotation</span><span class="op">(</span></span>
<span>        sex <span class="op">=</span> <span class="va">vsd</span><span class="op">$</span><span class="va">sex</span>,</span>
<span>        time <span class="op">=</span> <span class="va">vsd</span><span class="op">$</span><span class="va">time</span>,</span>
<span>        col <span class="op">=</span> <span class="fu">list</span><span class="op">(</span>sex <span class="op">=</span> <span class="fu">c</span><span class="op">(</span>Female <span class="op">=</span> <span class="st">"red"</span>, Male <span class="op">=</span> <span class="st">"blue"</span><span class="op">)</span>,</span>
<span>                   time <span class="op">=</span> <span class="fu">c</span><span class="op">(</span>Day0 <span class="op">=</span> <span class="st">"yellow"</span>, Day4 <span class="op">=</span> <span class="st">"forestgreen"</span>, Day8 <span class="op">=</span> <span class="st">"purple"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/04-exploratory-qc-rendered-unnamed-chunk-9-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="pca"><h2 class="section-heading">PCA<a class="anchor" aria-label="anchor" href="#pca"></a>
</h2>
<hr class="half-width">
<p>Principal component analysis is a dimensionality reduction method,
which projects the samples into a lower-dimensional space. This
lower-dimensional representation can be used for visualization, or as
the input for other analysis methods. The principal components are
defined in such a way that they are orthogonal, and that the projection
of the samples into the space they span contains as much variance as
possible. It is an <em>unsupervised</em> method in the sense that no
external information about the samples (e.g., the treatment condition)
is taken into account. In the plot below we represent the samples in a
two-dimensional principal component space. For each of the two
dimensions, we indicate the fraction of the total variance that is
represented by that component. By definition, the first principal
component will always represent more of the variance than the subsequent
ones. The fraction of explained variance is a measure of how much of the
‘signal’ in the data that is retained when we project the samples from
the original, high-dimensional space to the low-dimensional space for
visualization.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pcaData</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">plotPCA</span><span class="op">(</span><span class="va">vsd</span>, intgroup <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"time"</span><span class="op">)</span>,</span>
<span>                           returnData <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">percentVar</span> <span class="op">&lt;-</span> <span class="fu">round</span><span class="op">(</span><span class="fl">100</span> <span class="op">*</span> <span class="fu">attr</span><span class="op">(</span><span class="va">pcaData</span>, <span class="st">"percentVar"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">pcaData</span>, <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">PC1</span>, y <span class="op">=</span> <span class="va">PC2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>color <span class="op">=</span> <span class="va">sex</span>, shape <span class="op">=</span> <span class="va">time</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">theme_minimal</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">xlab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC1: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"PC2: "</span>, <span class="va">percentVar</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span>, <span class="st">"% variance"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">coord_fixed</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span> </span>
<span>    <span class="fu">scale_color_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span>Male <span class="op">=</span> <span class="st">"blue"</span>, Female <span class="op">=</span> <span class="st">"red"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/04-exploratory-qc-rendered-unnamed-chunk-10-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="session-info"><h2 class="section-heading">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessionInfo</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>R version 4.3.0 (2023-04-21)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 22.04.2 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0

locale:
 [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       
 [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   
 [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          
[10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   

time zone: UTC
tzcode source: system (glibc)

attached base packages:
[1] grid      stats4    stats     graphics  grDevices utils     datasets 
[8] methods   base     

other attached packages:
 [1] hexbin_1.28.3               RColorBrewer_1.1-3         
 [3] ComplexHeatmap_2.16.0       ggplot2_3.4.2              
 [5] vsn_3.68.0                  DESeq2_1.40.2              
 [7] SummarizedExperiment_1.30.2 Biobase_2.60.0             
 [9] MatrixGenerics_1.12.2       matrixStats_1.0.0          
[11] GenomicRanges_1.52.0        GenomeInfoDb_1.36.1        
[13] IRanges_2.34.1              S4Vectors_0.38.1           
[15] BiocGenerics_0.46.0        

loaded via a namespace (and not attached):
 [1] shape_1.4.6             circlize_0.4.15         gtable_0.3.3           
 [4] rjson_0.2.21            xfun_0.39               GlobalOptions_0.1.2    
 [7] lattice_0.21-8          vctrs_0.6.3             tools_4.3.0            
[10] bitops_1.0-7            generics_0.1.3          parallel_4.3.0         
[13] tibble_3.2.1            fansi_1.0.4             highr_0.10             
[16] cluster_2.1.4           pkgconfig_2.0.3         Matrix_1.5-1           
[19] lifecycle_1.0.3         GenomeInfoDbData_1.2.10 farver_2.1.1           
[22] compiler_4.3.0          munsell_0.5.0           codetools_0.2-19       
[25] clue_0.3-64             RCurl_1.98-1.12         yaml_2.3.7             
[28] preprocessCore_1.62.1   pillar_1.9.0            crayon_1.5.2           
[31] BiocParallel_1.34.2     affy_1.78.1             DelayedArray_0.26.6    
[34] limma_3.56.2            iterators_1.0.14        foreach_1.5.2          
[37] tidyselect_1.2.0        locfit_1.5-9.8          digest_0.6.33          
[40] dplyr_1.1.2             labeling_0.4.2          colorspace_2.1-0       
[43] cli_3.6.1               magrittr_2.0.3          S4Arrays_1.0.4         
[46] utf8_1.2.3              withr_2.5.0             scales_1.2.1           
[49] XVector_0.40.0          affyio_1.70.0           GetoptLong_1.0.5       
[52] png_0.1-8               evaluate_0.21           knitr_1.43             
[55] doParallel_1.0.17       rlang_1.1.1             Rcpp_1.0.11            
[58] glue_1.6.2              BiocManager_1.30.21     renv_1.0.0             
[61] rstudioapi_0.15.0       R6_2.5.1                zlibbioc_1.46.0        </code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>Exploratory analysis is essential for quality control and to detect
potential problems with a data set.</li>
<li>Different classes of exploratory analysis methods expect differently
preprocessed data. The most commonly used methods expect counts to be
normalized and log-transformed (or similar- more
sensitive/sophisticated), to be closer to homoskedastic. Other methods
work directly on the raw counts.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-05-differential-expression"><p>Content from <a href="05-differential-expression.html">Differential expression analysis</a></p>
<hr>
<p> Last updated on 2023-07-17 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/05-differential-expression.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What are the steps performed in a typical differential expression
analysis?</li>
<li>How does one interpret the output of DESeq2?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the steps involved in a differential expression
analysis.</li>
<li>Explain how to perform these steps in R, using DESeq2.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<div id="contribute" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute"></a>
</h3>
<div class="callout-content">
<p>This episode is intended to introduce the concepts required to
perform differential expression analysis with RNA-seq data. Explain
concepts like size factors, count modeling (Negative Binomial),
dispersion, interpretation of the test output, multiple testing
correction.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">suppressPackageStartupMessages</span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ExploreModelMatrix</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">cowplot</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ComplexHeatmap</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<section id="load-data"><h2 class="section-heading">Load data<a class="anchor" aria-label="anchor" href="#load-data"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span></code></pre>
</div>
</section><section id="create-deseqdataset"><h2 class="section-heading">Create DESeqDataSet<a class="anchor" aria-label="anchor" href="#create-deseqdataset"></a>
</h2>
<hr class="half-width">
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span><span class="op">[</span>, <span class="va">se</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"Cerebellum"</span><span class="op">]</span>,</span>
<span>                            design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning in DESeq2::DESeqDataSet(se[, se$tissue == "Cerebellum"], design = ~sex
+ : some variables in design formula are characters, converting to factors</code></pre>
</div>
</section><section id="run-deseq"><h2 class="section-heading">Run DESeq()<a class="anchor" aria-label="anchor" href="#run-deseq"></a>
</h2>
<hr class="half-width">
<div id="contribute-1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute-1" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute-1"></a>
</h3>
<div class="callout-content">
<p>The concepts may be clearer if the steps of DESeq() are first
performed separately, followed by a note that they can be performed in a
single step using DESeq().</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating size factors</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>estimating dispersions</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>gene-wise dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>mean-dispersion relationship</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>final dispersion estimates</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>fitting model and testing</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plotDispEsts</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/05-differential-expression-rendered-unnamed-chunk-6-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="extract-results-for-specific-contrasts"><h2 class="section-heading">Extract results for specific contrasts<a class="anchor" aria-label="anchor" href="#extract-results-for-specific-contrasts"></a>
</h2>
<hr class="half-width">
<div id="contribute-2" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute-2" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute-2"></a>
</h3>
<div class="callout-content">
<p>Refer back to the episode about experimental design.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Day 8 vs Day 0</span></span>
<span><span class="va">resTime</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">resTime</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 32652 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 4472, 14%
LFC &lt; 0 (down)     : 4276, 13%
outliers [1]       : 10, 0.031%
low counts [2]     : 8732, 27%
(mean count &lt; 1)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">resTime</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">resTime</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>log2 fold change (MLE): time Day8 vs Day0 
Wald test p-value: time Day8 vs Day0 
DataFrame with 6 rows and 6 columns
               baseMean log2FoldChange     lfcSE      stat      pvalue
              &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;
Asl             701.343        1.11733 0.0592541   18.8565 2.59885e-79
Apod          18765.146        1.44698 0.0805186   17.9708 3.30147e-72
Cyp2d22        2550.480        0.91020 0.0554756   16.4072 1.69794e-60
Klk6            546.503       -1.67190 0.1058989  -15.7877 3.78228e-56
Fcrls           184.235       -1.94701 0.1279847  -15.2128 2.90708e-52
A330076C08Rik   107.250       -1.74995 0.1154279  -15.1606 6.45112e-52
                     padj
                &lt;numeric&gt;
Asl           6.21386e-75
Apod          3.94690e-68
Cyp2d22       1.35326e-56
Klk6          2.26086e-52
Fcrls         1.39017e-48
A330076C08Rik 2.57077e-48</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">plotMA</span><span class="op">(</span><span class="va">resTime</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/05-differential-expression-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Male vs Female</span></span>
<span><span class="va">resSex</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">summary</span><span class="op">(</span><span class="va">resSex</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
out of 32652 with nonzero total read count
adjusted p-value &lt; 0.1
LFC &gt; 0 (up)       : 53, 0.16%
LFC &lt; 0 (down)     : 71, 0.22%
outliers [1]       : 10, 0.031%
low counts [2]     : 13717, 42%
(mean count &lt; 6)
[1] see 'cooksCutoff' argument of ?results
[2] see 'independentFiltering' argument of ?results</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">resSex</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">resSex</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">]</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>log2 fold change (MLE): sex Male vs Female 
Wald test p-value: sex Male vs Female 
DataFrame with 6 rows and 6 columns
               baseMean log2FoldChange     lfcSE      stat       pvalue
              &lt;numeric&gt;      &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;    &lt;numeric&gt;
Xist         22603.0359      -11.60429  0.336282  -34.5076 6.16852e-261
Ddx3y         2072.9436       11.87241  0.397493   29.8683 5.08722e-196
Eif2s3y       1410.8750       12.62514  0.565216   22.3369 1.62066e-110
Kdm5d          692.1672       12.55386  0.593627   21.1477  2.89566e-99
Uty            667.4375       12.01728  0.593591   20.2451  3.92780e-91
LOC105243748    52.9669        9.08325  0.597624   15.1989  3.59432e-52
                     padj
                &lt;numeric&gt;
Xist         1.16739e-256
Ddx3y        4.81378e-192
Eif2s3y      1.02237e-106
Kdm5d         1.37001e-95
Uty           1.48667e-87
LOC105243748  1.13371e-48</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">plotMA</span><span class="op">(</span><span class="va">resSex</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/05-differential-expression-rendered-unnamed-chunk-7-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="visualize-selected-set-of-genes"><h2 class="section-heading">Visualize selected set of genes<a class="anchor" aria-label="anchor" href="#visualize-selected-set-of-genes"></a>
</h2>
<hr class="half-width">
<div id="contribute-3" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="contribute-3" class="callout-inner">
<h3 class="callout-title">Contribute!<a class="anchor" aria-label="anchor" href="#contribute-3"></a>
</h3>
<div class="callout-content">
<p>Here we intend to practice how to interpret the results from the
differential expression analysis. Refer back to the exploratory/QC
episode.</p>
</div>
</div>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vsd</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">vst</span><span class="op">(</span><span class="va">dds</span>, blind <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="fu">head</span><span class="op">(</span><span class="va">resTime</span><span class="op">[</span><span class="fu">order</span><span class="op">(</span><span class="va">resTime</span><span class="op">$</span><span class="va">pvalue</span><span class="op">)</span>, <span class="op">]</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">heatmapData</span> <span class="op">&lt;-</span> <span class="fu">assay</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span><span class="va">genes</span>, <span class="op">]</span></span>
<span><span class="va">heatmapData</span> <span class="op">&lt;-</span> <span class="fu">t</span><span class="op">(</span><span class="fu">scale</span><span class="op">(</span><span class="fu">t</span><span class="op">(</span><span class="va">heatmapData</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">heatmapColAnnot</span> <span class="op">&lt;-</span> <span class="fu">data.frame</span><span class="op">(</span><span class="fu">colData</span><span class="op">(</span><span class="va">vsd</span><span class="op">)</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"sex"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">idx</span> <span class="op">&lt;-</span> <span class="fu">order</span><span class="op">(</span><span class="va">vsd</span><span class="op">$</span><span class="va">time</span><span class="op">)</span></span>
<span><span class="va">heatmapData</span> <span class="op">&lt;-</span> <span class="va">heatmapData</span><span class="op">[</span>, <span class="va">idx</span><span class="op">]</span></span>
<span><span class="va">heatmapColAnnot</span> <span class="op">&lt;-</span> <span class="fu">HeatmapAnnotation</span><span class="op">(</span>df <span class="op">=</span> <span class="va">heatmapColAnnot</span><span class="op">[</span><span class="va">idx</span>, <span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">ComplexHeatmap</span><span class="fu">::</span><span class="fu">Heatmap</span><span class="op">(</span><span class="va">heatmapData</span>,</span>
<span>                        top_annotation <span class="op">=</span> <span class="va">heatmapColAnnot</span>,</span>
<span>                        cluster_rows <span class="op">=</span> <span class="cn">TRUE</span>, cluster_columns <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/05-differential-expression-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>With DESeq2, the main steps of a differential expression analysis
(size factor estimation, dispersion estimation, calculation of test
statistics) are wrapped in a single function: DESeq().</li>
<li>Independent filtering of lowly expressed genes is often
beneficial.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-06-gene-set-analysis"><p>Content from <a href="06-gene-set-analysis.html">Gene set enrichment analysis</a></p>
<hr>
<p> Last updated on 2023-07-18 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/06-gene-set-analysis.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<style>
body h3 {
    margin-top: 50px;
}
</style>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>What is the aim of performing gene set enrichment analysis?</li>
<li>What is the method of over-representation analysis?</li>
<li>What are the commonly-used gene set databases?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Learn the method of gene set enrichment analysis.</li>
<li>Learn how to obtain gene sets from various resources in R.</li>
<li>Learn how to perform gene set enrichment analysis and how to
visualize enrichment results.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<p>After we have obtained a list of differentially expressed (DE) genes,
the next question naturally to ask is what biological functions these DE
genes may affect. Gene set enrichment analysis (GSEA) evaluates the
associations of a list of DE genes to a collection of pre-defined gene
sets, where each gene set has a specific biological meaning. Once DE
genes are significantly enriched in a gene set, the conclusion is made
that the corresponding biological meaning (e.g. a biological process or
a pathway) is significantly affected.</p>
<p>The definition of a gene set is very flexible and the construction of
gene sets is straightforward. In most cases, gene sets are from public
databases where huge efforts from scientific curators have already been
made to carefully categorize genes into gene sets with clear biological
meanings. Nevertheless, gene sets can also be self-defined from
individual studies, such as a set of genes in a network module from a
co-expression network analysis, or a set of genes that are up-regulated
in a certain disease.</p>
<p>There are a huge amount of methods available for GSEA analysis. In
this episode, we will learn the simplest but the mostly used one: the
over-representation analysis (ORA). ORA is directly applied to the list
of DE genes and it evaluates the association of the DE genes and the
gene set by the numbers of genes in different categories.</p>
<p>Please note, ORA is a universal method that it can not only be
applied to the DE gene list, but also any type of gene list of interest
to look for their statistically associated biological meanings.</p>
<p>In this episode, we will start with a tiny example to illustrate the
statistical method of ORA. Next we will go through several commonly-used
gene set databases and how to access them in R. Then, we will learn how
to perform ORA analysis with the Bioconductor package
<strong>clusterProfiler</strong>. And in the end, we will learn several
visualization methods on the GSEA results.</p>
<section id="the-statistical-method"><h2 class="section-heading">The statistical method<a class="anchor" aria-label="anchor" href="#the-statistical-method"></a>
</h2>
<hr class="half-width">
<p>To demonstrate the ORA analysis, we use a list of DE genes from a
comparison between genders. The following code performs
<strong>DESeq2</strong> analysis which you should have already learnt in
the previous episode. In the end, we have a list of DE genes filtered by
FDR &lt; 0.05, and save it in the object <code>sexDEgenes</code>.</p>
<p>The file <code>data/GSE96870_se.rds</code> contains a
<code>RangedSummarizedExperiment</code> and it is downloaded and
constructed in the script <a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/episodes/download_data.R" class="external-link"><code>download_data.R</code></a>.
In the following code, there are also comments that explain every step
of the analysis.</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># download example datasets and generate `GSE96870_se.rds`</span></span>
<span><span class="kw">source</span><span class="op">(</span><span class="st">"download_data.R"</span><span class="op">)</span></span>
<span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">DESeq2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># read the example dataset which is a `RangedSummarizedExperiment` object</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="fu">readRDS</span><span class="op">(</span><span class="st">"data/GSE96870_se.rds"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># only restrict to mRNA (protein-coding genes)</span></span>
<span><span class="va">se</span> <span class="op">&lt;-</span> <span class="va">se</span><span class="op">[</span><span class="fu">rowData</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">$</span><span class="va">gbkey</span> <span class="op">==</span> <span class="st">"mRNA"</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># construct a `DESeqDataSet` object where we also specify the experimental design</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeqDataSet</span><span class="op">(</span><span class="va">se</span><span class="op">[</span>, <span class="va">se</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"Cerebellum"</span><span class="op">]</span>,</span>
<span>                            design <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">time</span><span class="op">)</span></span>
<span><span class="co"># perform DESeq2 analysis</span></span>
<span><span class="va">dds</span> <span class="op">&lt;-</span> <span class="fu">DESeq</span><span class="op">(</span><span class="va">dds</span><span class="op">)</span></span>
<span><span class="co"># obtain DESeq2 results, here we only want Male vs Female in the "sex" variable</span></span>
<span><span class="va">resSex</span> <span class="op">&lt;-</span> <span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"sex"</span>, <span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># extract DE genes with padj &lt; 0.05</span></span>
<span><span class="va">sexDE</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resSex</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># the list of DE genes</span></span>
<span><span class="va">sexDEgenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">sexDE</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s check the number of DE genes and how they look like. It seems
the number of DE genes is very small, but it is OK for this example.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">sexDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 54</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb4">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">head</span><span class="op">(</span><span class="va">sexDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Lgr6"   "Myoc"   "Fibcd1" "Kcna4"  "Ctxn2"  "S100a9"</code></pre>
</div>
<p>Next we construct a gene set which contains genes on sex chromosomes
(let’s call it the “<em>XY gene set</em>”). Recall the
<code>RangedSummarizedExperiment</code> object also includes genomic
locations of genes, thus we can simply obtain sex genes by filtering the
chromosome names.</p>
<p>In the following code, <code>geneGR</code> is a <code>GRanges</code>
object on which <code>seqnames()</code> is applied to extract chromosome
names. <code>seqnames()</code> returns a special data format and we need
to explicitly convert it to a normal vector by
<code>as.vector()</code>.</p>
<div class="codewrapper sourceCode" id="cb6">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">geneGR</span> <span class="op">&lt;-</span> <span class="fu">rowRanges</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">totalGenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">XYGeneSet</span> <span class="op">&lt;-</span> <span class="va">totalGenes</span><span class="op">[</span><span class="fu">as.vector</span><span class="op">(</span><span class="fu">seqnames</span><span class="op">(</span><span class="va">geneGR</span><span class="op">)</span><span class="op">)</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"X"</span>, <span class="st">"Y"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">XYGeneSet</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "Gm21950"   "Gm14346"   "Gm14345"   "Gm14351"   "Spin2-ps1" "Gm3701"   </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb8">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">XYGeneSet</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1134</code></pre>
</div>
<p>The format of a single gene set is very straightforward, which is
simply a vector. The ORA analysis is applied on the DE gene vector and
gene set vector.</p>
<p>Before we move on, one thing worth to mention is that ORA deals with
two gene vectors. To correctly map between them, gene ID types must be
consistent in the two vectors. In this tiny example, since both DE genes
and the <em>XY gene set</em> are from the same object <code>se</code>,
they are ensured to be in the same gene ID types (the gene symbol). But
in general, DE genes and gene sets are from two different sources
(e.g. DE genes are from researcher’s experiment and gene sets are from
public databases), it is very possible that gene IDs are not consistent
in the two. Later in this episode, we will learn how to perform gene ID
conversion in the ORA analysis.</p>
<p>Since the DE genes and the gene set can be mathematically thought as
two sets, a natural way is to first visualize them with a Venn
diagram.</p>
<div class="codewrapper sourceCode" id="cb10">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">gplots</span><span class="op">)</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="fu">venn</span><span class="op">(</span><span class="fu">list</span><span class="op">(</span><span class="st">"sexDEgenes"</span>  <span class="op">=</span> <span class="va">sexDEgenes</span>, </span>
<span>               <span class="st">"XY gene set"</span> <span class="op">=</span> <span class="va">XYGeneSet</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">title</span><span class="op">(</span><span class="fu">paste0</span><span class="op">(</span><span class="st">"|universe| = "</span>, <span class="fu">length</span><span class="op">(</span><span class="va">totalGenes</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-gene-set-analysis-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In the Venn diagram, we can observe that around 1.1% (13/1134) of
genes in the <em>XY gene set</em> are DE. Compared to the global
fraction of DE genes (54/21198 = 0.25%), it seems there is a strong
relations between DE genes and the gene set. We can also compare the
fraction of DE genes that belong to the gene set (13/54 = 24.1%) and the
global fraction of <em>XY gene set</em> in the genome (1134/21198 =
5.3%). On the other hand, it is quite expected because the two events
are actually biologically relevant where one is from a comparison
between genders and the other is the set of gender-related genes.</p>
<p>Then, how to statistically measure the enrichment or
over-representation? Let’s go to the next section.</p>
<div class="section level3">
<h3 id="fishers-exact-test">Fisher’s exact test<a class="anchor" aria-label="anchor" href="#fishers-exact-test"></a>
</h3>
<p>To statistically measure the enrichment, the relationship of DE genes
and the gene set is normally formatted into the following 2x2 contigency
table, where in the table are the numbers of genes in different
categories. <span class="math inline">\(n_{+1}\)</span> is the size of
the <em>XY gene set</em> (i.e. the number of member genes), <span class="math inline">\(n_{1+}\)</span> is the number of DE genes, <span class="math inline">\(n\)</span> is the number of total genes.</p>
<center>
<table class="table">
<thead><tr class="header">
<th></th>
<th>In the gene set</th>
<th>Not in the gene set</th>
<th>Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>DE</strong></td>
<td><span class="math inline">\(n_{11}\)</span></td>
<td><span class="math inline">\(n_{12}\)</span></td>
<td><span class="math inline">\(n_{1+}\)</span></td>
</tr>
<tr class="even">
<td><strong>Not DE</strong></td>
<td><span class="math inline">\(n_{21}\)</span></td>
<td><span class="math inline">\(n_{22}\)</span></td>
<td><span class="math inline">\(n_{2+}\)</span></td>
</tr>
<tr class="odd">
<td><strong>Total</strong></td>
<td><span class="math inline">\(n_{+1}\)</span></td>
<td><span class="math inline">\(n_{+2}\)</span></td>
<td><span class="math inline">\(n\)</span></td>
</tr>
</tbody>
</table>
</center>
<p>These numbers can be obtained as in the following code<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>. Note we replace
<code>+</code> with <code>0</code> in the R variable names.</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n</span>    <span class="op">&lt;-</span> <span class="fu">nrow</span><span class="op">(</span><span class="va">se</span><span class="op">)</span></span>
<span><span class="va">n_01</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">XYGeneSet</span><span class="op">)</span></span>
<span><span class="va">n_10</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="va">sexDEgenes</span><span class="op">)</span></span>
<span><span class="va">n_11</span> <span class="op">&lt;-</span> <span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">sexDEgenes</span>, <span class="va">XYGeneSet</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<p>Other values can be obtained by:</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_12</span> <span class="op">&lt;-</span> <span class="va">n_10</span> <span class="op">-</span> <span class="va">n_11</span></span>
<span><span class="va">n_21</span> <span class="op">&lt;-</span> <span class="va">n_01</span> <span class="op">-</span> <span class="va">n_11</span></span>
<span><span class="va">n_20</span> <span class="op">&lt;-</span> <span class="va">n</span>    <span class="op">-</span> <span class="va">n_10</span></span>
<span><span class="va">n_02</span> <span class="op">&lt;-</span> <span class="va">n</span>    <span class="op">-</span> <span class="va">n_01</span></span>
<span><span class="va">n_22</span> <span class="op">&lt;-</span> <span class="va">n_02</span> <span class="op">-</span> <span class="va">n_12</span></span></code></pre>
</div>
<p>All the values are:</p>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">n_11</span>, <span class="va">n_12</span>, <span class="va">n_10</span>, <span class="va">n_21</span>, <span class="va">n_22</span>, <span class="va">n_20</span>, <span class="va">n_01</span>, <span class="va">n_02</span>, <span class="va">n</span><span class="op">)</span>,</span>
<span>    nrow <span class="op">=</span> <span class="fl">3</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     [,1]  [,2]  [,3]
[1,]   13    41    54
[2,] 1121 20023 21144
[3,] 1134 20064 21198</code></pre>
</div>
<p>And we fill these numbers into the 2x2 contigency table:</p>
<center>
<table class="table">
<thead><tr class="header">
<th></th>
<th>In the gene set</th>
<th>Not in the gene set</th>
<th>Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>DE</strong></td>
<td>13</td>
<td>41</td>
<td>54</td>
</tr>
<tr class="even">
<td><strong>Not DE</strong></td>
<td>1121</td>
<td>20023</td>
<td>21144</td>
</tr>
<tr class="odd">
<td><strong>Total</strong></td>
<td>1134</td>
<td>20064</td>
<td>21198</td>
</tr>
</tbody>
</table>
</center>
<p>Fisher’s exact test can be used to test the associations of the two
marginal attributes, i.e. is there a dependency of a gene to be a DE
gene and to be in the <em>XY gene set</em>? In R, we can use the
function <code>fisher.test()</code> to perform the test. The input is
the top-left 2x2 sub-matrix. We specify
<code>alternative = "greater"</code> in the function because we are only
interested in over-representation.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fisher.test</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">n_11</span>, <span class="va">n_12</span>, <span class="va">n_21</span>, <span class="va">n_22</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Fisher's Exact Test for Count Data

data:  matrix(c(n_11, n_12, n_21, n_22), nrow = 2, byrow = TRUE)
p-value = 3.906e-06
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 3.110607      Inf
sample estimates:
odds ratio 
  5.662486 </code></pre>
</div>
<p>In the output, we can see the <em>p</em>-value is very small
(<code>3.906e-06</code>), then we can conclude DE genes have a very
strong enrichment in the <em>XY gene set</em>.</p>
<p>Results of the Fisher’s Exact test can be saved into an object
<code>t</code>, which is a simple list, and the <em>p</em>-value can be
obtained by <code>t$p.value</code>.</p>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">t</span> <span class="op">&lt;-</span> <span class="fu">fisher.test</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">n_11</span>, <span class="va">n_12</span>, <span class="va">n_21</span>, <span class="va">n_22</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></span>
<span><span class="va">t</span><span class="op">$</span><span class="va">p.value</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.9059e-06</code></pre>
</div>
<p>Odd ratio from the Fisher’s exact test is defined as follows:</p>
<p><span class="math display">\[
\mathrm{Odd\_ratio} = \frac{n_{11}/n_{21}}{n_{12}/n_{22}} =
\frac{n_{11}/n_{12}}{n_{21}/n_{22}} = \frac{n_{11} * n_{22}}{n_{12} *
n_{21}}
\]</span></p>
<p>If there is no association between DE genes and the gene set, odd
ratio is expected to be 1. And it is larger than 1 if there is an
over-representation of DE genes on the gene set.</p>
<div id="further-reading" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="further-reading" class="callout-inner">
<h3 class="callout-title">Further reading<a class="anchor" aria-label="anchor" href="#further-reading"></a>
</h3>
<div class="callout-content">
<p>The 2x2 contigency table can be transposed and it does not affect the
Fisher’s exact test. E.g. let’s put whether genes are in the gene sets
on rows, and put whether genes are DE on columns.</p>
<center>
<table class="table">
<thead><tr class="header">
<th></th>
<th>DE</th>
<th>Not DE</th>
<th>Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>In the gene set</strong></td>
<td>13</td>
<td>1121</td>
<td>1134</td>
</tr>
<tr class="even">
<td><strong>Not in the gene set</strong></td>
<td>41</td>
<td>20023</td>
<td>20064</td>
</tr>
<tr class="odd">
<td><strong>Total</strong></td>
<td>54</td>
<td>21144</td>
<td>21198</td>
</tr>
</tbody>
</table>
</center>
<p>And the corresponding <code>fisher.test()</code> is:</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">fisher.test</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="fl">13</span>, <span class="fl">1121</span>, <span class="fl">41</span>, <span class="fl">20023</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>    alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
	Fisher's Exact Test for Count Data

data:  matrix(c(13, 1121, 41, 20023), nrow = 2, byrow = TRUE)
p-value = 3.906e-06
alternative hypothesis: true odds ratio is greater than 1
95 percent confidence interval:
 3.110607      Inf
sample estimates:
odds ratio 
  5.662486 </code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="the-hypergeometric-distribution">The hypergeometric distribution<a class="anchor" aria-label="anchor" href="#the-hypergeometric-distribution"></a>
</h3>
<p>We can look at the problem from another aspect. This time we treat
all genes as balls in a big box where all the genes have the same
probability to be picked up. Some genes are marked as DE genes (in red
in the figure) and other genes are marked as non-DE genes (in blue). We
grab <span class="math inline">\(n_{+1}\)</span> genes (the size of the
gene set) from the box and we want to ask <strong>what is the
probability of having <span class="math inline">\(n_{11}\)</span> DE
genes in our hand?</strong></p>
<figure><img src="fig/06-gene-set-analysis-rendered-unnamed-chunk-12-1.png" width="80%" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>We first calculate the total number of ways of picking <span class="math inline">\(n_{+1}\)</span> genes from total <span class="math inline">\(n\)</span> genes, without distinguishing whether
they are DE or not: <span class="math inline">\(\binom{n}{n_{+1}}\)</span>.</p>
<p>Next, in the <span class="math inline">\(n_{+1}\)</span> genes that
have been picked, there are <span class="math inline">\(n_{11}\)</span>
DE genes which can only be from the total <span class="math inline">\(n_{1+}\)</span> DE genes. Then the number of ways
of picking <span class="math inline">\(n_{11}\)</span> DE genes from
<span class="math inline">\(n_{1+}\)</span> total DE genes is <span class="math inline">\(\binom{n_{1+}}{n_{11}}\)</span>.</p>
<p>Similarly, there are still <span class="math inline">\(n_{21}\)</span> non-DE genes in our hand, which
can only be from the total <span class="math inline">\(n_{2+}\)</span>
non-DE genes. Then the number of ways of picking <span class="math inline">\(n_{21}\)</span> non-DE genes from <span class="math inline">\(n_{2+}\)</span> total non-DE genes: <span class="math inline">\(\binom{n_{2+}}{n_{21}}\)</span>.</p>
<p>Since picking DE genes and picking non-DE genes are independent, the
number of ways of picking <span class="math inline">\(n_{+1}\)</span>
genes which contain <span class="math inline">\(n_{11}\)</span> DE genes
and <span class="math inline">\(n_{21}\)</span> non-DE genes is their
multiplication: <span class="math inline">\(\binom{n_{1+}}{n_{11}}
\binom{n_{2+}}{n_{21}}\)</span>.</p>
<p>And the probability <span class="math inline">\(P\)</span> is:</p>
<p><span class="math display">\[P = \frac{\binom{n_{1+}}{n_{11}}
\binom{n_{2+}}{n_{21}}}{\binom{n}{n_{+1}}}
=  \frac{\binom{n_{1+}}{n_{11}} \binom{n - n_{1+}}{n_{+1}
-n_{11}}}{\binom{n}{n_{+1}}} \]</span></p>
<p>where in the denominator is the number of ways of picking <span class="math inline">\(n_{+1}\)</span> genes without distinguishing
whether they are DE or not.</p>
<p>If <span class="math inline">\(n\)</span> (number of total genes),
<span class="math inline">\(n_{1+}\)</span> (number of DE genes) and
<span class="math inline">\(n_{+1}\)</span> (size of gene set) are all
fixed values, the number of DE genes that are picked can be denoted as a
random variable <span class="math inline">\(X\)</span>. Then <span class="math inline">\(X\)</span> follows the hypergeometric distribution
with parameters <span class="math inline">\(n\)</span>, <span class="math inline">\(n_{1+}\)</span> and <span class="math inline">\(n_{+1}\)</span>, written as:</p>
<p><span class="math display">\[ X \sim \mathrm{Hyper}(n, n_{1+},
n_{+1})\]</span></p>
<p>The <em>p</em>-value of the enrichment is calculated as the
probability of having an observation equal to or larger than <span class="math inline">\(n_{11}\)</span> (or it can be understood as the
probability of being over-representated):</p>
<p><span class="math display">\[
\mathrm{Pr}( X \geqslant  n_{11} ) = \sum_{x \in \{ {n_{11}, n_{11}+1,
..., \min\{n_{1+}, n_{+1}\} \}}} \mathrm{Pr}(X = x)
\]</span></p>
<p>In R, the function <code>phyper()</code> calculates <em>p</em>-values
from the hypergeometric distribution. There are four arguments:</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">phyper</span><span class="op">(</span><span class="va">q</span>, <span class="va">m</span>, <span class="va">n</span>, <span class="va">k</span><span class="op">)</span></span></code></pre>
</div>
<p>which are:</p>
<ul>
<li>
<code>q</code>: the observation,</li>
<li>
<code>m</code>: number of DE genes,</li>
<li>
<code>n</code>: number of non-DE genes,</li>
<li>
<code>k</code>: size of the gene set.</li>
</ul>
<p><code>phyper()</code> calculates <span class="math inline">\(\mathrm{Pr}(X \leqslant q)\)</span>. To calculate
<span class="math inline">\(\mathrm{Pr}(X \geqslant q)\)</span>, we need
to transform it a little bit:</p>
<p><span class="math display">\[ \mathrm{Pr}(X \geqslant q) = 1 -
\mathrm{Pr}(X &lt; q) = 1 - \mathrm{Pr}(X \leqslant q-1)\]</span></p>
<p>Then, the correct use of <code>phyper()</code> is:</p>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">q</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">m</span>, <span class="va">n</span>, <span class="va">k</span><span class="op">)</span></span></code></pre>
</div>
<p>Let’s pluigin in our variables:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_10</span>, <span class="va">n_20</span>, <span class="va">n_01</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.9059e-06</code></pre>
</div>
<p>Optionally, <code>lower.tail</code> argument can be specified which
directly calculates <em>p</em>-values from the upper tail of the
distribution.</p>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_10</span>, <span class="va">n_20</span>, <span class="va">n_01</span>, lower.tail <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.9059e-06</code></pre>
</div>
<p>If we switch <code>n_01</code> and <code>n_10</code>, the
<em>p</em>-values are identical:</p>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_01</span>, <span class="va">n_02</span>, <span class="va">n_10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 3.9059e-06</code></pre>
</div>
<p><code>fisher.test()</code> and <code>phyper()</code> give the same
<em>p</em>-value. Actually the two methods are identical because in
Fisher’s exact test, hypergeometric distribution is the exact
distribution of its statistic.</p>
<p>Let’s test the runtime of the two functions:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">microbenchmark</span><span class="op">)</span></span>
<span><span class="fu">microbenchmark</span><span class="op">(</span></span>
<span>    fisher <span class="op">=</span> <span class="fu">fisher.test</span><span class="op">(</span><span class="fu">matrix</span><span class="op">(</span><span class="fu">c</span><span class="op">(</span><span class="va">n_11</span>, <span class="va">n_12</span>, <span class="va">n_21</span>, <span class="va">n_22</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, byrow <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,</span>
<span>        alternative <span class="op">=</span> <span class="st">"greater"</span><span class="op">)</span>,</span>
<span>    hyper <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_10</span>, <span class="va">n_20</span>, <span class="va">n_01</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>Unit: microseconds
   expr   min     lq    mean median    uq   max neval
 fisher 282.2 286.45 297.559  288.5 299.5 550.3   100
  hyper   1.7   1.90   2.473    2.3   2.6  18.8   100</code></pre>
</div>
<p>It is very astonishing that <code>phyper()</code> is hundreds of
times faster than <code>fisher.test()</code>. Main reason is in
<code>fisher.test()</code>, there are many additional calculations
besides calculating the <em>p</em>-value. So if you want to implement
ORA analysis by yourself, always consider to use <code>phyper()</code><a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>.</p>
<div id="further-reading-1" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="further-reading-1" class="callout-inner">
<h3 class="callout-title">Further reading<a class="anchor" aria-label="anchor" href="#further-reading-1"></a>
</h3>
<div class="callout-content">
<p>Current tools also use Binomial distribution or chi-square test for
ORA analysis. These two are just approximations. Please refer to <a href="https://doi.org/10.1093/bioinformatics/btl633" class="external-link">Rivals et al.,
Enrichment or depletion of a GO category within a class of genes: which
test? Bioinformatics 2007</a> which gives an overview of statistical
methods used in ORA analysis.</p>
</div>
</div>
</div>
</div>
</section><section id="gene-set-resources"><h2 class="section-heading">Gene set resources<a class="anchor" aria-label="anchor" href="#gene-set-resources"></a>
</h2>
<hr class="half-width">
<p>We have learnt the basic methods of ORA analysis. Now we go to the
second component of the analysis: the gene sets.</p>
<p>Gene sets represent prior knowledge of what is the general shared
biological attribute of genes in the gene set. For example, in a “cell
cycle” gene set, all the genes are involved in the cell cycle process.
Thus, if DE genes are significantly enriched in the “cell cycle” gene
set, which means there are significantly more cell cycle genes
differentially expressed than expected, we can conclude that the normal
function of cell cycle process may be affected.</p>
<p>As we have mentioned, genes in the gene set share the same
“biological attribute” where “the attribute” will be used for making
conclusions. The definition of “biological attribute” is very flexible.
It can be a biological process such as “cell cycle”. It can also be from
a wide range of other definitions, to name a few:</p>
<ul>
<li>Locations in the cell, e.g. cell membrane or cell nucleus.</li>
<li>Positions on chromosomes, e.g. sex chromosomes or the cytogenetic
band p13 on chromome 10.</li>
<li>Target genes of a transcription factor or a microRNA, e.g. all genes
that are transcriptionally regulationed by NF-κB.</li>
<li>Signature genes in a certain tumor type, i.e. genes that are
uniquely highly expressed in a tumor type.</li>
</ul>
<p>The <a href="https://www.gsea-msigdb.org/gsea/msigdb" class="external-link">MSigDB
database</a> contains gene sets in many topics. We will introduce it
later in this section.</p>
<p>You may have encountered many different ways to name gene sets: “gene
sets”, “biological terms”, “GO terms”, “GO gene sets”, “pathways”, and
so on. They basically refer to the same thing, but from different
aspects. As shown in the following figure, “gene set” corresponds to a
vector of genes and it is the representation of the data for
computation. “Biological term” is a textual entity that contains
description of its biological meaning; It corresponds to the knowledge
of the gene set and is for the inference of the analysis. “GO gene sets”
and “pathways” specifically refer to the enrichment analysis using GO
gene sets and pahtway gene sets.</p>
<figure><img src="fig/geneset.svg" class="figure mx-auto d-block"></figure><p>Before we touch the gene set databases, we first summarize the
general formats of gene sets in R. In most analysis, a gene set is
simply treated as a vector of genes. Thus, naturally, a collection of
gene sets can be represented as a list of vectors. In the following
example, there are three gene sets with 3, 5 and 2 genes. Some genes
exist in multiple gene sets.</p>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lt</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>gene_set_1 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_1"</span>, <span class="st">"gene_2"</span>, <span class="st">"gene_3"</span><span class="op">)</span>,</span>
<span>     gene_set_2 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_1"</span>, <span class="st">"gene_3"</span>, <span class="st">"gene_4"</span>, <span class="st">"gene_5"</span>, <span class="st">"gene_6"</span><span class="op">)</span>,</span>
<span>     gene_set_3 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_4"</span>, <span class="st">"gene_7"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">lt</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$gene_set_1
[1] "gene_1" "gene_2" "gene_3"

$gene_set_2
[1] "gene_1" "gene_3" "gene_4" "gene_5" "gene_6"

$gene_set_3
[1] "gene_4" "gene_7"</code></pre>
</div>
<p>It is also very common to store the relations of gene sets and genes
as a two-column data frame. The order of the gene set column and the
gene column, i.e. which column locates as the first column, are quite
arbitrary. Different tools may require differently.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">data.frame</span><span class="op">(</span>gene_set <span class="op">=</span> <span class="fu">rep</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">lt</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">lt</span>, <span class="va">length</span><span class="op">)</span><span class="op">)</span>,</span>
<span>           gene <span class="op">=</span> <span class="fu">unname</span><span class="op">(</span><span class="fu">unlist</span><span class="op">(</span><span class="va">lt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     gene_set   gene
1  gene_set_1 gene_1
2  gene_set_1 gene_2
3  gene_set_1 gene_3
4  gene_set_2 gene_1
5  gene_set_2 gene_3
6  gene_set_2 gene_4
7  gene_set_2 gene_5
8  gene_set_2 gene_6
9  gene_set_3 gene_4
10 gene_set_3 gene_7</code></pre>
</div>
<p>Or genes be in the first column:</p>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     gene   gene_set
1  gene_1 gene_set_1
2  gene_2 gene_set_1
3  gene_3 gene_set_1
4  gene_1 gene_set_2
5  gene_3 gene_set_2
6  gene_4 gene_set_2
7  gene_5 gene_set_2
8  gene_6 gene_set_2
9  gene_4 gene_set_3
10 gene_7 gene_set_3</code></pre>
</div>
<p>Not very often, gene sets are represented as a matrix where one
dimension corresponds to gene sets and the other dimension corresponds
to genes. The values in the matix are binary where a value of 1
represents the gene is a member of the corresponding gene sets. In some
methods, 1 is replaced by <span class="math inline">\(w_{ij}\)</span> to
weight the effect of the genes in the gene set.</p>
<pre><code><span><span class="co">#            gene_1 gene_2 gene_3 gene_4</span></span>
<span><span class="co"># gene_set_1      1      1      0      0</span></span>
<span><span class="co"># gene_set_2      1      0      1      1</span></span></code></pre>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>Can you convert between different gene set representations? E.g.
convert a list to a two-column data frame?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lt</span> <span class="op">&lt;-</span> <span class="fu">list</span><span class="op">(</span>gene_set_1 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_1"</span>, <span class="st">"gene_2"</span>, <span class="st">"gene_3"</span><span class="op">)</span>,</span>
<span>     gene_set_2 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_1"</span>, <span class="st">"gene_3"</span>, <span class="st">"gene_4"</span>, <span class="st">"gene_5"</span>, <span class="st">"gene_6"</span><span class="op">)</span>,</span>
<span>     gene_set_3 <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene_4"</span>, <span class="st">"gene_7"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre>
</div>
<p>To convert <code>lt</code> to a data frame (e.g. let’s put gene sets
in the first column):</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">df</span> <span class="op">=</span> <span class="fu">data.frame</span><span class="op">(</span>gene_set <span class="op">=</span> <span class="fu">rep</span><span class="op">(</span><span class="fu">names</span><span class="op">(</span><span class="va">lt</span><span class="op">)</span>, times <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">lt</span>, <span class="va">length</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                gene <span class="op">=</span> <span class="fu">unname</span><span class="op">(</span><span class="fu">unlist</span><span class="op">(</span><span class="va">lt</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">df</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     gene_set   gene
1  gene_set_1 gene_1
2  gene_set_1 gene_2
3  gene_set_1 gene_3
4  gene_set_2 gene_1
5  gene_set_2 gene_3
6  gene_set_2 gene_4
7  gene_set_2 gene_5
8  gene_set_2 gene_6
9  gene_set_3 gene_4
10 gene_set_3 gene_7</code></pre>
</div>
<p>To convert <code>df</code> back to the list:</p>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">split</span><span class="op">(</span><span class="va">df</span><span class="op">$</span><span class="va">gene</span>, <span class="va">df</span><span class="op">$</span><span class="va">gene_set</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$gene_set_1
[1] "gene_1" "gene_2" "gene_3"

$gene_set_2
[1] "gene_1" "gene_3" "gene_4" "gene_5" "gene_6"

$gene_set_3
[1] "gene_4" "gene_7"</code></pre>
</div>
</div>
</div>
</div>
</div>
<p>Next, let’s go through gene sets from several major databases: the
GO, KEGG and MSigDB databases.</p>
<div class="section level3">
<h3 id="gene-ontology-gene-sets">Gene Ontology gene sets<a class="anchor" aria-label="anchor" href="#gene-ontology-gene-sets"></a>
</h3>
<p>Gene Ontology (GO) is the standard source for gene set enrichment
analysis. GO contains three namespaces of biological process (BP),
cellular components (CC) and molecular function (MF) which describe a
biological entity from different aspect. The associations between GO
terms and genes are integrated in the Bioconductor standard packages:
<em>the oragnism annotation packages</em>. In current Bioconductor
release 3.17, there are the following organism packages:</p>
<center>
<table class="table">
<colgroup>
<col width="27%">
<col width="25%">
<col width="36%">
<col width="10%">
</colgroup>
<thead><tr class="header">
<th>Package</th>
<th>Organism</th>
<th>Package</th>
<th>Organism</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>org.Hs.eg.db</td>
<td>Human</td>
<td>org.Mm.eg.db</td>
<td>Mouse</td>
</tr>
<tr class="even">
<td>org.Rn.eg.db</td>
<td>Rat</td>
<td>org.Dm.eg.db</td>
<td>Fly</td>
</tr>
<tr class="odd">
<td>org.At.tair.db</td>
<td>Arabidopsis</td>
<td>org.Dr.eg.db</td>
<td>Zebrafish</td>
</tr>
<tr class="even">
<td>org.Sc.sgd.db</td>
<td>Yeast</td>
<td>org.Ce.eg.db</td>
<td>Worm</td>
</tr>
<tr class="odd">
<td>org.Bt.eg.db</td>
<td>Bovine</td>
<td>org.Gg.eg.db</td>
<td>Chicken</td>
</tr>
<tr class="even">
<td>org.Ss.eg.db</td>
<td>Pig</td>
<td>org.Mmu.eg.db</td>
<td>Rhesus</td>
</tr>
<tr class="odd">
<td>org.Cf.eg.db</td>
<td>Canine</td>
<td>org.EcK12.eg.db</td>
<td>E coli strain K12</td>
</tr>
<tr class="even">
<td>org.Xl.eg.db</td>
<td>Xenopus</td>
<td>org.Pt.eg.db</td>
<td>Chimp</td>
</tr>
<tr class="odd">
<td>org.Ag.eg.db</td>
<td>Anopheles</td>
<td>org.EcSakai.eg.db</td>
<td>E coli strain Sakai</td>
</tr>
</tbody>
</table>
</center>
<p>There are four sections in the name of an organism package. The
naming convention is: <code>org</code> simply means “organism”. The
second section corresponds to a specific organism, e.g. <code>Hs</code>
for human and <code>Mm</code> for mouse. The third section corresponds
to the primary gene ID type used in the package, where normally
<code>eg</code> is used which means “Entrez genes” because data is
mostly retrieved from the NCBI database. However, for some organisms,
the primary ID can be from its own primary database,
e.g. <code>sgd</code> for Yeast which corresponds to the <a href="https://www.yeastgenome.org/" class="external-link">Saccharomyces Genome Database</a>,
the primary database for yeast. The last section is always “db”, which
simply implies it is a database package.</p>
<p>Taking the <strong>org.Hs.eg.db</strong> package as an example, all
the data is stored in a database object <code>org.Hs.eg.db</code> in the
<code>OrgDb</code> class. The object contains a connection to a local
SQLite database. Users can simply think <code>org.Hs.eg.db</code> as a
huge table that contains ID mappings between various databases. GO gene
sets are essentially mappings between GO terms and genes. Let’s try to
extract it from the <code>org.Hs.eg.db</code> object.</p>
<p>All the columns (the key column or the source column) can be obtained
by <code>keytypes()</code>:</p>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span>
<span><span class="fu">keytypes</span><span class="op">(</span><span class="va">org.Hs.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code> [1] "ACCNUM"       "ALIAS"        "ENSEMBL"      "ENSEMBLPROT"  "ENSEMBLTRANS"
 [6] "ENTREZID"     "ENZYME"       "EVIDENCE"     "EVIDENCEALL"  "GENENAME"    
[11] "GENETYPE"     "GO"           "GOALL"        "IPI"          "MAP"         
[16] "OMIM"         "ONTOLOGY"     "ONTOLOGYALL"  "PATH"         "PFAM"        
[21] "PMID"         "PROSITE"      "REFSEQ"       "SYMBOL"       "UCSCKG"      
[26] "UNIPROT"     </code></pre>
</div>
<p>To get the GO gene sets, we first obtain all GO IDs under the BP
(biological process) namespace. As shown in the output from
<code>keytype()</code>, <code>"ONTOLOGY"</code> is also a valid “key
column”, thus we can query “<em>select all GO IDs where the
corresponding ONTOLOGY is BP</em>”, which is translated into the
following code:</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BP_Id</span> <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keys <span class="op">=</span> <span class="st">"BP"</span>, keytype <span class="op">=</span> <span class="st">"ONTOLOGY"</span>, </span>
<span>    column <span class="op">=</span> <span class="st">"GO"</span>, multiVals <span class="op">=</span> <span class="st">"list"</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">BP_Id</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "GO:0008150" "GO:0001553" "GO:0001869" "GO:0002438" "GO:0006953"
[6] "GO:0007584"</code></pre>
</div>
<p><code>mapIds()</code> maps IDs between two sources. Since a GO
namespace have more than one GO terms, we have to set
<code>multiVals = "list"</code> to obtain all GO terms under that
namespace. And since we only query for one GO “ONTOLOGY”, we directly
take the first element from the list returned by
<code>mapIds()</code>.</p>
<p>Next we do mapping from GO IDs to gene Entrez IDs. Now the query
becomes “<em>providing a vector of GO IDs, select ENTREZIDs which
correspond to every one of them</em>”.</p>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BPGeneSets</span> <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Hs.eg.db</span>, keys <span class="op">=</span> <span class="va">BP_Id</span>, keytype <span class="op">=</span> <span class="st">"GOALL"</span>, </span>
<span>    column <span class="op">=</span> <span class="st">"ENTREZID"</span>, multiVals <span class="op">=</span> <span class="st">"list"</span><span class="op">)</span></span></code></pre>
</div>
<p>You may have noticed there is a “GO” key column as well a “GOALL”
column in the database. As GO has a hierarchical structure where a child
term is a sub-class of a parent term. All the genes annotated to a child
term are also annotated to its parent terms. To reduce the duplicated
information when annotating genes to GO terms, genes are normally
annotated to the most specific offspring terms in the GO hierarchy.
Upstream merging of gene annotations should be done by the tools which
perform analysis. In this way, the mapping between <code>"GO"</code> and
<code>"ENTREZID"</code> only contains “primary” annotations which is not
complete, and mapping between <code>"GOALL"</code> and
<code>"ENTREZID"</code> is the correct one to use.</p>
<p>We filter out GO gene sets with no gene annotated.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">BPGeneSets</span> <span class="op">=</span> <span class="va">BPGeneSets</span><span class="op">[</span><span class="fu">sapply</span><span class="op">(</span><span class="va">BPGeneSets</span>, <span class="va">length</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">]</span></span>
<span><span class="va">BPGeneSets</span><span class="op">[</span><span class="fl">2</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span> <span class="co"># BPGeneSets[[1]] is too long</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$`GO:0001553`
 [1] "2"     "2516"  "2661"  "2661"  "3624"  "4313"  "5156"  "5798"  "6777" 
[10] "8322"  "8879"  "56729" "59338"

$`GO:0001869`
[1] "2"   "710"</code></pre>
</div>
<p>In most cases, because <code>OrgDb</code> is a standard Bioconductor
data structure, most tools can automatically construct GO gene sets
internally. There is no need for users to touch such low-level
processings.</p>
<div id="further-reading-2" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="further-reading-2" class="callout-inner">
<h3 class="callout-title">Further reading<a class="anchor" aria-label="anchor" href="#further-reading-2"></a>
</h3>
<div class="callout-content">
<p>Mapping between various databases can also be done with the general
<code>select()</code> interface. If the <code>OrgDb</code> object is
provided by a package such as <strong>org.Hs.eg.db</strong>, there is
also a separated object that specifically contains mapping between GO
terms and genes. Readers can check the documentation of
<code>org.Hs.egGO2ALLEGS</code>. Additional information on GO terms such
as GO names and long descriptions are available in the package
<strong>GO.db</strong>.</p>
<p>Bioconductor has already provided a large number of organism
packages. However, if the organism you are working on is not supported
there, you may consider to look for it with the
<strong>AnnotationHub</strong> package, which additionally provide
<code>OrgDb</code> objects for approximately 2000 organisms. The
<code>OrgDb</code> object can be directly used in the ORA analysis
introduced in the next section.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="kegg-gene-sets">KEGG gene sets<a class="anchor" aria-label="anchor" href="#kegg-gene-sets"></a>
</h3>
<p>A biological pathway is a series of interactions among molecules in a
cell that leads to a certain product or a change in a cell<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>. A pathway involes a
list of genes playing different roles which constructs the “pathway gene
set”. <a href="https://www.genome.jp/kegg/pathway.html" class="external-link">KEGG pathway</a>
is the mostly used database for pathways. It provides its data via a
REST API (<a href="https://rest.kegg.jp/" class="external-link uri">https://rest.kegg.jp/</a>). There are several commands to
retrieve specific types of data. To retrieve the pathway gene sets, we
can use the “link” command as shown in the following URL (“link” means
to link genes to pathways). When you enter the URL in the web
brower:</p>
<pre><code>https://rest.kegg.jp/link/pathway/hsa</code></pre>
<p>there will be a text table which contains a column of genes and a
column of pathway IDs.</p>
<pre><code>hsa:10327   path:hsa00010
hsa:124 path:hsa00010
hsa:125 path:hsa00010
hsa:126 path:hsa00010</code></pre>
<p>We can directly read the text output with <code>read.table()</code>.
Wrapping the URL with the function <code>url()</code>, you can pretend
to directly read data from the remote web server.</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keggGeneSets</span> <span class="op">=</span> <span class="fu">read.table</span><span class="op">(</span><span class="fu">url</span><span class="op">(</span><span class="st">"https://rest.kegg.jp/link/pathway/hsa"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">keggGeneSets</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>         V1            V2
1 hsa:10327 path:hsa00010
2   hsa:124 path:hsa00010
3   hsa:125 path:hsa00010
4   hsa:126 path:hsa00010
5   hsa:127 path:hsa00010
6   hsa:128 path:hsa00010</code></pre>
</div>
<p>In this two-column table, the first column contains genes in the
Entrez ID type. Let’s remove the <code>"hsa:"</code> prefix, also we
remove the <code>"path:"</code> prefix for pathway IDs in the second
column.</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keggGeneSets</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span> <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"hsa:"</span>, <span class="st">""</span>, <span class="va">keggGeneSets</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">keggGeneSets</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span> <span class="op">=</span> <span class="fu">gsub</span><span class="op">(</span><span class="st">"path:"</span>, <span class="st">""</span>, <span class="va">keggGeneSets</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">keggGeneSets</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>     V1       V2
1 10327 hsa00010
2   124 hsa00010
3   125 hsa00010
4   126 hsa00010
5   127 hsa00010
6   128 hsa00010</code></pre>
</div>
<p>The full pathway names can be obtained via the “list” command.</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keggNames</span> <span class="op">=</span> <span class="fu">read.table</span><span class="op">(</span><span class="fu">url</span><span class="op">(</span><span class="st">"https://rest.kegg.jp/list/pathway/hsa"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">keggNames</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>        V1                                                     V2
1 hsa01100              Metabolic pathways - Homo sapiens (human)
2 hsa01200               Carbon metabolism - Homo sapiens (human)
3 hsa01210 2-Oxocarboxylic acid metabolism - Homo sapiens (human)
4 hsa01212           Fatty acid metabolism - Homo sapiens (human)
5 hsa01230     Biosynthesis of amino acids - Homo sapiens (human)
6 hsa01232           Nucleotide metabolism - Homo sapiens (human)</code></pre>
</div>
<p>In both commands, we obtained data for human where the corresponding
KEGG code is <code>"hsa"</code>. The code for other organisms can be
found from the <a href="https://www.genome.jp/kegg/" class="external-link">KEGG website</a>
(e.g. <code>"mmu"</code> for mouse), or via <a href="https://rest.kegg.jp/list/organism" class="external-link uri">https://rest.kegg.jp/list/organism</a>.</p>
<p>Keep in mind, KEGG pathways are only free for academic users. If you
use it for commercial purposes, <a href="https://www.kegg.jp/kegg/legal.html" class="external-link">please contact the KEGG team
to get a licence</a>.</p>
<div id="further-reading-3" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="further-reading-3" class="callout-inner">
<h3 class="callout-title">Further reading<a class="anchor" aria-label="anchor" href="#further-reading-3"></a>
</h3>
<div class="callout-content">
<p>Instead directly reading from the URLs, there are also R packages
which help to obtain data from the KEGG database, such as the
<strong>KEGGREST</strong> package or the <code>download_KEGG()</code>
function from the <strong>clusterProfiler</strong> package. But
essentially, they all obtain KEGG data with the REST API.</p>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="msigdb-gene-sets">MSigDB gene sets<a class="anchor" aria-label="anchor" href="#msigdb-gene-sets"></a>
</h3>
<p><a href="https://www.gsea-msigdb.org/gsea/msigdb/" class="external-link">Molecular
signature database</a> (MSigDB) is a manually curated gene set database.
Initially, it was proposed as a supplementary dataset for <a href="https://www.nature.com/articles/ng1180" class="external-link">the original GSEA
paper</a>. Later it has been separated out and developed independently.
In the first version in 2005, there were only two gene sets collections
and in total 843 gene sets. Now in the newest version of MSigDB
(v2023.1.Hs), it has grown into nine gene sets collections, covering
&gt; 30K gene sets. It provides gene sets on a variety of topics.</p>
<p>MSigDB categorizes gene sets into nine collections where each
collection focuses on a specific topic. For some collections, they are
additionally split into sub-collections. There are several ways to
obtain gene sets from MSigDB. One convenient way is to use the
<strong>msigdbr</strong> package. The advantages is
<strong>msigdbr</strong> supports many other organisms by mapping to
orthologs.</p>
<p>Let’s check which organisms are supported and which gene sets
collections it provides.</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">msigdbr</span><span class="op">)</span></span>
<span><span class="fu">msigdbr_species</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 20 × 2
   species_name                    species_common_name                          
   &lt;chr&gt;                           &lt;chr&gt;                                        
 1 Anolis carolinensis             Carolina anole, green anole                  
 2 Bos taurus                      bovine, cattle, cow, dairy cow, domestic cat…
 3 Caenorhabditis elegans          &lt;NA&gt;                                         
 4 Canis lupus familiaris          dog, dogs                                    
 5 Danio rerio                     leopard danio, zebra danio, zebra fish, zebr…
 6 Drosophila melanogaster         fruit fly                                    
 7 Equus caballus                  domestic horse, equine, horse                
 8 Felis catus                     cat, cats, domestic cat                      
 9 Gallus gallus                   bantam, chicken, chickens, Gallus domesticus 
10 Homo sapiens                    human                                        
11 Macaca mulatta                  rhesus macaque, rhesus macaques, Rhesus monk…
12 Monodelphis domestica           gray short-tailed opossum                    
13 Mus musculus                    house mouse, mouse                           
14 Ornithorhynchus anatinus        duck-billed platypus, duckbill platypus, pla…
15 Pan troglodytes                 chimpanzee                                   
16 Rattus norvegicus               brown rat, Norway rat, rat, rats             
17 Saccharomyces cerevisiae        baker's yeast, brewer's yeast, S. cerevisiae 
18 Schizosaccharomyces pombe 972h- &lt;NA&gt;                                         
19 Sus scrofa                      pig, pigs, swine, wild boar                  
20 Xenopus tropicalis              tropical clawed frog, western clawed frog    </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">msigdbr_collections</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 23 × 3
   gs_cat gs_subcat         num_genesets
   &lt;chr&gt;  &lt;chr&gt;                    &lt;int&gt;
 1 C1     ""                         299
 2 C2     "CGP"                     3384
 3 C2     "CP"                        29
 4 C2     "CP:BIOCARTA"              292
 5 C2     "CP:KEGG"                  186
 6 C2     "CP:PID"                   196
 7 C2     "CP:REACTOME"             1615
 8 C2     "CP:WIKIPATHWAYS"          664
 9 C3     "MIR:MIRDB"               2377
10 C3     "MIR:MIR_Legacy"           221
11 C3     "TFT:GTRD"                 518
12 C3     "TFT:TFT_Legacy"           610
13 C4     "CGN"                      427
14 C4     "CM"                       431
15 C5     "GO:BP"                   7658
16 C5     "GO:CC"                   1006
17 C5     "GO:MF"                   1738
18 C5     "HPO"                     5071
19 C6     ""                         189
20 C7     "IMMUNESIGDB"             4872
21 C7     "VAX"                      347
22 C8     ""                         700
23 H      ""                          50</code></pre>
</div>
<p>The first column in the above output is the primary category of gene
sets. Some gene sets collections may have sub-collections, and they are
shown in the second column. The description of gene sets collections
from MSigDB is in the following figure.</p>
<figure><img src="fig/msigdb.png" class="figure mx-auto d-block"></figure><p>The core function <code>msigdbr()</code> retrieves gene sets in a
specific category (or a subcategory if it exists).</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">msigdbr</span><span class="op">(</span><span class="va">species</span>, <span class="va">category</span>, <span class="va">subcategory</span><span class="op">)</span></span></code></pre>
</div>
<p>For example, we want to obtain the hallmark gene sets for mouse.</p>
<div class="codewrapper sourceCode" id="cb62">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MSigDBGeneSets</span> <span class="op">=</span> <span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"mouse"</span>, category <span class="op">=</span> <span class="st">"H"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">MSigDBGeneSets</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 18
  gs_cat gs_subcat gs_name               gene_symbol entrez_gene ensembl_gene   
  &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;                 &lt;chr&gt;             &lt;int&gt; &lt;chr&gt;          
1 H      ""        HALLMARK_ADIPOGENESIS Abca1             11303 ENSMUSG0000001…
2 H      ""        HALLMARK_ADIPOGENESIS Abcb8             74610 ENSMUSG0000002…
3 H      ""        HALLMARK_ADIPOGENESIS Acaa2             52538 ENSMUSG0000003…
4 H      ""        HALLMARK_ADIPOGENESIS Acadl             11363 ENSMUSG0000002…
5 H      ""        HALLMARK_ADIPOGENESIS Acadm             11364 ENSMUSG0000006…
6 H      ""        HALLMARK_ADIPOGENESIS Acads             11409 ENSMUSG0000002…
# ℹ 12 more variables: human_gene_symbol &lt;chr&gt;, human_entrez_gene &lt;int&gt;,
#   human_ensembl_gene &lt;chr&gt;, gs_id &lt;chr&gt;, gs_pmid &lt;chr&gt;, gs_geoid &lt;chr&gt;,
#   gs_exact_source &lt;chr&gt;, gs_url &lt;chr&gt;, gs_description &lt;chr&gt;, taxon_id &lt;int&gt;,
#   ortholog_sources &lt;chr&gt;, num_ortholog_sources &lt;dbl&gt;</code></pre>
</div>
<p>The output is in the <code>tibble</code> class. If you have no
experience with it, don’t worry, just take it as a table. As you can
see, The three major gene IDs type <code>"gene_symbol"</code>,
<code>"entrez_gene"</code> and <code>"ensembl_gene"</code> are all
included in the table. So users can easily pick one with the same gene
ID type as in the DE gene list. For example:</p>
<div class="codewrapper sourceCode" id="cb64">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">MSigDBGeneSets</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"gs_name"</span>, <span class="st">"ensembl_gene"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 7,384 × 2
   gs_name               ensembl_gene      
   &lt;chr&gt;                 &lt;chr&gt;             
 1 HALLMARK_ADIPOGENESIS ENSMUSG00000015243
 2 HALLMARK_ADIPOGENESIS ENSMUSG00000028973
 3 HALLMARK_ADIPOGENESIS ENSMUSG00000036880
 4 HALLMARK_ADIPOGENESIS ENSMUSG00000026003
 5 HALLMARK_ADIPOGENESIS ENSMUSG00000062908
 6 HALLMARK_ADIPOGENESIS ENSMUSG00000029545
 7 HALLMARK_ADIPOGENESIS ENSMUSG00000020917
 8 HALLMARK_ADIPOGENESIS ENSMUSG00000022477
 9 HALLMARK_ADIPOGENESIS ENSMUSG00000020777
10 HALLMARK_ADIPOGENESIS ENSMUSG00000022994
# ℹ 7,374 more rows</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb66">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># or put genes in the first column</span></span>
<span><span class="va">MSigDBGeneSets</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"ensembl_gene"</span>, <span class="st">"gs_name"</span><span class="op">)</span><span class="op">]</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 7,384 × 2
   ensembl_gene       gs_name              
   &lt;chr&gt;              &lt;chr&gt;                
 1 ENSMUSG00000015243 HALLMARK_ADIPOGENESIS
 2 ENSMUSG00000028973 HALLMARK_ADIPOGENESIS
 3 ENSMUSG00000036880 HALLMARK_ADIPOGENESIS
 4 ENSMUSG00000026003 HALLMARK_ADIPOGENESIS
 5 ENSMUSG00000062908 HALLMARK_ADIPOGENESIS
 6 ENSMUSG00000029545 HALLMARK_ADIPOGENESIS
 7 ENSMUSG00000020917 HALLMARK_ADIPOGENESIS
 8 ENSMUSG00000022477 HALLMARK_ADIPOGENESIS
 9 ENSMUSG00000020777 HALLMARK_ADIPOGENESIS
10 ENSMUSG00000022994 HALLMARK_ADIPOGENESIS
# ℹ 7,374 more rows</code></pre>
</div>
<p>If you only want to use a sub-category gene sets,
e.g. <code>"CP:KEGG"</code> from <code>"C2"</code>, you can simply
specify both <code>category</code> and <code>subcategory</code>
arguments:</p>
<div class="codewrapper sourceCode" id="cb68">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">msigdbr</span><span class="op">(</span>species <span class="op">=</span> <span class="st">"mouse"</span>, category <span class="op">=</span> <span class="st">"C2"</span>, subcategory <span class="op">=</span> <span class="st">"CP:KEGG"</span><span class="op">)</span></span></code></pre>
</div>
</div>
</section><section id="ora-with-clusterprofiler"><h2 class="section-heading">ORA with clusterProfiler<a class="anchor" aria-label="anchor" href="#ora-with-clusterprofiler"></a>
</h2>
<hr class="half-width">
<p>The ORA method itself is quite simple and it has been implemented in
a large number of R packages. Among them, the
<strong>clusterProfiler</strong> package especially does a good job that
it has a seamless integration to the Bioconductor annotation resources
which allows extending GSEA analysis to other organisms easily; it has
pre-defined functions for common analysis tasks, e.g. GO enrichment,
KEGG enrichment; and it implements a variaty of different visualization
methods on the GSEA results. In this section, we will learn how to
perform ORA anlaysis with <strong>clusterProfiler</strong>.</p>
<p>Here we will use a list of DE genes from a different comparison. As
you may still remember, there are only 54 DE genes between genders,
which may not be a good case for GSEA analysis, since after overlapping
to a collection fo gene sets, majority of the gene sets will have few or
even no gene overlapped. In this example we use the list of DE genes
from the comparison between different time points.</p>
<p>Another thing worth to mention is, in the following code where we
filter DE genes, we additionally add a filtering on the log2 fold
change. This is recommended when the number of DE genes is too large.
The filtering on log2 fold change can be thought as a filtering from the
biology aspect.</p>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTime</span> <span class="op">&lt;-</span> <span class="fu">DESeq2</span><span class="fu">::</span><span class="fu">results</span><span class="op">(</span><span class="va">dds</span>, contrast <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"time"</span>, <span class="st">"Day8"</span>, <span class="st">"Day0"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">timeDE</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resTime</span>, </span>
<span>                               <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="fu">abs</span><span class="op">(</span><span class="va">log2FoldChange</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span></span>
<span>                       <span class="op">)</span><span class="op">)</span></span>
<span><span class="va">timeDEgenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">timeDE</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">timeDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] "3110035E14Rik" "Sgk3"          "Kcnb2"         "Sbspon"       
[5] "Gsta3"         "Lman2l"       </code></pre>
</div>
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">length</span><span class="op">(</span><span class="va">timeDEgenes</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[1] 1134</code></pre>
</div>
<p>Let’s confirm that there are around one thousand DE genes, and the DE
genes are in gene symbols.</p>
<div class="section level3">
<h3 id="go-enrichment">GO enrichment<a class="anchor" aria-label="anchor" href="#go-enrichment"></a>
</h3>
<p>In <strong>clusterProfiler</strong>, there is an
<code>enrichGO()</code> function which performs ORA on GO gene sets. To
use it, we need to provide the DE genes, the organism <code>OrgDb</code>
object which is from the organism package <strong>org.Mm.eg.db</strong>
(because our data is from mouse), also the GO namespace (one of
<code>"BP"</code>, <code>"CC"</code> and <code>"MF"</code>). The GO gene
sets are automatically retrieved and processed from
<code>org.Mm.eg.db</code> in <code>enrichGO()</code>.</p>
<div class="codewrapper sourceCode" id="cb73">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">clusterProfiler</span><span class="op">)</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="va">resTimeGO</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                     ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                     OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>--&gt; No gene can be mapped....</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>--&gt; Expected input gene ID: 16449,23967,635253,72938,15375,12416</code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>--&gt; return NULL...</code></pre>
</div>
<p>Oops, something seems wrong. Well, this is a common mistake where the
gene ID types do not match between DE genes and the gene sets.
Thankfully, the message clearly explains the reason. The ID type for
gene sets is Entrez ID and it cannot match any DE gene.</p>
<p>There are two ways to solve this problem. 1. Convert gene IDs in
<code>timeDEgenes</code> to Entrez IDs in advance; or 2. Simply specify
the ID type of DE genes and let <code>enrichGO()</code> do the
conversion job (recall various gene ID types are also stored in the
<code>OrgDb</code> object). Let’s choose the second way.</p>
<p>In the next code, we additionally specify
<code>keyType = "SYMBOL"</code> to explicitly tell the function that DE
genes are in gene symbols. Recall that all valid values for
<code>keyType</code> are in <code>keytypes(org.Mm.eg.db)</code>.</p>
<div class="codewrapper sourceCode" id="cb77">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeGO</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                     keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                     ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                     OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span><span class="op">)</span></span>
<span><span class="va">resTimeGOTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                   ID                 Description GeneRatio   BgRatio
GO:0050900 GO:0050900         leukocyte migration    49/969 386/28564
GO:0030595 GO:0030595        leukocyte chemotaxis    35/969 230/28564
GO:0071674 GO:0071674  mononuclear cell migration    30/969 185/28564
GO:0060326 GO:0060326             cell chemotaxis    39/969 308/28564
GO:0035456 GO:0035456 response to interferon-beta    17/969  71/28564
GO:0097529 GO:0097529 myeloid leukocyte migration    31/969 242/28564
                 pvalue     p.adjust       qvalue
GO:0050900 2.163956e-15 1.096260e-11 8.402981e-12
GO:0030595 1.066463e-13 2.701351e-10 2.070622e-10
GO:0071674 1.143223e-12 1.930522e-09 1.479771e-09
GO:0060326 1.728979e-12 2.189752e-09 1.678475e-09
GO:0035456 1.681085e-10 1.703276e-07 1.305584e-07
GO:0097529 2.447325e-10 2.066358e-07 1.583892e-07
                                                                                                                                                                                                                                                                                       geneID
GO:0050900 Tnfsf18/Sell/Slamf9/Fut7/Itga4/Mdk/Grem1/Ada/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Enpp1/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Emp2/Apod/Retnlg/Plg/Dusp1/Ager/Il33/Ch25h
GO:0030595                                                                           Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Lgals3/Ptk2b/Retnlg/Dusp1/Ch25h
GO:0071674                                                                                                         Tnfsf18/Slamf9/Fut7/Itga4/Mdk/Grem1/Il12a/Nbl1/Padi2/Alox5/Trpm4/Hsd3b7/Adam8/Calr/Ccl17/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Lgals3/Ptk2b/Apod/Retnlg/Plg/Dusp1/Ager/Ch25h
GO:0060326                                                  Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Lpar1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Ccl28/Lgals3/Ptk2b/Nr4a1/Retnlg/Dusp1/Ch25h/Plxnb3
GO:0035456                                                                                                                                                                      Aim2/Ifi204/Gbp6/Oas1c/Ifitm6/Bst2/Irgm1/Tgtp1/Tgtp2/Ifi47/Igtp/Irgm2/Ifitm7/Gm4951/F830016B08Rik/Iigp1/Ifit1
GO:0097529                                                                                                    Tnfsf18/Sell/Fut7/Mdk/Grem1/Prex1/Edn3/P2ry12/S100a8/S100a9/Nbl1/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Itgam/Adam8/Ccl17/Enpp1/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Lgals3/Ptk2b/Emp2/Retnlg/Dusp1/Ager
           Count
GO:0050900    49
GO:0030595    35
GO:0071674    30
GO:0060326    39
GO:0035456    17
GO:0097529    31</code></pre>
</div>
<p>Now <code>enrichGO()</code> went through! The returned object
<code>resTimeGO</code> is in a special format which looks like a table
but actually not! To get rid of the confusion, in the code it is
converted to a real data frame <code>resTimeGOTable</code>.</p>
<p>In the output data frame, there are the following columns:</p>
<ul>
<li>
<code>ID</code>: ID of the gene set. In this example analysis, it is
the GO ID.</li>
<li>
<code>Description</code>: Readable description. Here is the name of
the GO term.</li>
<li>
<code>GeneRatio</code>: Number of DE genes in the gene set / total
number of DE genes</li>
<li>
<code>BgRatio</code>: size of the gene set / total number of
genes.</li>
<li>
<code>pvalue</code>: <em>p</em>-value calculated from the
hypergeometric distribution.</li>
<li>
<code>p.adjust</code>: Adjusted <em>p</em>-value by the BH
method.</li>
<li>
<code>qvalue</code>: <em>q</em>-value which is another way for
controlling false positives in multiple testings.</li>
<li>
<code>geneID</code>: A list of DE genes in the gene set.</li>
<li>
<code>Count</code>: Number of DE genes in the gene set.</li>
</ul>
<p>You may have found the total number of DE genes changes. There are
1134 in <code>timeDEgenes</code>, but only 983 DE genes are included in
the enrichment result table (in the <code>GeneRatio</code> column). The
main reason is by default DE genes not annotated to any GO gene set are
filtered out. This relates to the “universe” of all genes in ORA, which
we will touch in the end of this section.</p>
<p>There are several additional arguments in
<code>enrichGO()</code>:</p>
<ul>
<li>
<code>universe</code>: the universe set of genes, i.e. total genes
to use. By default it uses union of genes in all gene sets. If it is
set, DE genes and all gene sets will take intersections with it. We will
discuss it in the end of this section.</li>
<li>
<code>minGSSize</code>: Minimal size of gene sets. Normally gene
sets with very small size have very specific biological meanings which
are not helpful too much for the interpretation. Gene sets with size
smaller than it will be removed from the analysis. By default it is
10.</li>
<li>
<code>maxGSSize</code>: Maximal size of gene sets. Normally gene
sets with huge size provide too general biological meanings and are not
helpful either. By default is 500.</li>
<li>
<code>pvalueCutoff</code>: Cutoff for both <em>p</em>-values and
adjusted <em>p</em>-values. By default is 0.05.</li>
<li>
<code>qvalueCutoff</code>: Cutoff for <em>q</em>-values. by default
is 0.2.</li>
</ul>
<p>Note, <code>enrichGO()</code> only returns significant gene sets that
pass the cutoffs<a href="#fn4" class="footnote-ref" id="fnref4"><sup>4</sup></a>. This function design might not be proper
because a function should return all the results no matter they are
significant or not. Later users may need to use the complete enrichment
table for downstream anlaysis. Second, the meaning of
<code>pvalueCutoff</code> is not precise and there is redundancy between
<code>pvalueCutoff</code> and <code>pvalueCutoff</code> (adjusted
<em>p</em>-values and <em>q</em>-values are always non-smaller than raw
<em>p</em>-values). Thus it is suggested to set both
<code>pvalueCutoff</code> and <code>qvalueCutoff</code> to 1 in
<code>enrichGO()</code>.</p>
<div class="codewrapper sourceCode" id="cb79">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeGO</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                     keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                     ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                     OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                     pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                   ID                 Description GeneRatio   BgRatio
GO:0050900 GO:0050900         leukocyte migration    49/969 386/28564
GO:0030595 GO:0030595        leukocyte chemotaxis    35/969 230/28564
GO:0071674 GO:0071674  mononuclear cell migration    30/969 185/28564
GO:0060326 GO:0060326             cell chemotaxis    39/969 308/28564
GO:0035456 GO:0035456 response to interferon-beta    17/969  71/28564
GO:0097529 GO:0097529 myeloid leukocyte migration    31/969 242/28564
                 pvalue     p.adjust       qvalue
GO:0050900 2.163956e-15 1.096260e-11 8.402981e-12
GO:0030595 1.066463e-13 2.701351e-10 2.070622e-10
GO:0071674 1.143223e-12 1.930522e-09 1.479771e-09
GO:0060326 1.728979e-12 2.189752e-09 1.678475e-09
GO:0035456 1.681085e-10 1.703276e-07 1.305584e-07
GO:0097529 2.447325e-10 2.066358e-07 1.583892e-07
                                                                                                                                                                                                                                                                                       geneID
GO:0050900 Tnfsf18/Sell/Slamf9/Fut7/Itga4/Mdk/Grem1/Ada/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Enpp1/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Emp2/Apod/Retnlg/Plg/Dusp1/Ager/Il33/Ch25h
GO:0030595                                                                           Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Lgals3/Ptk2b/Retnlg/Dusp1/Ch25h
GO:0071674                                                                                                         Tnfsf18/Slamf9/Fut7/Itga4/Mdk/Grem1/Il12a/Nbl1/Padi2/Alox5/Trpm4/Hsd3b7/Adam8/Calr/Ccl17/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Lgals3/Ptk2b/Apod/Retnlg/Plg/Dusp1/Ager/Ch25h
GO:0060326                                                  Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Lpar1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Ccl28/Lgals3/Ptk2b/Nr4a1/Retnlg/Dusp1/Ch25h/Plxnb3
GO:0035456                                                                                                                                                                      Aim2/Ifi204/Gbp6/Oas1c/Ifitm6/Bst2/Irgm1/Tgtp1/Tgtp2/Ifi47/Igtp/Irgm2/Ifitm7/Gm4951/F830016B08Rik/Iigp1/Ifit1
GO:0097529                                                                                                    Tnfsf18/Sell/Fut7/Mdk/Grem1/Prex1/Edn3/P2ry12/S100a8/S100a9/Nbl1/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Itgam/Adam8/Ccl17/Enpp1/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Lgals3/Ptk2b/Emp2/Retnlg/Dusp1/Ager
           Count
GO:0050900    49
GO:0030595    35
GO:0071674    30
GO:0060326    39
GO:0035456    17
GO:0097529    31</code></pre>
</div>
<div id="perform-go-enrichment-on-other-organisms" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="perform-go-enrichment-on-other-organisms" class="callout-inner">
<h3 class="callout-title">Perform GO enrichment on other organisms<a class="anchor" aria-label="anchor" href="#perform-go-enrichment-on-other-organisms"></a>
</h3>
<div class="callout-content">
<p>Gene sets are provided as an <code>OrgDb</code> object in
<code>enrichGO()</code>, thus you can perform ORA analysis on any
organism as long as there is a corresponding <code>OrgDb</code>
object.</p>
<ul>
<li>For model organisms, the <code>OrgDb</code> object can be obtained
from the corresponding <strong>org.*.db</strong> package.</li>
<li>For other organisms, the <code>OrgDb</code> object can be found with
the <strong>AnnotationHub</strong> package.</li>
</ul>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="kegg-pathway-enrichment">KEGG pathway enrichment<a class="anchor" aria-label="anchor" href="#kegg-pathway-enrichment"></a>
</h3>
<p>To perform KEGG pathway enrichment analysis, there is also a function
<code>enrichKEGG()</code> for that. Unfortunately, it cannot perform
gene ID conversion automatically. Thus, if the ID type is not Entrez ID,
we have to convert it by hand.</p>
<p>Note if you have also set <code>universe</code>, it should be
convereted to Entrez IDs as well.</p>
<p>We use the <code>mapIds()</code> function to convert genes from
symbols to Entrez IDs. Since the ID mapping is not always one-to-one. We
only take the first one if there are multiple hits by setting
<code>multiVals = "first"</code>(but of course you can choose other
options for <code>multiVals</code>, check the documentation). We also
remove genes with no mapping available (with <code>NA</code> after the
mapping)<a href="#fn5" class="footnote-ref" id="fnref5"><sup>5</sup></a>.</p>
<div class="codewrapper sourceCode" id="cb81">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EntrezIDs</span> <span class="op">=</span> <span class="fu">mapIds</span><span class="op">(</span><span class="va">org.Mm.eg.db</span>, keys <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>    keytype <span class="op">=</span> <span class="st">"SYMBOL"</span>, column <span class="op">=</span> <span class="st">"ENTREZID"</span>, multiVals <span class="op">=</span> <span class="st">"first"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>'select()' returned 1:1 mapping between keys and columns</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb83">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">EntrezIDs</span> <span class="op">=</span> <span class="va">EntrezIDs</span><span class="op">[</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">EntrezIDs</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">EntrezIDs</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>    Sgk3    Kcnb2   Sbspon    Gsta3   Lman2l  Ankrd39 
"170755"  "98741" "226866"  "14859" "214895" "109346" </code></pre>
</div>
<p>We have to set the KEGG organism code if it is not human. Similarly
it is suggested to set <code>pvalueCutoff</code> and
<code>qvalueCutoff</code> both to 1 and convert the result to a data
frame.</p>
<div class="codewrapper sourceCode" id="cb85">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeKEGG</span> <span class="op">=</span> <span class="fu">enrichKEGG</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">EntrezIDs</span>, </span>
<span>                         organism <span class="op">=</span> <span class="st">"mmu"</span>,</span>
<span>                         pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                         qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeKEGGTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeKEGG</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeKEGGTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>               ID
mmu00590 mmu00590
mmu00565 mmu00565
mmu00592 mmu00592
mmu00591 mmu00591
mmu04913 mmu04913
mmu04061 mmu04061
                                                                                        Description
mmu00590                                   Arachidonic acid metabolism - Mus musculus (house mouse)
mmu00565                                        Ether lipid metabolism - Mus musculus (house mouse)
mmu00592                               alpha-Linolenic acid metabolism - Mus musculus (house mouse)
mmu00591                                      Linoleic acid metabolism - Mus musculus (house mouse)
mmu04913                                       Ovarian steroidogenesis - Mus musculus (house mouse)
mmu04061 Viral protein interaction with cytokine and cytokine receptor - Mus musculus (house mouse)
         GeneRatio BgRatio       pvalue     p.adjust       qvalue
mmu00590    16/452 85/9327 2.536407e-06 0.0007761404 0.0006861647
mmu00565    11/452 48/9327 1.358365e-05 0.0015199802 0.0013437733
mmu00592     8/452 25/9327 1.490177e-05 0.0015199802 0.0013437733
mmu00591    11/452 50/9327 2.056762e-05 0.0015734230 0.0013910207
mmu04913    12/452 63/9327 4.053030e-05 0.0024804542 0.0021929024
mmu04061    14/452 95/9327 1.785817e-04 0.0081798521 0.0072315858
                                                                                                       geneID
mmu00590 18783/19215/211429/329502/78390/19223/67103/242546/13118/18781/18784/11689/232889/15446/237625/11687
mmu00565                               18783/211429/329502/78390/22239/18781/18784/232889/320981/237625/53897
mmu00592                                                  18783/211429/329502/78390/18781/18784/232889/237625
mmu00591                               18783/211429/329502/78390/242546/18781/18784/13113/232889/237625/11687
mmu04913                          18783/211429/329502/78390/242546/11689/232889/13076/13070/15485/13078/16867
mmu04061                  16174/20311/57349/56744/14825/20295/20296/20306/20304/20305/12775/56838/16185/16186
         Count
mmu00590    16
mmu00565    11
mmu00592     8
mmu00591    11
mmu04913    12
mmu04061    14</code></pre>
</div>
<div id="perform-kegg-pathway-enrichment-on-other-organisms" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="perform-kegg-pathway-enrichment-on-other-organisms" class="callout-inner">
<h3 class="callout-title">Perform KEGG pathway enrichment on other
organisms<a class="anchor" aria-label="anchor" href="#perform-kegg-pathway-enrichment-on-other-organisms"></a>
</h3>
<div class="callout-content">
<p>Extending ORA to other organisms is rather simple.</p>
<ol style="list-style-type: decimal">
<li>Make sure the DE genes are Entrez IDs.</li>
<li>Choose the corresponding KEGG organism code.</li>
</ol>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="msigdb-enrichment">MSigDB enrichment<a class="anchor" aria-label="anchor" href="#msigdb-enrichment"></a>
</h3>
<p>For MSigDB gene sets, there is no pre-defined enrichment function. We
need to directly use the low-level enrichment function
<code>enricher()</code> which accepts self-defined gene sets. The gene
sets should be in a format of a two-column data frame of genes and gene
sets (or a class that can be converted to a data frame).</p>
<div class="codewrapper sourceCode" id="cb87">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">msigdbr</span><span class="op">)</span></span>
<span><span class="va">gene_sets</span> <span class="op">=</span> <span class="fu">msigdbr</span><span class="op">(</span>category <span class="op">=</span> <span class="st">"H"</span>, species <span class="op">=</span> <span class="st">"mouse"</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 6 × 18
  gs_cat gs_subcat gs_name               gene_symbol entrez_gene ensembl_gene   
  &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;                 &lt;chr&gt;             &lt;int&gt; &lt;chr&gt;          
1 H      ""        HALLMARK_ADIPOGENESIS Abca1             11303 ENSMUSG0000001…
2 H      ""        HALLMARK_ADIPOGENESIS Abcb8             74610 ENSMUSG0000002…
3 H      ""        HALLMARK_ADIPOGENESIS Acaa2             52538 ENSMUSG0000003…
4 H      ""        HALLMARK_ADIPOGENESIS Acadl             11363 ENSMUSG0000002…
5 H      ""        HALLMARK_ADIPOGENESIS Acadm             11364 ENSMUSG0000006…
6 H      ""        HALLMARK_ADIPOGENESIS Acads             11409 ENSMUSG0000002…
# ℹ 12 more variables: human_gene_symbol &lt;chr&gt;, human_entrez_gene &lt;int&gt;,
#   human_ensembl_gene &lt;chr&gt;, gs_id &lt;chr&gt;, gs_pmid &lt;chr&gt;, gs_geoid &lt;chr&gt;,
#   gs_exact_source &lt;chr&gt;, gs_url &lt;chr&gt;, gs_description &lt;chr&gt;, taxon_id &lt;int&gt;,
#   ortholog_sources &lt;chr&gt;, num_ortholog_sources &lt;dbl&gt;</code></pre>
</div>
<p>As mentioned before, it is important the gene ID type in the gene
sets should be the same as in the DE genes, so here we choose the
<code>"gene_symbol"</code> column.</p>
<div class="codewrapper sourceCode" id="cb89">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeHallmark</span> <span class="op">=</span> <span class="fu">enricher</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                           TERM2GENE <span class="op">=</span> <span class="va">gene_sets</span><span class="op">[</span>, <span class="fu">c</span><span class="op">(</span><span class="st">"gs_name"</span>, <span class="st">"gene_symbol"</span><span class="op">)</span><span class="op">]</span>,</span>
<span>                           pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                           qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeHallmarkTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeHallmark</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeHallmarkTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                                                             ID
HALLMARK_MYOGENESIS                         HALLMARK_MYOGENESIS
HALLMARK_COMPLEMENT                         HALLMARK_COMPLEMENT
HALLMARK_COAGULATION                       HALLMARK_COAGULATION
HALLMARK_ALLOGRAFT_REJECTION       HALLMARK_ALLOGRAFT_REJECTION
HALLMARK_ESTROGEN_RESPONSE_LATE HALLMARK_ESTROGEN_RESPONSE_LATE
HALLMARK_INFLAMMATORY_RESPONSE   HALLMARK_INFLAMMATORY_RESPONSE
                                                    Description GeneRatio
HALLMARK_MYOGENESIS                         HALLMARK_MYOGENESIS    31/291
HALLMARK_COMPLEMENT                         HALLMARK_COMPLEMENT    26/291
HALLMARK_COAGULATION                       HALLMARK_COAGULATION    20/291
HALLMARK_ALLOGRAFT_REJECTION       HALLMARK_ALLOGRAFT_REJECTION    23/291
HALLMARK_ESTROGEN_RESPONSE_LATE HALLMARK_ESTROGEN_RESPONSE_LATE    23/291
HALLMARK_INFLAMMATORY_RESPONSE   HALLMARK_INFLAMMATORY_RESPONSE    22/291
                                 BgRatio       pvalue     p.adjust       qvalue
HALLMARK_MYOGENESIS             201/4394 5.710171e-06 0.0002740882 0.0002284068
HALLMARK_COMPLEMENT             196/4394 4.300844e-04 0.0103220246 0.0086016872
HALLMARK_COAGULATION            139/4394 7.088433e-04 0.0113414921 0.0094512434
HALLMARK_ALLOGRAFT_REJECTION    201/4394 6.397659e-03 0.0767719033 0.0639765861
HALLMARK_ESTROGEN_RESPONSE_LATE 206/4394 8.584183e-03 0.0824081536 0.0686734614
HALLMARK_INFLAMMATORY_RESPONSE  201/4394 1.256794e-02 0.0960740794 0.0800617329
                                                                                                                                                                                                                geneID
HALLMARK_MYOGENESIS             Myl1/Casq1/Aplnr/Tnnc2/Ptgis/Gja5/Col15a1/Cav3/Tnnt1/Ryr1/Cox6a2/Tnni2/Lsp1/Nqo1/Hspb2/Cryab/Erbb3/Stc2/Gpx3/Sparc/Myh3/Myh1/Col1a1/Cacng1/Itgb4/Bdkrb2/Mapk12/Apod/Spdef/Cdkn1a/Actn3
HALLMARK_COMPLEMENT                                                Serpinb2/Pla2g4a/Cd46/Hspa5/Cp/Gnb4/S100a9/Cda/Cxcl1/Apoc1/Itgam/Irf7/Klkb1/Mmp15/Mmp8/Ccl5/Lgals3/Gzmb/Tmprss6/Maff/Plg/Psmb9/Hspa1a/C3/Dusp5/Phex
HALLMARK_COAGULATION                                                                                         Serpinb2/Ctse/C8g/Ctsk/Masp2/Pf4/Sh2b2/Vwf/Apoc1/Klkb1/Mmp15/Mmp8/Trf/Sparc/Itgb3/Dct/Tmprss6/Maff/Plg/C3
HALLMARK_ALLOGRAFT_REJECTION                                                           Il18rap/Cd247/Il12a/Pf4/Capg/Ccnd2/Igsf6/Irf7/Il12rb1/Icosl/Itk/Nos2/Ccl2/Ccl7/Ccl5/Gpr65/Gzmb/Il2rb/Tap1/Tap2/H2-T23/Cfp/Il2rg
HALLMARK_ESTROGEN_RESPONSE_LATE                                                    Clic3/Ass1/Mdk/S100a9/Cdc20/Nbl1/Cav1/Idh2/Th/Pdlim3/Ascl1/Xbp1/Top2a/Dcxr/Fos/Serpina3n/Dhrs2/Tst/Emp2/Fkbp5/Mapk13/Cyp4f15/Kif20a
HALLMARK_INFLAMMATORY_RESPONSE                                                              Il18rap/Sell/Aplnr/Abca1/Lpar1/Cxcl5/Adm/Irf7/Msr1/Bst2/Ccl17/Icosl/Ccl2/Ccl7/Ccl5/Ccr7/Itgb3/Has2/Il2rb/Cdkn1a/Cd70/Best1
                                Count
HALLMARK_MYOGENESIS                31
HALLMARK_COMPLEMENT                26
HALLMARK_COAGULATION               20
HALLMARK_ALLOGRAFT_REJECTION       23
HALLMARK_ESTROGEN_RESPONSE_LATE    23
HALLMARK_INFLAMMATORY_RESPONSE     22</code></pre>
</div>
<div id="further-reading-4" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="further-reading-4" class="callout-inner">
<h3 class="callout-title">Further reading<a class="anchor" aria-label="anchor" href="#further-reading-4"></a>
</h3>
<div class="callout-content">
<p>Implementing ORA is rather simple. The following function
<code>ora()</code> performs ORA on a list of gene sets. Try to read and
understand the code.</p>
<div class="codewrapper sourceCode" id="cb91">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ora</span> <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">genes</span>, <span class="va">gene_sets</span>, <span class="va">universe</span> <span class="op">=</span> <span class="cn">NULL</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span><span class="op">(</span><span class="fu">is.null</span><span class="op">(</span><span class="va">universe</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">universe</span> <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="fu">unlist</span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">universe</span> <span class="op">=</span> <span class="fu">unique</span><span class="op">(</span><span class="va">universe</span><span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co"># make sure genes are unique</span></span>
<span>    <span class="va">genes</span> <span class="op">=</span> <span class="fu">intersect</span><span class="op">(</span><span class="va">genes</span>, <span class="va">universe</span><span class="op">)</span></span>
<span>    <span class="va">gene_sets</span> <span class="op">=</span> <span class="fu">lapply</span><span class="op">(</span><span class="va">gene_sets</span>, <span class="va">intersect</span>, <span class="va">universe</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># calculate different numbers</span></span>
<span>    <span class="va">n_11</span> <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">gene_sets</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">genes</span>, <span class="va">x</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">n_10</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">genes</span><span class="op">)</span></span>
<span>    <span class="va">n_01</span> <span class="op">=</span> <span class="fu">sapply</span><span class="op">(</span><span class="va">gene_sets</span>, <span class="va">length</span><span class="op">)</span></span>
<span>    <span class="va">n</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">universe</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># calculate p-values</span></span>
<span>    <span class="va">p</span> <span class="op">=</span> <span class="fl">1</span> <span class="op">-</span> <span class="fu">phyper</span><span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="fl">1</span>, <span class="va">n_10</span>, <span class="va">n</span> <span class="op">-</span> <span class="va">n_10</span>, <span class="va">n_01</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">df</span> <span class="op">=</span> <span class="fu">data.frame</span><span class="op">(</span></span>
<span>        gene_set <span class="op">=</span> <span class="fu">names</span><span class="op">(</span><span class="va">gene_sets</span><span class="op">)</span>,</span>
<span>        hits <span class="op">=</span> <span class="va">n_11</span>,</span>
<span>        n_genes <span class="op">=</span> <span class="va">n_10</span>,</span>
<span>        gene_set_size <span class="op">=</span> <span class="va">n_01</span>,</span>
<span>        n_total <span class="op">=</span> <span class="va">n</span>,</span>
<span>        p_value <span class="op">=</span> <span class="va">p</span>,</span>
<span>        p_adjust <span class="op">=</span> <span class="fu">p.adjust</span><span class="op">(</span><span class="va">p</span>, <span class="st">"BH"</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre>
</div>
<p>Test on the MSigDB hallmark gene sets:</p>
<div class="codewrapper sourceCode" id="cb92">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">HallmakrGeneSets</span> <span class="op">=</span> <span class="fu">split</span><span class="op">(</span><span class="va">gene_sets</span><span class="op">$</span><span class="va">gene_symbol</span>, <span class="va">gene_sets</span><span class="op">$</span><span class="va">gs_name</span><span class="op">)</span></span>
<span><span class="va">df</span> <span class="op">=</span> <span class="fu">ora</span><span class="op">(</span><span class="va">timeDEgenes</span>, <span class="va">HallmakrGeneSets</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">df</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                                                 gene_set hits n_genes
HALLMARK_ADIPOGENESIS               HALLMARK_ADIPOGENESIS    9    1134
HALLMARK_ALLOGRAFT_REJECTION HALLMARK_ALLOGRAFT_REJECTION   23    1134
HALLMARK_ANDROGEN_RESPONSE     HALLMARK_ANDROGEN_RESPONSE    6    1134
HALLMARK_ANGIOGENESIS               HALLMARK_ANGIOGENESIS    4    1134
HALLMARK_APICAL_JUNCTION         HALLMARK_APICAL_JUNCTION   14    1134
HALLMARK_APICAL_SURFACE           HALLMARK_APICAL_SURFACE    3    1134
                             gene_set_size n_total     p_value    p_adjust
HALLMARK_ADIPOGENESIS                  185   21198 0.662392856 0.871569547
HALLMARK_ALLOGRAFT_REJECTION           191   21198 0.000234807 0.002935087
HALLMARK_ANDROGEN_RESPONSE             107   21198 0.512665196 0.753919406
HALLMARK_ANGIOGENESIS                   36   21198 0.124351058 0.260206520
HALLMARK_APICAL_JUNCTION               186   21198 0.124899130 0.260206520
HALLMARK_APICAL_SURFACE                 41   21198 0.376915129 0.588929889</code></pre>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="choose-a-proper-universe">Choose a proper universe<a class="anchor" aria-label="anchor" href="#choose-a-proper-universe"></a>
</h3>
<p>Finally, it is time to talk about the “universe” of ORA analysis
which is normally ignored in many analyses. In current tools, there are
mainly following different universe settings:</p>
<ol style="list-style-type: decimal">
<li>Using all genes in the genome, this also includes non-protein coding
genes. For human, the size of universe is 60k ~ 70k.</li>
<li>Using all protein-coding genes. For human, the size of universe is ~
20k.</li>
<li>In the era of microarray, total genes that are measured on the chip
is taken as the universe. For RNASeq, since reads are aligned to all
genes, we can set a cutoff and only use those “expressed” genes as the
universe.</li>
<li>Using all genes in a gene sets collection. Then the size of the
universe depends on the size of the gene sets collection. For example GO
gene sets collection is much larger than the KEGG pathway gene sets
collection.</li>
</ol>
<p>If the universe is set, DE genes as well as genes in the gene sets
are first intersected to the universe. However, in general the universe
affects three values of <span class="math inline">\(n_{22}\)</span>,
<span class="math inline">\(n_{02}\)</span> and <span class="math inline">\(n_{20}\)</span> more, which correspond to the
non-DE genes or non-gene-set genes.</p>
<center>
<table class="table">
<colgroup>
<col width="19%">
<col width="28%">
<col width="38%">
<col width="13%">
</colgroup>
<thead><tr class="header">
<th></th>
<th>In the gene set</th>
<th>Not in the gene set</th>
<th>Total</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><strong>DE</strong></td>
<td><span class="math inline">\(n_{11}\)</span></td>
<td><span class="math inline">\(n_{12}\)</span></td>
<td><span class="math inline">\(n_{1+}\)</span></td>
</tr>
<tr class="even">
<td><strong>Not DE</strong></td>
<td><span class="math inline">\(n_{21}\)</span></td>
<td><span class="math inline">\(\color{red}{n_{22}}\)</span></td>
<td><span class="math inline">\(\color{red}{n_{2+}}\)</span></td>
</tr>
<tr class="odd">
<td><strong>Total</strong></td>
<td><span class="math inline">\(n_{+1}\)</span></td>
<td><span class="math inline">\(\color{red}{n_{+2}}\)</span></td>
<td><span class="math inline">\(\color{red}{n}\)</span></td>
</tr>
</tbody>
</table>
</center>
<p>In the contigency table, we are testing the dependency of whether
genes being DE and whether genes being in the gene set. In the model,
each gene has a definite attribute of being either DE or non-DE and each
gene has a second definite attribute of either belonging to the gene set
or not. If a larger universe is used, such as total genes where there
are genes not measured nor will never annotated to gene sets (let’s call
them <em>non-informative genes</em>, e.g. non-proteining coding genes or
not-expressed genes), all the non-informative genes are implicitly
assigned with an attribute of being non-DE or not in the gene set. This
implicit assignment is not proper because these genes provide no
information and they should not be included in the analysis. Adding them
to the analysis increases <span class="math inline">\(n_{22}\)</span>,
<span class="math inline">\(n_{02}\)</span> or <span class="math inline">\(n_{20}\)</span>, makes the observation <span class="math inline">\(n_{11}\)</span> getting further away from the null
distribution, eventually generates a smaller <em>p</em>-value. For the
similar reason, small universes tend to generate large
<em>p</em>-values.</p>
<p>In
<code>enrichGO()</code>/<code>enrichKEGG()</code>/<code>enricher()</code>,
universe genes can be set via the <code>universe</code> argument. By
default the universe is the total genes in the gene sets collection.
When a self-defined universe is provided, this might be different from
what you may think, the universe is <a href="https://github.com/YuLab-SMU/DOSE/blob/93a4b981c0251e5c6eb1f2413f4a02b8e6d06ff5/R/enricher_internal.R#L65C43-L65C51" class="external-link"><em>the
intersection of user-provided universe and total genes in the gene set
collection</em></a>. Thus the universe setting in
<strong>clusterProfiler</strong> is very conservative.</p>
<p>Check the more discusstions at <a href="https://twitter.com/mdziemann/status/1626407797939384320" class="external-link uri">https://twitter.com/mdziemann/status/1626407797939384320</a>.</p>
<p>We can do a simple experiment on the small MSigDB hallmark gene sets.
We use the <code>ora()</code> function which we have implemented in
previous “Further reading” section and we compare three different
universe settings.</p>
<div class="codewrapper sourceCode" id="cb94">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># all genes in the gene sets collection ~ 4k genes</span></span>
<span><span class="va">df1</span> <span class="op">=</span> <span class="fu">ora</span><span class="op">(</span><span class="va">timeDEgenes</span>, <span class="va">HallmakrGeneSets</span><span class="op">)</span></span>
<span><span class="co"># all protein-coding genes, ~ 20k genes</span></span>
<span><span class="va">df2</span> <span class="op">=</span> <span class="fu">ora</span><span class="op">(</span><span class="va">timeDEgenes</span>, <span class="va">HallmakrGeneSets</span>, <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># all genes in org.Mm.eg.db ~ 70k genes</span></span>
<span><span class="va">df3</span> <span class="op">=</span> <span class="fu">ora</span><span class="op">(</span><span class="va">timeDEgenes</span>, <span class="va">HallmakrGeneSets</span>, </span>
<span>    <span class="fu">keys</span><span class="op">(</span><span class="va">org.Mm.eg.db</span>, keytype <span class="op">=</span> <span class="st">"SYMBOL"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># df1, df2, and df3 are in the same row order, </span></span>
<span><span class="co"># so we can directly compare them</span></span>
<span><span class="fu">plot</span><span class="op">(</span><span class="va">df1</span><span class="op">$</span><span class="va">p_value</span>, <span class="va">df2</span><span class="op">$</span><span class="va">p_value</span>, col <span class="op">=</span> <span class="fl">2</span>, pch <span class="op">=</span> <span class="fl">16</span>, </span>
<span>    xlim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, ylim <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>, </span>
<span>    xlab <span class="op">=</span> <span class="st">"all hallmark genes as universe (p-values)"</span>, ylab <span class="op">=</span> <span class="st">"p-values"</span>,</span>
<span>    main <span class="op">=</span> <span class="st">"compare universes"</span><span class="op">)</span></span>
<span><span class="fu">points</span><span class="op">(</span><span class="va">df1</span><span class="op">$</span><span class="va">p_value</span>, <span class="va">df3</span><span class="op">$</span><span class="va">p_value</span>, col <span class="op">=</span> <span class="fl">4</span>, pch <span class="op">=</span> <span class="fl">16</span><span class="op">)</span></span>
<span><span class="fu">abline</span><span class="op">(</span>a <span class="op">=</span> <span class="fl">0</span>, b <span class="op">=</span> <span class="fl">1</span>, lty <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu">legend</span><span class="op">(</span><span class="st">"topleft"</span>, legend <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"all protein-coding genes as universe"</span>, <span class="st">"all genes as universe"</span><span class="op">)</span>, </span>
<span>    pch <span class="op">=</span> <span class="fl">16</span>, col <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-gene-set-analysis-rendered-unnamed-chunk-45-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>It is very straightforward to see, with a larger universe, there are
more significant gene sets, which may produce potentially more false
positives. This is definitely worse when using all genes in the genome
as universe.</p>
<p>Based on the discussion in this section, the recommendation of using
universe is:</p>
<ol style="list-style-type: decimal">
<li>using protein-coding genes,</li>
<li>using measured genes,</li>
<li>or using a conservative way with
<strong>clusterProfiler</strong>.</li>
</ol>
</div>
</section><section id="visualization"><h2 class="section-heading">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<hr class="half-width">
<p><strong>clusterProfiler</strong> provides a rich set of visualization
methods on the GSEA results, from simple visualization to complex ones.
Complex visualizations are normally visually fancy but do not transfer
too much useful information, and they should only be applied in very
specific scenarios under very specific settings; while simple graphs
normally do better jobs. Recently the visualization code in
<strong>clusterProfiler</strong> has been moved to a new package
<strong>enrichplot</strong>. Let’s first load the
<strong>enrichplot</strong> package. The full sets of visualizations
that <strong>enrichplot</strong> supports can be found from <a href="https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html" class="external-link uri">https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html</a>.</p>
<p>We first re-generate the enrichment table.</p>
<div class="codewrapper sourceCode" id="cb95">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">enrichplot</span><span class="op">)</span></span>
<span><span class="va">resTimeGO</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEgenes</span>, </span>
<span>                     keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                     ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                     OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                     pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                     qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGO</span><span class="op">)</span></span>
<span><span class="fu">head</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                   ID                 Description GeneRatio   BgRatio
GO:0050900 GO:0050900         leukocyte migration    49/969 386/28564
GO:0030595 GO:0030595        leukocyte chemotaxis    35/969 230/28564
GO:0071674 GO:0071674  mononuclear cell migration    30/969 185/28564
GO:0060326 GO:0060326             cell chemotaxis    39/969 308/28564
GO:0035456 GO:0035456 response to interferon-beta    17/969  71/28564
GO:0097529 GO:0097529 myeloid leukocyte migration    31/969 242/28564
                 pvalue     p.adjust       qvalue
GO:0050900 2.163956e-15 1.096260e-11 8.402981e-12
GO:0030595 1.066463e-13 2.701351e-10 2.070622e-10
GO:0071674 1.143223e-12 1.930522e-09 1.479771e-09
GO:0060326 1.728979e-12 2.189752e-09 1.678475e-09
GO:0035456 1.681085e-10 1.703276e-07 1.305584e-07
GO:0097529 2.447325e-10 2.066358e-07 1.583892e-07
                                                                                                                                                                                                                                                                                       geneID
GO:0050900 Tnfsf18/Sell/Slamf9/Fut7/Itga4/Mdk/Grem1/Ada/Prex1/Edn3/P2ry12/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Enpp1/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Aoc3/Itgb3/Ccl28/Lgals3/Ptk2b/Emp2/Apod/Retnlg/Plg/Dusp1/Ager/Il33/Ch25h
GO:0030595                                                                           Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Lgals3/Ptk2b/Retnlg/Dusp1/Ch25h
GO:0071674                                                                                                         Tnfsf18/Slamf9/Fut7/Itga4/Mdk/Grem1/Il12a/Nbl1/Padi2/Alox5/Trpm4/Hsd3b7/Adam8/Calr/Ccl17/Aire/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Itgb3/Lgals3/Ptk2b/Apod/Retnlg/Plg/Dusp1/Ager/Ch25h
GO:0060326                                                  Tnfsf18/Sell/Slamf9/Mdk/Grem1/Prex1/Edn3/Il12a/S100a8/S100a9/Lpar1/Nbl1/Padi2/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Ptn/Alox5/Trpm4/Hsd3b7/Itgam/Adam8/Calr/Ccl17/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Ccl28/Lgals3/Ptk2b/Nr4a1/Retnlg/Dusp1/Ch25h/Plxnb3
GO:0035456                                                                                                                                                                      Aim2/Ifi204/Gbp6/Oas1c/Ifitm6/Bst2/Irgm1/Tgtp1/Tgtp2/Ifi47/Igtp/Irgm2/Ifitm7/Gm4951/F830016B08Rik/Iigp1/Ifit1
GO:0097529                                                                                                    Tnfsf18/Sell/Fut7/Mdk/Grem1/Prex1/Edn3/P2ry12/S100a8/S100a9/Nbl1/Bst1/Cxcl5/Ppbp/Pf4/Cxcl1/Itgam/Adam8/Ccl17/Enpp1/Ccl2/Ccl7/Ccl5/Ccl6/Ccr7/Lgals3/Ptk2b/Emp2/Retnlg/Dusp1/Ager
           Count
GO:0050900    49
GO:0030595    35
GO:0071674    30
GO:0060326    39
GO:0035456    17
GO:0097529    31</code></pre>
</div>
<p><code>barplot()</code> and <code>dotplot()</code> generate plots for
a small number of significant gene sets. Note the two functions are
directly applied on <code>resTimeGO</code> returned by
<code>enrichGO()</code>.</p>
<div class="codewrapper sourceCode" id="cb97">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">barplot</span><span class="op">(</span><span class="va">resTimeGO</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-gene-set-analysis-rendered-unnamed-chunk-47-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb98">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">dotplot</span><span class="op">(</span><span class="va">resTimeGO</span>, showCategory <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-gene-set-analysis-rendered-unnamed-chunk-47-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Barplots can map two variables to the plot, one to the height of bars
and the other to the colors of bars; while for dotplot, sizes of dots
can be mapped to a third variable. The variable names are in the colum
names of the result table. Both plots include the top 20 most
significant terms. On dotplot, terms are ordered by the values on x-axis
(the <code>GeneRatio</code>).</p>
<p>Now we need to talk about “what is a good visualization?”. The
essential two questions are “<em>what is the key message a plot transfer
to readers?</em>” and “<em>what is the major graphical element in the
plot?</em>”. In the barplot or dotplot, the major graphical element
which readers may notice the easiest is the height of bars or the offset
of dots to the origin. The most important message of the ORA anaysis is
of course “the enrichment”. The two examples from <code>barplot()</code>
and <code>dotplot()</code> actually fail to transfer such information to
readers. In the first barplot where <code>"Count"</code> is used as
values on x-axis, the numer of DE genes in gene sets is not a good
measure of the enrichment because it has a positive relation to the size
of gene sets. A high value of <code>"Count"</code> does not mean the
gene set is more enriched.</p>
<p>It is the same reason for dotplot where <code>"GeneRatio"</code> is
used as values on x-axis. Gene ratio is calculated as the fraction of DE
genes from a certain gene set (GeneRatio = Count/Total_DE_Genes). The
dotplot puts multiple gene sets in the same plot and the aim is to
compare between gene sets, thus gene sets should be “scaled” to make
them comparible. <code>"GeneRatio"</code> is not scaled for different
gene sets and it still has a positive relation to the gene set size,
which can be observed in the dotplot where higher the gene ratio, larger
the dot size. Actually “GeneRatio” has the same effect as “Count”
(GeneRatio = Count/Total_DE_Genes), so as has been explained in the
previous paragraph, <code>"GeneRatio"</code> is not a good measure for
enrichment either.</p>
<p>Now let’s try to make a more reasonable barplot and dotplot to show
the enrichment of ORA.</p>
<p>First, let’s define some metrics which measure the “enrichment” of DE
genes on gene sets. Recall the denotations in the 2x2 contigency table
(we are too far from that!). Let’s take these numbers from the
enrichment table.</p>
<div class="codewrapper sourceCode" id="cb99">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">n_11</span> <span class="op">=</span> <span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">Count</span></span>
<span><span class="va">n_10</span> <span class="op">=</span> <span class="fl">983</span>  <span class="co"># length(intersect(resTimeGO@gene, resTimeGO@universe))</span></span>
<span><span class="va">n_01</span> <span class="op">=</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="fu">gsub</span><span class="op">(</span><span class="st">"/.*$"</span>, <span class="st">""</span>, <span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">BgRatio</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fl">28943</span>  <span class="co"># length(resTimeGO@universe)</span></span></code></pre>
</div>
<p>Instead of using <code>GeneRatio</code>, we use the fraction of DE
genes in the gene sets which are kind like a “scaled” value for all gene
sets. Let’s calculate it:</p>
<div class="codewrapper sourceCode" id="cb100">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">DE_Ratio</span> <span class="op">=</span> <span class="va">n_11</span><span class="op">/</span><span class="va">n_01</span></span>
<span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">GS_size</span> <span class="op">=</span> <span class="va">n_01</span>  <span class="co"># size of gene sets</span></span></code></pre>
</div>
<p>Then intuitively, if a gene set has a higher <code>DE_Ratio</code>
value, we could say DE genes have a higher enrichment<a href="#fn6" class="footnote-ref" id="fnref6"><sup>6</sup></a> in it.</p>
<p>We can measure the enrichment in two other ways. First, the log2 fold
enrichment, defined as:</p>
<p><span class="math display">\[ \log_2(\mathrm{Fold\_enrichment}) =
\frac{n_{11}/n_{10}}{n_{01}/n} = \frac{n_{11}/n_{01}}{n_{10}/n} =
\frac{n_{11}n}{n_{10}n_{01}} \]</span></p>
<p>which is the log2 of the ratio of <em>DE% in the gene set</em> and
<em>DE% in the universe</em> or the log2 of the ratio of <em>gene_set%
in the DE genes</em> and <em>gene_set% in the universe</em>. The two are
identical.</p>
<div class="codewrapper sourceCode" id="cb101">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">log2_Enrichment</span> <span class="op">=</span> <span class="fu">log</span><span class="op">(</span> <span class="op">(</span><span class="va">n_11</span><span class="op">/</span><span class="va">n_10</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="va">n_01</span><span class="op">/</span><span class="va">n</span><span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<p>Second, it is also common to use <em>z</em>-score which is</p>
<p><span class="math display">\[ z = \frac{n_{11} - \mu}{\sigma}
\]</span></p>
<p>where <span class="math inline">\(\mu\)</span> and <span class="math inline">\(\sigma\)</span> are <a href="https://en.wikipedia.org/wiki/Hypergeometric_distribution" class="external-link">the
mean and standard deviation of the hypergeometric distribution</a>. They
can be calculated as:</p>
<div class="codewrapper sourceCode" id="cb102">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hyper_mean</span> <span class="op">=</span> <span class="va">n_01</span><span class="op">*</span><span class="va">n_10</span><span class="op">/</span><span class="va">n</span></span>
<span></span>
<span><span class="va">n_02</span> <span class="op">=</span> <span class="va">n</span> <span class="op">-</span> <span class="va">n_01</span></span>
<span><span class="va">n_20</span> <span class="op">=</span> <span class="va">n</span> <span class="op">-</span> <span class="va">n_10</span></span>
<span><span class="va">hyper_var</span> <span class="op">=</span> <span class="va">n_01</span><span class="op">*</span><span class="va">n_10</span><span class="op">/</span><span class="va">n</span> <span class="op">*</span> <span class="va">n_20</span><span class="op">*</span><span class="va">n_02</span><span class="op">/</span><span class="va">n</span><span class="op">/</span><span class="op">(</span><span class="va">n</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">zScore</span> <span class="op">=</span> <span class="op">(</span><span class="va">n_11</span> <span class="op">-</span> <span class="va">hyper_mean</span><span class="op">)</span><span class="op">/</span><span class="fu">sqrt</span><span class="op">(</span><span class="va">hyper_var</span><span class="op">)</span></span></code></pre>
</div>
<p>We will use log2 fold change as the primary variable to map to bar
heights and <code>DE_Ratio</code> as the secondary variable to map to
colors. This can be done directly with the <strong>ggplot2</strong>
package. We also add the adjusted <em>p</em>-values as labels on the
bars.</p>
<p>In <code>resTimeGOTable</code>, gene sets are already ordered by the
significance, so we take the first 10 gene sets which are the 10 most
significant gene sets.</p>
<div class="codewrapper sourceCode" id="cb103">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">library</span><span class="op">(</span><span class="va">ggplot2</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span>, </span>
<span>        <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, y <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">Description</span>, levels <span class="op">=</span> <span class="fu">rev</span><span class="op">(</span><span class="va">Description</span><span class="op">)</span><span class="op">)</span>, </span>
<span>            fill <span class="op">=</span> <span class="va">DE_Ratio</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_text</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, </span>
<span>        label <span class="op">=</span> <span class="fu">sprintf</span><span class="op">(</span><span class="st">"%.2e"</span>, <span class="va">p.adjust</span><span class="op">)</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-gene-set-analysis-rendered-unnamed-chunk-52-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In the next example, we use <em>z</em>-score as the primary variable
to map to the offset to origin, <code>DE_Ratio</code> and
<code>Count</code> to map to dot colors and sizes.</p>
<div class="codewrapper sourceCode" id="cb104">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">resTimeGOTable</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, <span class="op">]</span>, </span>
<span>        <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">zScore</span>, y <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">Description</span>, levels <span class="op">=</span> <span class="fu">rev</span><span class="op">(</span><span class="va">Description</span><span class="op">)</span><span class="op">)</span>, </span>
<span>            col <span class="op">=</span> <span class="va">DE_Ratio</span>, size <span class="op">=</span> <span class="va">Count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-gene-set-analysis-rendered-unnamed-chunk-53-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Both plots can highlight the gene set “leukocyte migration involved
in inflammatory response” is relatively small but highly enriched.</p>
<p>Another useful visualization is the volcano plot. You may be aware of
in differential expression analysis, in the volcano plot, x-axis
corresponds to log2 fold changes of the differential expression, and
y-axis corresponds to the adjusted <em>p</em>-values. It is actually
similar here where we use log2 fold enrichment on x-axis.</p>
<p>Since we only look at the over-representation, the volcano plot is
one-sided. We can set two cutoffs on the log2 fold enrichment and
adjusted <em>p</em>-values, then the gene sets on the top right region
can be thought as being both statistically significant and also
biologically sensible.</p>
<div class="codewrapper sourceCode" id="cb105">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">ggplot</span><span class="op">(</span><span class="va">resTimeGOTable</span>, </span>
<span>    <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="va">p.adjust</span><span class="op">)</span>, </span>
<span>        color <span class="op">=</span> <span class="va">DE_Ratio</span>, size <span class="op">=</span> <span class="va">GS_size</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_point</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_hline</span><span class="op">(</span>yintercept <span class="op">=</span> <span class="op">-</span><span class="fu">log10</span><span class="op">(</span><span class="fl">0.01</span><span class="op">)</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"#444444"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_vline</span><span class="op">(</span>xintercept <span class="op">=</span> <span class="fl">1.5</span>, lty <span class="op">=</span> <span class="fl">2</span>, col <span class="op">=</span> <span class="st">"#444444"</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-gene-set-analysis-rendered-unnamed-chunk-54-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>In the “volcano plot”, we can observe the plot is composed by a list
of curves. The trends are especially clear in the right bottom of the
plot. Actually each “curve” corresponds to a same <code>"Count"</code>
value (number of DE genes in a gene set). The volcano plot shows very
clearly that the enrichment has a positive relation to the gene set size
where a large gene set can easily reach a small <em>p</em>-value with a
small DE_ratio and a small log2 fold enrichment, while a small gene set
needs to have a large DE ratio to be significant.</p>
<p>It is also common that we perform ORA analysis on up-regulated genes
and down-regulated separately. And we want to combine the significant
gene sets from the two ORA analysis in one plot. In the following code,
we first generate two enrichment tables for up-regulated genes and
down-regulated separately.</p>
<div class="codewrapper sourceCode" id="cb106">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># up-regulated genes</span></span>
<span><span class="va">timeDEup</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resTime</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&gt;</span> <span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">timeDEupGenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">timeDEup</span><span class="op">)</span></span>
<span></span>
<span><span class="va">resTimeGOup</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEupGenes</span>, </span>
<span>                       keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                       ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                       OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                       universe <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,</span>
<span>                       pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOupTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGOup</span><span class="op">)</span></span>
<span><span class="va">n_11</span> <span class="op">=</span> <span class="va">resTimeGOupTable</span><span class="op">$</span><span class="va">Count</span></span>
<span><span class="va">n_10</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">resTimeGOup</span><span class="op">@</span><span class="va">gene</span>, <span class="va">resTimeGOup</span><span class="op">@</span><span class="va">universe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n_01</span> <span class="op">=</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="fu">gsub</span><span class="op">(</span><span class="st">"/.*$"</span>, <span class="st">""</span>, <span class="va">resTimeGOupTable</span><span class="op">$</span><span class="va">BgRatio</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">resTimeGOup</span><span class="op">@</span><span class="va">universe</span><span class="op">)</span></span>
<span><span class="va">resTimeGOupTable</span><span class="op">$</span><span class="va">log2_Enrichment</span> <span class="op">=</span> <span class="fu">log</span><span class="op">(</span> <span class="op">(</span><span class="va">n_11</span><span class="op">/</span><span class="va">n_10</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="va">n_01</span><span class="op">/</span><span class="va">n</span><span class="op">)</span> <span class="op">)</span></span>
<span></span>
<span><span class="co"># down-regulated genes</span></span>
<span><span class="va">timeDEdown</span> <span class="op">&lt;-</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="fu">subset</span><span class="op">(</span><span class="va">resTime</span>, <span class="va">padj</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;</span> <span class="va">log2FoldChange</span> <span class="op">&lt;</span> <span class="op">-</span><span class="fu">log2</span><span class="op">(</span><span class="fl">1.5</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">timeDEdownGenes</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">timeDEdown</span><span class="op">)</span></span>
<span></span>
<span><span class="va">resTimeGOdown</span> <span class="op">=</span> <span class="fu">enrichGO</span><span class="op">(</span>gene <span class="op">=</span> <span class="va">timeDEdownGenes</span>, </span>
<span>                       keyType <span class="op">=</span> <span class="st">"SYMBOL"</span>,</span>
<span>                       ont <span class="op">=</span> <span class="st">"BP"</span>, </span>
<span>                       OrgDb <span class="op">=</span> <span class="va">org.Mm.eg.db</span>,</span>
<span>                       universe <span class="op">=</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">se</span><span class="op">)</span>,</span>
<span>                       pvalueCutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       qvalueCutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">resTimeGOdownTable</span> <span class="op">=</span> <span class="fu">as.data.frame</span><span class="op">(</span><span class="va">resTimeGOdown</span><span class="op">)</span></span>
<span><span class="va">n_11</span> <span class="op">=</span> <span class="va">resTimeGOdownTable</span><span class="op">$</span><span class="va">Count</span></span>
<span><span class="va">n_10</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="fu">intersect</span><span class="op">(</span><span class="va">resTimeGOdown</span><span class="op">@</span><span class="va">gene</span>, <span class="va">resTimeGOdown</span><span class="op">@</span><span class="va">universe</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n_01</span> <span class="op">=</span> <span class="fu">as.numeric</span><span class="op">(</span><span class="fu">gsub</span><span class="op">(</span><span class="st">"/.*$"</span>, <span class="st">""</span>, <span class="va">resTimeGOdownTable</span><span class="op">$</span><span class="va">BgRatio</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">n</span> <span class="op">=</span> <span class="fu">length</span><span class="op">(</span><span class="va">resTimeGOdown</span><span class="op">@</span><span class="va">universe</span><span class="op">)</span></span>
<span><span class="va">resTimeGOdownTable</span><span class="op">$</span><span class="va">log2_Enrichment</span> <span class="op">=</span> <span class="fu">log</span><span class="op">(</span> <span class="op">(</span><span class="va">n_11</span><span class="op">/</span><span class="va">n_10</span><span class="op">)</span><span class="op">/</span><span class="op">(</span><span class="va">n_01</span><span class="op">/</span><span class="va">n</span><span class="op">)</span> <span class="op">)</span></span></code></pre>
</div>
<p>As an example, let’s simply take the first 5 most significant terms
for up-regulated genes and the first 5 most significant terms for
down-regulated genes. The following <strong>ggplot2</strong> code should
be easy to read.</p>
<div class="codewrapper sourceCode" id="cb107">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># The name of the 4th term is too long, we wrap it into two lines.</span></span>
<span><span class="va">resTimeGOupTable</span><span class="op">[</span><span class="fl">4</span>, <span class="st">"Description"</span><span class="op">]</span> <span class="op">=</span> <span class="fu">paste</span><span class="op">(</span><span class="fu">strwrap</span><span class="op">(</span><span class="va">resTimeGOupTable</span><span class="op">[</span><span class="fl">4</span>, <span class="st">"Description"</span><span class="op">]</span><span class="op">)</span>, collapse <span class="op">=</span> <span class="st">"\n"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">direction</span> <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="fu">rep</span><span class="op">(</span><span class="st">"up"</span>, <span class="fl">5</span><span class="op">)</span>, <span class="fu">rep</span><span class="op">(</span><span class="st">"down"</span>, <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">ggplot</span><span class="op">(</span><span class="fu">rbind</span><span class="op">(</span><span class="va">resTimeGOupTable</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span>,</span>
<span>             <span class="va">resTimeGOdownTable</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span><span class="op">)</span>,</span>
<span>        <span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, y <span class="op">=</span> <span class="fu">factor</span><span class="op">(</span><span class="va">Description</span>, levels <span class="op">=</span> <span class="fu">rev</span><span class="op">(</span><span class="va">Description</span><span class="op">)</span><span class="op">)</span>, </span>
<span>            fill <span class="op">=</span> <span class="va">direction</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_bar</span><span class="op">(</span>stat <span class="op">=</span> <span class="st">"identity"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">scale_fill_manual</span><span class="op">(</span>values <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"up"</span> <span class="op">=</span> <span class="st">"red"</span>, <span class="st">"down"</span> <span class="op">=</span> <span class="st">"darkgreen"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">geom_text</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>x <span class="op">=</span> <span class="va">log2_Enrichment</span>, </span>
<span>        label <span class="op">=</span> <span class="fu">sprintf</span><span class="op">(</span><span class="st">"%.2e"</span>, <span class="va">p.adjust</span><span class="op">)</span><span class="op">)</span>, hjust <span class="op">=</span> <span class="fl">1</span>, col <span class="op">=</span> <span class="st">"white"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">ylab</span><span class="op">(</span><span class="st">""</span><span class="op">)</span></span></code></pre>
</div>
<figure><img src="fig/06-gene-set-analysis-rendered-unnamed-chunk-56-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>Specifically for GO enrichment, it is often that GO enrichment
returns a long list of significant GO terms (e.g. several hundreds).
This makes it difficult to summarize the common functions from the long
list. The last package we will introduce is the
<strong>simplifyEnrichment</strong> package which partitions GO terms
into clusters based on their semantic similarity<a href="#fn7" class="footnote-ref" id="fnref7"><sup>7</sup></a> and summarizes their
common functions via word clouds.</p>
<p>The input of the <code>simplifyGO()</code> function is a vector of GO
IDs. It is recommended to have at least 100 GO IDs for summarization and
visualization.</p>
<div class="codewrapper sourceCode" id="cb108">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">GO_ID</span> <span class="op">=</span> <span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">ID</span><span class="op">[</span><span class="va">resTimeGOTable</span><span class="op">$</span><span class="va">p.adjust</span> <span class="op">&lt;</span> <span class="fl">0.1</span><span class="op">]</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="va">simplifyEnrichment</span><span class="op">)</span></span>
<span><span class="fu">simplifyGO</span><span class="op">(</span><span class="va">GO_ID</span><span class="op">)</span></span></code></pre>
</div>
<!-- Carpentry CI cannot install RcppParallel which is indirectly depended by simplifyenrichment,
    so the plot is not generated on the fly.
 -->
<figure><img src="simplifyEnrichment.png" class="figure mx-auto d-block"></figure><div id="further-reading-5" class="callout callout">
<div class="callout-square">
<i class="callout-icon" data-feather="bell"></i>
</div>
<div id="further-reading-5" class="callout-inner">
<h3 class="callout-title">Further reading<a class="anchor" aria-label="anchor" href="#further-reading-5"></a>
</h3>
<div class="callout-content">
<p>ORA analysis actually applies a binary conversion on genes where
genes pass the cutoff are set as 1 (DE gene) and others are set as 0
(non-DE gene). This binary transformation over-simplifies the problem
and a lot of information are lost. There is second class of gene set
enrichment analysis methods which takes the continuous gene-level score
as input and weights the importance of a gene in the gene set. Please
refer to <a href="https://www.pnas.org/doi/10.1073/pnas.0506580102" class="external-link">Subramanian et.
al. Gene set enrichment analysis: A knowledge-based approach for
interpreting genome-wide expression profiles, PNAS 2005</a> for more
information.</p>
</div>
</div>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>ORA analysis is based on the gene counts and it is based on Fisher’s
exact test or the hypergeometric distribution.</li>
<li>In R, it is easy to obtain gene sets from a large number of
sources.</li>
</ul>
</div>
</div>
</div>
</section><div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>Genes must be unique in each vector.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Also note <code>phyper()</code> can be vectorized.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>The definition is from Wikipedia: <a href="https://en.wikipedia.org/wiki/Biological_pathway" class="external-link uri">https://en.wikipedia.org/wiki/Biological_pathway</a>.<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
<li id="fn4"><p>This is actually not true. Indeed
<code>as.data.frame(resTimeGO)</code> only returns the significant GO
terms, but the complete enrichment table is still stored in
<code>resTimeGO@result</code>. However, directly retrieving the slot of
an S4 object is highly unrecommended.<a href="#fnref4" class="footnote-back">↩︎</a></p></li>
<li id="fn5"><p>You can also use <code>select()</code> function:
<code>select(org.Mm.eg.db, keys = timeDEgenes, keytype = "SYMBOL", column = "ENTREZID")</code><a href="#fnref5" class="footnote-back">↩︎</a></p></li>
<li id="fn6"><p>If here the term “enrichment” does mean statistically.<a href="#fnref6" class="footnote-back">↩︎</a></p></li>
<li id="fn7"><p>The semantic similarity between GO terms considers the
topological relations in the GO hierarchy.<a href="#fnref7" class="footnote-back">↩︎</a></p></li>
</ol>
</div></section><section id="aio-07-next-steps"><p>Content from <a href="07-next-steps.html">Next steps</a></p>
<hr>
<p> Last updated on 2023-07-17 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/07-next-steps.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How to go further from here?</li>
<li>What other types of analyses can be done with RNA-seq data?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Get an overview of usages of RNA-seq data that are not covered in
this workshop.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="transcript-level-analyses"><h2 class="section-heading">Transcript-level analyses<a class="anchor" aria-label="anchor" href="#transcript-level-analyses"></a>
</h2>
<hr class="half-width">
<p>The analyses covered in this workshop all assumed that we have
generated a matrix with read counts for each <em>gene</em>. Some
questions, however, require expression estimates on a more fine-grained
level, typically individual transcript isoforms. This is the case, for
example, if we would like to look for differences in expression of
individual isoforms, or changes in splicing patterns induced by a
specific treatment. As already mentioned in episode 1, some
quantification tools (such as Salmon, kallisto and RSEM) do in fact
estimate abundances on the transcript level. To perform transcript-level
differential expression analysis, these estimates would be used
directly, without aggregating them on the gene level. Some caution is
warranted, however, as expression estimates for individual transcripts
are often more noisy than after aggregation to the gene level. This is
due to the high similarity often observed between different isoforms of
a gene, which leads to higher uncertainty when mapping reads to
transcripts. For this reason, several approaches have been developed
specifically for performing differential expression analysis on the
transcript level <span class="citation">[@Zhu2019-swish,
@Baldoni2023-catchsalmon]</span>.</p>
<p>Analysis of differential splicing, or differential transcript usage,
differs from the differential expression analyses mentioned so far as it
is concerned with changes in the <em>relative</em> abundances of the
transcripts of a gene. Hence, considering the transcripts in isolation
is no longer sufficient. A good resource for learning more about
differential transcript usage with Bioconductor is provided by <span class="citation">@Love2018-swimming</span>.</p>
</section><section id="de-novo-transcript-assembly"><h2 class="section-heading">De novo transcript assembly<a class="anchor" aria-label="anchor" href="#de-novo-transcript-assembly"></a>
</h2>
<hr class="half-width">
<p>In this lesson, we have assumed that we are working with a
well-annotated transcriptome, that is, that we know the sequence of all
expressed isoforms and how they are grouped together into genes.
Depending on the organism or disease type we are interested in, this may
or may not be a reasonable assumption. In situations where we do not
believe that the annotated transcriptome is complete enough, we can use
the RNA-seq data to <em>assemble</em> transcripts and create a custom
annotation. This assembly can be performed either guided by a genome
sequence (if one is available), or completely de novo. It should be
noted that transcript assembly is a challenging task, which requires a
deeply sequenced library to get the best results. In addition, data from
more recent long-read sequencing technologies can be very helpful, as
they are in principle able to sequence entire transcript molecules, thus
circumventing the need for assembly. Methods for transcript assembly
represent an active area of research. Recent review are provided e.g. by
<span class="citation">@Raghavan2022-denovo</span> and <span class="citation">@Amarasinghe2020-longread</span>.</p>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>RNA-seq data is very versatile and can be used for a number of
different purposes. It is important, however, to carefully plan one’s
analyses, to make sure that enough data is available and that abundances
for appropriate features (e.g., genes, transcripts, or exons) are
quantified.</li>
</ul>
</div>
</div>
</div>
</section></section><section id="aio-08-extra-design"><p>Content from <a href="08-extra-design.html">Extra exploration of design matrices</a></p>
<hr>
<p> Last updated on 2023-07-17 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/episodes/08-extra-design.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
<div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: replacing previous import 'S4Arrays::makeNindexFromArrayViewport' by
'DelayedArray::makeNindexFromArrayViewport' when loading 'SummarizedExperiment'</code></pre>
</div>
<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul>
<li>How can one translate biological questions and comparisons to
statistical terms suitable for use with RNA-seq analysis packages?</li>
</ul>
</div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul>
<li>Explain the formula notation and design matrices.</li>
<li>Explore different designs and learn how to interpret
coefficients.</li>
</ul>
</div>
</div>
</div>
</div>
</div>
<section id="loading-required-packages-and-reading-data"><h2 class="section-heading">Loading required packages and reading data<a class="anchor" aria-label="anchor" href="#loading-required-packages-and-reading-data"></a>
</h2>
<hr class="half-width">
<p>We start by loading a few packages that will be needed in this
episode. In particular, the <a href="https://bioconductor.org/packages/ExploreModelMatrix/" class="external-link">ExploreModelMatrix</a>
package provides resources for exploring design matrices in a graphical
fashion, for easier interpretation.</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">suppressPackageStartupMessages</span><span class="op">(</span><span class="op">{</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">SummarizedExperiment</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">ExploreModelMatrix</span><span class="op">)</span></span>
<span>    <span class="kw">library</span><span class="op">(</span><span class="va">dplyr</span><span class="op">)</span></span>
<span><span class="op">}</span><span class="op">)</span></span></code></pre>
</div>
<p>Next, we read the metadata table for our data set. As seen in
previous episodes, the metadata contains information about the age, sex,
infection status, time of measurement and tissue of the collected
samples. Note that Day0 always corresponds to non-infected samples, and
that infected samples are collected on days 4 and 8. Moreover, all mice
have the same age (8 weeks). Hence, in the first part of this episode we
consider only the sex, tissue and time variables further.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu">read.csv</span><span class="op">(</span><span class="st">"data/GSE96870_coldata_all.csv"</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">meta</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545336 CNS_RNA-seq_10C    GSM2545336 Mus musculus 8 weeks Female
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545339 CNS_RNA-seq_13C    GSM2545339 Mus musculus 8 weeks Female
GSM2545340 CNS_RNA-seq_14C    GSM2545340 Mus musculus 8 weeks   Male
GSM2545341 CNS_RNA-seq_17C    GSM2545341 Mus musculus 8 weeks   Male
GSM2545342  CNS_RNA-seq_1C    GSM2545342 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545344 CNS_RNA-seq_21C    GSM2545344 Mus musculus 8 weeks Female
GSM2545345 CNS_RNA-seq_22C    GSM2545345 Mus musculus 8 weeks   Male
GSM2545346 CNS_RNA-seq_25C    GSM2545346 Mus musculus 8 weeks   Male
GSM2545347 CNS_RNA-seq_26C    GSM2545347 Mus musculus 8 weeks   Male
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545350 CNS_RNA-seq_29C    GSM2545350 Mus musculus 8 weeks   Male
GSM2545351  CNS_RNA-seq_2C    GSM2545351 Mus musculus 8 weeks Female
GSM2545352 CNS_RNA-seq_30C    GSM2545352 Mus musculus 8 weeks Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545355 CNS_RNA-seq_571    GSM2545355 Mus musculus 8 weeks   Male
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545359 CNS_RNA-seq_585    GSM2545359 Mus musculus 8 weeks Female
GSM2545360 CNS_RNA-seq_589    GSM2545360 Mus musculus 8 weeks   Male
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545362  CNS_RNA-seq_5C    GSM2545362 Mus musculus 8 weeks Female
GSM2545363  CNS_RNA-seq_6C    GSM2545363 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
GSM2545368 CNS_RNA-seq_728    GSM2545368 Mus musculus 8 weeks   Male
GSM2545369 CNS_RNA-seq_729    GSM2545369 Mus musculus 8 weeks   Male
GSM2545370 CNS_RNA-seq_730    GSM2545370 Mus musculus 8 weeks Female
GSM2545371 CNS_RNA-seq_731    GSM2545371 Mus musculus 8 weeks Female
GSM2545372 CNS_RNA-seq_733    GSM2545372 Mus musculus 8 weeks   Male
GSM2545373 CNS_RNA-seq_735    GSM2545373 Mus musculus 8 weeks   Male
GSM2545374 CNS_RNA-seq_736    GSM2545374 Mus musculus 8 weeks Female
GSM2545375 CNS_RNA-seq_738    GSM2545375 Mus musculus 8 weeks Female
GSM2545376 CNS_RNA-seq_740    GSM2545376 Mus musculus 8 weeks Female
GSM2545377 CNS_RNA-seq_741    GSM2545377 Mus musculus 8 weeks Female
GSM2545378 CNS_RNA-seq_742    GSM2545378 Mus musculus 8 weeks   Male
GSM2545379 CNS_RNA-seq_743    GSM2545379 Mus musculus 8 weeks   Male
GSM2545380  CNS_RNA-seq_9C    GSM2545380 Mus musculus 8 weeks Female
             infection  strain time     tissue mouse
GSM2545336  InfluenzaA C57BL/6 Day8 Cerebellum    14
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545339  InfluenzaA C57BL/6 Day4 Cerebellum    15
GSM2545340  InfluenzaA C57BL/6 Day4 Cerebellum    18
GSM2545341  InfluenzaA C57BL/6 Day8 Cerebellum     6
GSM2545342  InfluenzaA C57BL/6 Day8 Cerebellum     5
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11
GSM2545344  InfluenzaA C57BL/6 Day4 Cerebellum    22
GSM2545345  InfluenzaA C57BL/6 Day4 Cerebellum    13
GSM2545346  InfluenzaA C57BL/6 Day8 Cerebellum    23
GSM2545347  InfluenzaA C57BL/6 Day8 Cerebellum    24
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7
GSM2545350  InfluenzaA C57BL/6 Day4 Cerebellum     1
GSM2545351  InfluenzaA C57BL/6 Day8 Cerebellum    16
GSM2545352  InfluenzaA C57BL/6 Day4 Cerebellum    21
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2
GSM2545355  InfluenzaA C57BL/6 Day4 Spinalcord     1
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545359  InfluenzaA C57BL/6 Day8 Spinalcord     5
GSM2545360  InfluenzaA C57BL/6 Day8 Spinalcord     6
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545362  InfluenzaA C57BL/6 Day4 Cerebellum    20
GSM2545363  InfluenzaA C57BL/6 Day4 Cerebellum    12
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11
GSM2545368  InfluenzaA C57BL/6 Day4 Spinalcord    12
GSM2545369  InfluenzaA C57BL/6 Day4 Spinalcord    13
GSM2545370  InfluenzaA C57BL/6 Day8 Spinalcord    14
GSM2545371  InfluenzaA C57BL/6 Day4 Spinalcord    15
GSM2545372  InfluenzaA C57BL/6 Day8 Spinalcord    17
GSM2545373  InfluenzaA C57BL/6 Day4 Spinalcord    18
GSM2545374  InfluenzaA C57BL/6 Day8 Spinalcord    19
GSM2545375  InfluenzaA C57BL/6 Day4 Spinalcord    20
GSM2545376  InfluenzaA C57BL/6 Day4 Spinalcord    21
GSM2545377  InfluenzaA C57BL/6 Day4 Spinalcord    22
GSM2545378  InfluenzaA C57BL/6 Day8 Spinalcord    23
GSM2545379  InfluenzaA C57BL/6 Day8 Spinalcord    24
GSM2545380  InfluenzaA C57BL/6 Day8 Cerebellum    19</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">time</span>, <span class="va">meta</span><span class="op">$</span><span class="va">infection</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>      
       InfluenzaA NonInfected
  Day0          0          15
  Day4         16           0
  Day8         14           0</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">table</span><span class="op">(</span><span class="va">meta</span><span class="op">$</span><span class="va">age</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
8 weeks 
     45 </code></pre>
</div>
<p>We can start by visualizing the number of observations for each
combination of the three predictor variables.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">tissue</span> <span class="op">+</span> <span class="va">time</span> <span class="op">+</span> <span class="va">sex</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">cooccurrenceplots</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$`tissue = Cerebellum`</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-5-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
$`tissue = Spinalcord`</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-5-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="challenge" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge"></a>
</h3>
<div class="callout-content">
<p>Based on this visualization, would you say that the data set is
balanced, or are there combinations of predictor variables that are
severely over- or underrepresented?</p>
</div>
</div>
</div>
</section><section id="compare-males-and-females-non-infected-spinal-cord"><h2 class="section-heading">Compare males and females, non-infected spinal cord<a class="anchor" aria-label="anchor" href="#compare-males-and-females-non-infected-spinal-cord"></a>
</h2>
<hr class="half-width">
<p>Next, we will set up our first design matrix. Here, we will focus on
the uninfected (Day0) spinal cord samples, and our aim is to compare the
male and female mice. Thus, we first subset the metadata to only the
samples of interest, and next set up and visualize the design matrix
with a single predictor variable (sex). By defining the design formula
as <code>~ sex</code>, we tell R to include an intercept in the design.
This intercept will represent the ‘baseline’ level of the predictor
variable, which in this case is selected to be the Female mice. If not
explicitly specified, R will order the values of the predictor in
alphabetical order and select the first one as the reference or baseline
level.</p>
<div class="codewrapper sourceCode" id="cb12">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Subset metadata</span></span>
<span><span class="va">meta_noninf_spc</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span> <span class="op">&amp;</span> </span>
<span>                                       <span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span><span class="op">)</span></span>
<span><span class="va">meta_noninf_spc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb14">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Use ExploreModelMatrix to create a design matrix and visualizations, given </span></span>
<span><span class="co">## the desired design formula. </span></span>
<span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_noninf_spc</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) sexMale
GSM2545356           1       1
GSM2545357           1       1
GSM2545358           1       0
GSM2545361           1       1
GSM2545364           1       0
GSM2545365           1       0
GSM2545366           1       0
GSM2545367           1       1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb16">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-6-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper sourceCode" id="cb18">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Note that we can also generate the design matrix like this</span></span>
<span><span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span> <span class="va">sex</span>, data <span class="op">=</span> <span class="va">meta_noninf_spc</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) sexMale
GSM2545356           1       1
GSM2545357           1       1
GSM2545358           1       0
GSM2545361           1       1
GSM2545364           1       0
GSM2545365           1       0
GSM2545366           1       0
GSM2545367           1       1
attr(,"assign")
[1] 0 1
attr(,"contrasts")
attr(,"contrasts")$sex
[1] "contr.treatment"</code></pre>
</div>
<div id="challenge-1" class="callout discussion">
<div class="callout-square">
<i class="callout-icon" data-feather="message-circle"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge-1"></a>
</h3>
<div class="callout-content">
<p>With this design, what is the interpretation of the
<code>sexMale</code> coefficient?</p>
</div>
</div>
</div>
<div id="challenge-2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge-2"></a>
</h3>
<div class="callout-content">
<p>Set up the design formula to compare male and female spinal cord
samples from Day0 as above, but instruct R to not include an intercept
in the model. How does this change the interpretation of the
coefficients? What contrast would have to be specified to compare the
mean expression of a gene between male and female mice?</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" aria-labelledby="headingSolution1" data-bs-parent="#accordionSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb20">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_noninf_spc</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span> <span class="op">&amp;</span> </span>
<span>                                       <span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span><span class="op">)</span></span>
<span><span class="va">meta_noninf_spc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb22">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_noninf_spc</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="fl">0</span> <span class="op">+</span> <span class="va">sex</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           sexFemale sexMale
GSM2545356         0       1
GSM2545357         0       1
GSM2545358         1       0
GSM2545361         0       1
GSM2545364         1       0
GSM2545365         1       0
GSM2545366         1       0
GSM2545367         0       1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb24">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-7-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
<div id="challenge-3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge-3"></a>
</h3>
<div class="callout-content">
<p>Set up the design formula to compare the three time points (Day0,
Day4, Day8) in the male spinal cord samples, and visualize it using
<code>ExploreModelMatrix</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" aria-labelledby="headingSolution2" data-bs-parent="#accordionSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb26">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_male_spc</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span> <span class="op">&amp;</span> <span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span><span class="op">)</span></span>
<span><span class="va">meta_male_spc</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age  sex   infection
GSM2545355 CNS_RNA-seq_571    GSM2545355 Mus musculus 8 weeks Male  InfluenzaA
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks Male NonInfected
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks Male NonInfected
GSM2545360 CNS_RNA-seq_589    GSM2545360 Mus musculus 8 weeks Male  InfluenzaA
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks Male NonInfected
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks Male NonInfected
GSM2545368 CNS_RNA-seq_728    GSM2545368 Mus musculus 8 weeks Male  InfluenzaA
GSM2545369 CNS_RNA-seq_729    GSM2545369 Mus musculus 8 weeks Male  InfluenzaA
GSM2545372 CNS_RNA-seq_733    GSM2545372 Mus musculus 8 weeks Male  InfluenzaA
GSM2545373 CNS_RNA-seq_735    GSM2545373 Mus musculus 8 weeks Male  InfluenzaA
GSM2545378 CNS_RNA-seq_742    GSM2545378 Mus musculus 8 weeks Male  InfluenzaA
GSM2545379 CNS_RNA-seq_743    GSM2545379 Mus musculus 8 weeks Male  InfluenzaA
            strain time     tissue mouse
GSM2545355 C57BL/6 Day4 Spinalcord     1
GSM2545356 C57BL/6 Day0 Spinalcord     2
GSM2545357 C57BL/6 Day0 Spinalcord     3
GSM2545360 C57BL/6 Day8 Spinalcord     6
GSM2545361 C57BL/6 Day0 Spinalcord     7
GSM2545367 C57BL/6 Day0 Spinalcord    11
GSM2545368 C57BL/6 Day4 Spinalcord    12
GSM2545369 C57BL/6 Day4 Spinalcord    13
GSM2545372 C57BL/6 Day8 Spinalcord    17
GSM2545373 C57BL/6 Day4 Spinalcord    18
GSM2545378 C57BL/6 Day8 Spinalcord    23
GSM2545379 C57BL/6 Day8 Spinalcord    24</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb28">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_male_spc</span>, designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">time</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) timeDay4 timeDay8
GSM2545355           1        1        0
GSM2545356           1        0        0
GSM2545357           1        0        0
GSM2545360           1        0        1
GSM2545361           1        0        0
GSM2545367           1        0        0
GSM2545368           1        1        0
GSM2545369           1        1        0
GSM2545372           1        0        1
GSM2545373           1        1        0
GSM2545378           1        0        1
GSM2545379           1        0        1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb30">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-8-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure>
</div>
</div>
</div>
</div>
</section><section id="factorial-design-without-interactions"><h2 class="section-heading">Factorial design without interactions<a class="anchor" aria-label="anchor" href="#factorial-design-without-interactions"></a>
</h2>
<hr class="half-width">
<p>Next, we again consider only non-infected mice, but fit a model
incorporating both sex and tissue as predictors. We assume that the
tissue differences are the same for both male and female mice, and
consequently fit an additive model, without interaction terms.</p>
<div class="codewrapper sourceCode" id="cb32">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_noninf</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span><span class="op">)</span></span>
<span><span class="va">meta_noninf</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb34">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_noninf</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">+</span> <span class="va">tissue</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) sexMale tissueSpinalcord
GSM2545337           1       0                0
GSM2545338           1       0                0
GSM2545343           1       1                0
GSM2545348           1       0                0
GSM2545349           1       1                0
GSM2545353           1       0                0
GSM2545354           1       1                0
GSM2545356           1       1                1
GSM2545357           1       1                1
GSM2545358           1       0                1
GSM2545361           1       1                1
GSM2545364           1       0                1
GSM2545365           1       0                1
GSM2545366           1       0                1
GSM2545367           1       1                1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb36">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-9-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="factorial-design-with-interactions"><h2 class="section-heading">Factorial design with interactions<a class="anchor" aria-label="anchor" href="#factorial-design-with-interactions"></a>
</h2>
<hr class="half-width">
<p>In the previous model, we assumed that the tissue differences were
the same for both male and female mice. To allow for the estimation of
sex-specific tissue differences (at the expense of having one additional
coefficient to estimate from the data), we can include an interaction
term in the model.</p>
<div class="codewrapper sourceCode" id="cb38">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_noninf</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span><span class="op">)</span></span>
<span><span class="va">meta_noninf</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545343 CNS_RNA-seq_20C    GSM2545343 Mus musculus 8 weeks   Male
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545349 CNS_RNA-seq_28C    GSM2545349 Mus musculus 8 weeks   Male
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545354  CNS_RNA-seq_4C    GSM2545354 Mus musculus 8 weeks   Male
GSM2545356 CNS_RNA-seq_574    GSM2545356 Mus musculus 8 weeks   Male
GSM2545357 CNS_RNA-seq_575    GSM2545357 Mus musculus 8 weeks   Male
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545361 CNS_RNA-seq_590    GSM2545361 Mus musculus 8 weeks   Male
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545367 CNS_RNA-seq_713    GSM2545367 Mus musculus 8 weeks   Male
             infection  strain time     tissue mouse
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545343 NonInfected C57BL/6 Day0 Cerebellum    11
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545349 NonInfected C57BL/6 Day0 Cerebellum     7
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545354 NonInfected C57BL/6 Day0 Cerebellum     2
GSM2545356 NonInfected C57BL/6 Day0 Spinalcord     2
GSM2545357 NonInfected C57BL/6 Day0 Spinalcord     3
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545361 NonInfected C57BL/6 Day0 Spinalcord     7
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545367 NonInfected C57BL/6 Day0 Spinalcord    11</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb40">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Define a design including an interaction term</span></span>
<span><span class="co">## Note that ~ sex * tissue is equivalent to </span></span>
<span><span class="co">## ~ sex + tissue + sex:tissue</span></span>
<span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_noninf</span>, </span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">sex</span> <span class="op">*</span> <span class="va">tissue</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) sexMale tissueSpinalcord sexMale:tissueSpinalcord
GSM2545337           1       0                0                        0
GSM2545338           1       0                0                        0
GSM2545343           1       1                0                        0
GSM2545348           1       0                0                        0
GSM2545349           1       1                0                        0
GSM2545353           1       0                0                        0
GSM2545354           1       1                0                        0
GSM2545356           1       1                1                        1
GSM2545357           1       1                1                        1
GSM2545358           1       0                1                        0
GSM2545361           1       1                1                        1
GSM2545364           1       0                1                        0
GSM2545365           1       0                1                        0
GSM2545366           1       0                1                        0
GSM2545367           1       1                1                        1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb42">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-10-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="paired-design"><h2 class="section-heading">Paired design<a class="anchor" aria-label="anchor" href="#paired-design"></a>
</h2>
<hr class="half-width">
<p>In this particular data set the samples are paired - the same mice
have contributed both the cerebellum and spinal cord samples. This
information was not included in the previous models. However, accounting
for it can increase power to detect tissue differences by eliminating
variability in baseline expression levels between mice. Here, we define
a paired design for the female non-infected mice, aimed at testing for
differences between tissues after accounting for baseline differences
between mice.</p>
<div class="codewrapper sourceCode" id="cb44">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_fem_day0</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span> <span class="op">&amp;</span> </span>
<span>                                     <span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># ensure that mouse is treated as a categorical variable</span></span>
<span><span class="va">meta_fem_day0</span><span class="op">$</span><span class="va">mouse</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">meta_fem_day0</span><span class="op">$</span><span class="va">mouse</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta_fem_day0</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
             infection  strain time     tissue mouse
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb46">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_fem_day0</span>,</span>
<span>                      designFormula <span class="op">=</span> <span class="op">~</span> <span class="va">mouse</span> <span class="op">+</span> <span class="va">tissue</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) mouse8 mouse9 mouse10 tissueSpinalcord
GSM2545337           1      0      1       0                0
GSM2545338           1      0      0       1                0
GSM2545348           1      1      0       0                0
GSM2545353           1      0      0       0                0
GSM2545358           1      0      0       0                1
GSM2545364           1      1      0       0                1
GSM2545365           1      0      1       0                1
GSM2545366           1      0      0       1                1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb48">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>[[1]]</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-11-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure></section><section id="within--and-between-subject-comparisons"><h2 class="section-heading">Within- and between-subject comparisons<a class="anchor" aria-label="anchor" href="#within--and-between-subject-comparisons"></a>
</h2>
<hr class="half-width">
<p>In some situations, we need to combine the types of models considered
above. For example, let’s say that we want to investigate if the tissue
differences are different for infected and non-infected female mice. In
this case, each mice only contributes to one of the infection groups
(each mice is either infected or non-infected), but contributes both a
cerebellum and a spinal cord sample. One way to view this type of design
is as two paired experiments, one for each infection group (see the <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/edgeR/inst/doc/edgeRUsersGuide.pdf" class="external-link">edgeR
user guide section 3.5</a>). We can then easily compare the two tissues
in each infection group, and contrast the tissue differences between the
infection groups.</p>
<div class="codewrapper sourceCode" id="cb50">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta_fem_day04</span> <span class="op">&lt;-</span> <span class="va">meta</span> <span class="op">%&gt;%</span> </span>
<span>    <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span> <span class="op">&amp;</span> </span>
<span>               <span class="va">time</span> <span class="op">%in%</span> <span class="fu">c</span><span class="op">(</span><span class="st">"Day0"</span>, <span class="st">"Day4"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">droplevels</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co"># ensure that mouse is treated as a categorical variable</span></span>
<span><span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">mouse</span> <span class="op">&lt;-</span> <span class="fu">factor</span><span class="op">(</span><span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">mouse</span><span class="op">)</span></span>
<span></span>
<span><span class="va">meta_fem_day04</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>                     title geo_accession     organism     age    sex
GSM2545337 CNS_RNA-seq_11C    GSM2545337 Mus musculus 8 weeks Female
GSM2545338 CNS_RNA-seq_12C    GSM2545338 Mus musculus 8 weeks Female
GSM2545339 CNS_RNA-seq_13C    GSM2545339 Mus musculus 8 weeks Female
GSM2545344 CNS_RNA-seq_21C    GSM2545344 Mus musculus 8 weeks Female
GSM2545348 CNS_RNA-seq_27C    GSM2545348 Mus musculus 8 weeks Female
GSM2545352 CNS_RNA-seq_30C    GSM2545352 Mus musculus 8 weeks Female
GSM2545353  CNS_RNA-seq_3C    GSM2545353 Mus musculus 8 weeks Female
GSM2545358 CNS_RNA-seq_583    GSM2545358 Mus musculus 8 weeks Female
GSM2545362  CNS_RNA-seq_5C    GSM2545362 Mus musculus 8 weeks Female
GSM2545364 CNS_RNA-seq_709    GSM2545364 Mus musculus 8 weeks Female
GSM2545365 CNS_RNA-seq_710    GSM2545365 Mus musculus 8 weeks Female
GSM2545366 CNS_RNA-seq_711    GSM2545366 Mus musculus 8 weeks Female
GSM2545371 CNS_RNA-seq_731    GSM2545371 Mus musculus 8 weeks Female
GSM2545375 CNS_RNA-seq_738    GSM2545375 Mus musculus 8 weeks Female
GSM2545376 CNS_RNA-seq_740    GSM2545376 Mus musculus 8 weeks Female
GSM2545377 CNS_RNA-seq_741    GSM2545377 Mus musculus 8 weeks Female
             infection  strain time     tissue mouse
GSM2545337 NonInfected C57BL/6 Day0 Cerebellum     9
GSM2545338 NonInfected C57BL/6 Day0 Cerebellum    10
GSM2545339  InfluenzaA C57BL/6 Day4 Cerebellum    15
GSM2545344  InfluenzaA C57BL/6 Day4 Cerebellum    22
GSM2545348 NonInfected C57BL/6 Day0 Cerebellum     8
GSM2545352  InfluenzaA C57BL/6 Day4 Cerebellum    21
GSM2545353 NonInfected C57BL/6 Day0 Cerebellum     4
GSM2545358 NonInfected C57BL/6 Day0 Spinalcord     4
GSM2545362  InfluenzaA C57BL/6 Day4 Cerebellum    20
GSM2545364 NonInfected C57BL/6 Day0 Spinalcord     8
GSM2545365 NonInfected C57BL/6 Day0 Spinalcord     9
GSM2545366 NonInfected C57BL/6 Day0 Spinalcord    10
GSM2545371  InfluenzaA C57BL/6 Day4 Spinalcord    15
GSM2545375  InfluenzaA C57BL/6 Day4 Spinalcord    20
GSM2545376  InfluenzaA C57BL/6 Day4 Spinalcord    21
GSM2545377  InfluenzaA C57BL/6 Day4 Spinalcord    22</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb52">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">model.matrix</span><span class="op">(</span><span class="op">~</span> <span class="va">mouse</span>, data <span class="op">=</span> <span class="va">meta_fem_day04</span><span class="op">)</span></span>
<span><span class="va">design</span> <span class="op">&lt;-</span> <span class="fu">cbind</span><span class="op">(</span><span class="va">design</span>, </span>
<span>                Spc.Day0 <span class="op">=</span> <span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span> <span class="op">&amp;</span> </span>
<span>                    <span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day0"</span>,</span>
<span>                Spc.Day4 <span class="op">=</span> <span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">tissue</span> <span class="op">==</span> <span class="st">"Spinalcord"</span> <span class="op">&amp;</span> </span>
<span>                    <span class="va">meta_fem_day04</span><span class="op">$</span><span class="va">time</span> <span class="op">==</span> <span class="st">"Day4"</span><span class="op">)</span></span>
<span><span class="fu">rownames</span><span class="op">(</span><span class="va">design</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">rownames</span><span class="op">(</span><span class="va">meta_fem_day04</span><span class="op">)</span></span>
<span><span class="va">design</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) mouse8 mouse9 mouse10 mouse15 mouse20 mouse21 mouse22
GSM2545337           1      0      1       0       0       0       0       0
GSM2545338           1      0      0       1       0       0       0       0
GSM2545339           1      0      0       0       1       0       0       0
GSM2545344           1      0      0       0       0       0       0       1
GSM2545348           1      1      0       0       0       0       0       0
GSM2545352           1      0      0       0       0       0       1       0
GSM2545353           1      0      0       0       0       0       0       0
GSM2545358           1      0      0       0       0       0       0       0
GSM2545362           1      0      0       0       0       1       0       0
GSM2545364           1      1      0       0       0       0       0       0
GSM2545365           1      0      1       0       0       0       0       0
GSM2545366           1      0      0       1       0       0       0       0
GSM2545371           1      0      0       0       1       0       0       0
GSM2545375           1      0      0       0       0       1       0       0
GSM2545376           1      0      0       0       0       0       1       0
GSM2545377           1      0      0       0       0       0       0       1
           Spc.Day0 Spc.Day4
GSM2545337        0        0
GSM2545338        0        0
GSM2545339        0        0
GSM2545344        0        0
GSM2545348        0        0
GSM2545352        0        0
GSM2545353        0        0
GSM2545358        1        0
GSM2545362        0        0
GSM2545364        1        0
GSM2545365        1        0
GSM2545366        1        0
GSM2545371        0        1
GSM2545375        0        1
GSM2545376        0        1
GSM2545377        0        1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb54">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span> <span class="op">&lt;-</span> <span class="fu">VisualizeDesign</span><span class="op">(</span>sampleData <span class="op">=</span> <span class="va">meta_fem_day04</span> <span class="op">%&gt;%</span></span>
<span>                          <span class="fu">select</span><span class="op">(</span><span class="va">time</span>, <span class="va">tissue</span>, <span class="va">mouse</span><span class="op">)</span>,</span>
<span>                      designFormula <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                      designMatrix <span class="op">=</span> <span class="va">design</span>, flipCoordFitted <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">vd</span><span class="op">$</span><span class="va">designmatrix</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>           (Intercept) mouse8 mouse9 mouse10 mouse15 mouse20 mouse21 mouse22
GSM2545337           1      0      1       0       0       0       0       0
GSM2545338           1      0      0       1       0       0       0       0
GSM2545339           1      0      0       0       1       0       0       0
GSM2545344           1      0      0       0       0       0       0       1
GSM2545348           1      1      0       0       0       0       0       0
GSM2545352           1      0      0       0       0       0       1       0
GSM2545353           1      0      0       0       0       0       0       0
GSM2545358           1      0      0       0       0       0       0       0
GSM2545362           1      0      0       0       0       1       0       0
GSM2545364           1      1      0       0       0       0       0       0
GSM2545365           1      0      1       0       0       0       0       0
GSM2545366           1      0      0       1       0       0       0       0
GSM2545371           1      0      0       0       1       0       0       0
GSM2545375           1      0      0       0       0       1       0       0
GSM2545376           1      0      0       0       0       0       1       0
GSM2545377           1      0      0       0       0       0       0       1
           Spc.Day0 Spc.Day4
GSM2545337        0        0
GSM2545338        0        0
GSM2545339        0        0
GSM2545344        0        0
GSM2545348        0        0
GSM2545352        0        0
GSM2545353        0        0
GSM2545358        1        0
GSM2545362        0        0
GSM2545364        1        0
GSM2545365        1        0
GSM2545366        1        0
GSM2545371        0        1
GSM2545375        0        1
GSM2545376        0        1
GSM2545377        0        1</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb56">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vd</span><span class="op">$</span><span class="va">plotlist</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>$`time = Day0`</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-12-1.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code>
$`time = Day4`</code></pre>
</div>
<figure><img src="fig/08-extra-design-rendered-unnamed-chunk-12-2.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul>
<li>The formula framework in R allows creation of design matrices, which
details the variables expected to be associated with systematic
differences in gene expression levels.</li>
<li>Comparisons of interest can be defined using contrasts, which are
linear combinations of the model coefficients.</li>
</ul>
</div>
</div>
</div>
</section></section>
</div>
    </main>
</div>
<!-- END  : inst/pkgdown/templates/content-extra.html -->

      </div>
<!--/div.row-->
      		<footer class="row footer mx-md-3"><hr>
<div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/bioc-rnaseq/edit/main/README.md" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/carpentries-incubator/bioc-rnaseq/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-rnaseq/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:charlotte.soneson@fmi.ch">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper" class="external-link">sandpaper (0.12.5)</a>,
        <a href="https://github.com/carpentries/pegboard" class="external-link">pegboard (0.5.3)</a>,
      and <a href="https://github.com/mblue9/varnish/tree/3403ad0dd37f7641c6442330fcc182937e7964c5" class="external-link">varnish (0.2.16)</a>.</p>
			</div>
		</footer>
</div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/bioc-rnaseq/aio.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "All in One View",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-rnaseq/aio.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-rnaseq/aio.html",
  "dateCreated": "2020-09-15",
  "dateModified": "2023-07-18",
  "datePublished": "2023-07-18"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code -->
</body>
</html><!-- END:   inst/pkgdown/templates/layout.html-->

