---
source: Rmd
title: Importing and annotating quantified data into R
teaching: XX
exercises: XX
---

```{r, echo=FALSE, purl=FALSE, message=FALSE}
source("download_data.R")
```

::::::::::::::::::::::::::::::::::::::: objectives
-   Learn how to import the quantifications into a SummarizedExperiment object.
-   Learn how to add additional gene annotations to the object.
::::::::::::::::::::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::: questions
-   How can one import quantified gene expression data into an object suitable for downstream statistical analysis in R?
-   What types of gene identifiers are typically used, and how are mappings between them done?
::::::::::::::::::::::::::::::::::::::::::::::::::


```{r}
suppressPackageStartupMessages({
    library(AnnotationDbi)
    library(org.Mm.eg.db)
    library(SummarizedExperiment)
})

## Read the data

In the last episode, we used R to download 4 files from the internet and saved them on our computer. But we do not have these files loaded into R yet so that we can work with them. The original experimental design in [Blackmore et al. 2017](https://pubmed.ncbi.nlm.nih.gov/28696309/) was fairly complex: 8 week old male and female mice were collected at Day 0 (before influenza infection), Day 4 and Day 8 after influenza infection. From each mouse, cerebellum and spinal cord tissues were taken for RNA-Seq. There were originally 4 mice per Sex X Day group, but a few were lost along the way for a total of 45 samples. For this workshop we are going to simplify the analysis by only using the 22 cerebellum samples. Expression quantification was done using STAR to align to the mouse genome and then counting reads that map to genes. In addition to the counts per gene per sample, we also need information on which sample belongs to which Sex/Timepoint/Replicate . And for the genes, it is helpful to have extra information called annotation.
Let't read in the data files that we downloaded in the last episode:


### Counts

```{r}
counts <- read.csv("data/GSE96870_counts_cerebellum.csv", 
                   row.names = 1)
dim(counts)
# View(counts)
```

Genes are in rows and samples are in columns, so we have counts for 41,786 genes and 22 samples. The `View()` command has been comment out for the website, but running it will open a tab in RStudio that lets us look at the data and even sort the table by a particular column. However, the viewer cannot change the data inside the `counts` object, so we can only look, not permanently sort nor edit the entries. When finished, close the viewer using the X in the tab. Looks like the rownames are gene symbols and the column names are the GEO sample IDs, which are not very informative for telling us which sample is what.

### Sample annotations

Next read in the sample annotations. Because samples are in columns in the count matrix, we will name the object `coldata`:

```{r}
coldata <- read.csv("data/GSE96870_coldata_cerebellum.csv",
                    row.names = 1)
dim(coldata)
# View(coldata)
```

Now samples are in rows with the GEO sample IDs as the rownames, and we have 10 columns of information. The columns that are the most useful for this workshop are `geo_accession` (GEO sample IDs again), `sex` and `time`.

### Gene annotations
The counts only have gene symbols, which while short and somewhat recognizable to the human brain, are not always good absolute identifiers for exactly what gene was measured. For this we need additional gene annotations that were provided by the authors. The `count` and `coldata` files were in comma separated value (.csv) format, but we cannot use that for our gene annotation file because the descriptions can contain commas that would prevent a .csv file from being read in correctly. Instead the gene annotation file is in tab separated value (.tsv) format. Likewise, the descriptions can contain the single quote ' (e.g., 5'), which by default R assumes indicates a character entry. So we have to use a more generic function `read.delim()` with extra arguments to specify that we have tab-separated data (`sep = "\t"`) with no quotes used (`quote = ""`). We also put in other arguments to specify that the first row contains our column names (`header`), the gene symbols that should be our `row.names` are in the 5th column, and that NCBI's species-specific gene ENTREZIDs should be read in as character data even though they look like numbers.

```{r}
rowranges <- read.delim("data/GSE96870_rowranges.tsv", sep = "\t", 
                        colClasses = c(ENTREZID = "character"),
                        header = TRUE, quote = "", row.names = 5)
dim(rowranges)
# View(rowranges)
```

For each of the 41,786 genes, we have the `seqnames` (e.g., chromosome), `start` and `end` position, `strand`, `ENTREZID`, `product` description and the type of gene (`gbkey`). Depending on who generates your count data, you might not have a nice file of additional gene annotations. There may only be the count row names, which could be symbols or ENTREZIDs or another database's ID. Bioconductor has many packages and functions that can help you get additional annotation information for your genes. This is covered in more detail in [Episode 6 Gene set enrichment analysis](https://carpentries-incubator.github.io/bioc-rnaseq/06-gene-set-analysis.html#gene-set-resources) but here is a short example:

The main function we will use is `mapIds`:
```
mapIds(annopkg, keys, column, keytype, ..., multiVals)
```

Where 
- *annopkg* is the annotation package    
- *keys* are the IDs that we **know**    
- *column* is the value we **want**    
- *keytype* is the type of key used

```{r}
mapIds(org.Mm.eg.db, keys = "497097", column = "SYMBOL", keytype = "ENTREZID")
```

Different from the `select` function, `mapIds` function handles 1:many mapping 
between keys and columns through an additional argument, 'multiVals'.

```{r echo=FALSE, eval=FALSE}
library(hgu95av2.db) # Affymetrix Human Genome U95 Set annotation data
keys <- head(keys(hgu95av2.db, "ENTREZID"))
last <- function(x){x[[length(x)]]}

mapIds(hgu95av2.db, keys = keys, column = "ALIAS", keytype = "ENTREZID")
mapIds(hgu95av2.db, keys = keys, column = "ALIAS", keytype = "ENTREZID", multiVals = last)
mapIds(hgu95av2.db, keys = keys, column = "ALIAS", keytype = "ENTREZID", multiVals = "list")
```

```{r}
packageVersion("org.Mm.eg.db") # double-check the version of gene annotation
```

There are many gene annotations available, including RefGene, Ensemble, and 
UCSC. Characteristics of gene annotations differ based on their annotation 
strategies and information sources. For example, RefSeq human gene models 
(i.e., Entrez from NCBI) are well supported and broadly used in various studies.
The UCSC Known Genes dataset is based on protein data from Swiss-Prot/TrEMBL 
(UniProt) and the associated mRNA data from GenBank, and serves as a foundation 
for the UCSC Genome Browser. Ensembl genes contain both automated genome 
annotation and manual curation.

You can find more information in Bioconductor [Annotation Workshop](https://jmacdon.github.io/Bioc2022Anno/articles/AnnotationWorkshop.html)
material.

Check feature types

The `gbkey` column shows us what types of genes we have, so we can check what they are and how many genes of each we have:


```{r}
table(rowranges$gbkey)
```

[have the below be a Challenge discussion ala https://carpentries-incubator.github.io/bioc-intro/10-data-organisation.html#challenge-discuss-the-following-points-with-your-neighbour ? ]

Suppose we decide we only want to analyze mRNA genes and not any of the others. We could remove these genes from the `rowranges` object, but we also have to remember to remove them correctly from the `counts` object. Likewise, we may decide one or more samples are outliers and we would have to remove them from both the `counts` columns and the `coldata` rows. You can see how this could easily lead to mis-matches between our counts and our annotations.

Instead, Bioconductor has created a specialized S4 class called a `SummarizedExperiment`. The details of a `SummarizedExperiment` were covered extensively at the end of the [Introduction to data analysis with R and Bioconductor](https://carpentries-incubator.github.io/bioc-intro/60-next-steps.html#next-steps) workshop. It was designed to hold any type of quantitative 'omics data ("`assay`") along with linked sample annotations ("`colData`") and feature annotations with (`rowRanges`) or without (`rowData`) chromosome, start and stop positions. Once these three tables are (correctly!) linked, subsetting either samples and/or features will correctly subset the `assay`, `colData` and `rowRanges`. Additionally, most Bioconductor packages are built around the same core data infrastructure so they will recognize and be able to manipulate `SummarizedExperiment` objects. Two of the most popular RNA-Seq statistical analysis packages have their own extended S4 classes similar to a `SummarizedExperiment` with addition slots for statistical results: [DESeq2's](http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#the-deseqdataset) `DESeqDataSet` and [edgeR's](https://www.rdocumentation.org/packages/edgeR/versions/3.14.0/topics/DGEList-class) `DGEList`. No matter which one you end up using for statistical analysis, you can start by putting your data in a `SummarizedExperiment`. 

## SummarizedExperiment

### What is SummarizedExperiment?
`SummarizedExperiment` objects are widely used to represent expression data 
and other rectangular data (feature x sample data). The _SummarizedExperiment_ 
package contains two classes: 
`SummarizedExperiment` and `RangedSummarizedExperiment`.

`SummarizedExperiment` is a matrix-like container where rows represent features
of interest (e.g., genes, transcripts, exons, etc.) and columns represent
samples. The objects contain one or more assays, each represented by a
matrix-like object of numeric or other mode. 

Information about these features is stored in a `DataFrame` object, accessible 
using the function `rowData()`. Each row of the `DataFrame` provides information 
on the feature in the corresponding row of the `SummarizedExperiment` object. 
Columns of the `DataFrame` represent different attributes of the features of 
interest (e.g., gene or transcript IDs, etc). Sample meta-data describing the 
samples can be accessed using `colData()`, which returns a `DataFrame` that can 
store any number of descriptive columns for each sample row.

`RangedSummarizedExperiment` is the child of the `SummarizedExperiment` class
which means that all the methods on `SummarizedExperiment` also work on a
`RangedSummarizedExperiment`.

The fundamental difference between the two classes is that the rows of a
`RangedSummarizedExperiment` object represent genomic ranges of interest
instead of a `DataFrame` of features, which is accessible using the
`rowRanges()` function.

![Summarized Experiment](https://raw.githubusercontent.com/Bioconductor/SummarizedExperiment/devel/vignettes/SE.svg)

### Assemble SummarizedExperiment
We will create a `SummarizedExperiment` from these objects:

- The `count` object will be used as the **`assay`**

- The `coldata` object with sample information will be used as the **sample
  metadata** `colData` slot

- The 'rowranges' object describing the genes will be used as the **features
  metadata** `rowRanges` slot

Before we put them together, you ABSOLUTELY MUST MAKE SURE THE SAMPLES AND GENES ARE IN THE SAME ORDER! Even though we saw that `count` and `coldata` had the same number of samples and `count` and `rowranges` had the same number of genes, we never explicitly checked to see if they were in the same order. One way to check:


```{r}
all.equal(colnames(counts), rownames(coldata))
all.equal(rownames(counts), rownames(rowranges))

# If both not TRUE, you could do something like:

if(sum(colnames(counts) %in%  rownames(coldata)) == ncol(counts)) {
  tempindex <- match(colnames(counts), rownames(coldata))
  coldata <- coldata[tempindex, ]
} else {
  print("Warning: the number of samples are not the same in counts and coldata")
}

```

[or we could do a challenge to write the other if else statement ]

Once we have verified that samples and genes are in the same order, we can then create our object after loading the `SummarizedExperiment` package:

```{r}

# One final check:

stopifnot(rownames(rowranges) == rownames(counts),
          rownames(coldata) == colnames(counts))

library("SummarizedExperiment")

se <- SummarizedExperiment(
    assays = list(counts = as.matrix(counts)),
    rowRanges = as(rowranges, "GRanges"),
    colData = coldata
)
```


A brief recap of how to access the various data slots in a `SummarizedExperiment`:

```{r}
# Access the counts

head(assay(se))
dim(assay(se))

# Access the sample annotations

colData(se)
dim(colData(se))

# Access the gene annotations

head(rowData(se))
dim(rowData(se))

```



## Save SummarizedExperiment

This was a bit of code and time to create our `SummarizedExperiment` object. We will need to keep using it throughout the workshop, so it can be useful to save it as an actual single file on our computer to read it back in to R's memory if we have to shut down RStudio. To save an R-specific file  we can use the `saveRDS()` function and later read it back into R using the `readRDS()` function. 

```{r}
saveRDS(se, "data/GSE96870_se.rds")
rm(se) # remove the object!
se <- readRDS("data/GSE96870_se.rds")
```

:::::::::::::::::::::::::::::::::::::::::  callout

## Important!

Saving an object as an .rds file and reading it back in can be a convenient time saver for an extended analysis. However, there is no record in the object or .rds file of how it was made or manipulated. Your code to make the .rds file can become lost or corrupted, such that you cannot create the same object again from input data files. If this happens you should consider the .rds file as lost/corrupted as well. For reproducibility, the codes are more important than the actual .rds file.

::::::::::::::::::::::::::::::::::::::::::::::::::


## Session info

```{r}
sessionInfo()
```

::: keypoints
-   Depending on the gene expression quantification tool used, there are different ways (often distributed in Bioconductor packages) to read the output into a SummarizedExperiment or DGEList object for further processing in R.
-   Stable gene identifiers such as Ensembl or Entrez IDs should preferably be used as the main identifiers throughout an RNA-seq analysis, with gene symbols added for easier interpretation.
:::
